{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge genes.\n",
    "\"\"\"\n",
    "import os\n",
    "from data_preparation import merge_csv\n",
    "for c in range(4, 24):\n",
    "    chr = f\"chr{c+1}\"\n",
    "    src_dir = os.path.join(\"workspace\", \"seq2seq-stride.384\", chr)\n",
    "    dest_file = os.path.join(\"workspace\", \"seq2seq-stride.384\", chr, \"bundle.csv\")\n",
    "    src_files = [os.path.join(src_dir, fname) for fname in os.listdir(src_dir)] \n",
    "    status = merge_csv(src_files, dest_file)\n",
    "    if not status:\n",
    "        print(f\"Something wrong with merging files in directory {src_dir}.\")\n",
    "        raise Exception(\"something wrong.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Make sample from bundle.csv.\n",
    "\"\"\"\n",
    "from utils.utils import create_n_sample\n",
    "import os\n",
    "\n",
    "n_sample = 100\n",
    "for c in range(24):\n",
    "    bundle_csv = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.csv\")\n",
    "    bundle_sample_csv = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.sample.csv\")\n",
    "    create_n_sample(bundle_csv, n_sample, bundle_sample_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Make train, validation, and test bundle for each chr.\n",
    "\"\"\"\n",
    "from utils.utils import split_and_store_csv\n",
    "import os\n",
    "\n",
    "fractions = [0.7, 0.2, 0.1]\n",
    "file_types = [\"train\", \"validation\", \"test\"]\n",
    "dest_dir = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\") for c in range(24)]\n",
    "src_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.csv\") for c in range(24)]\n",
    "dest_train_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.train.csv\") for c in range(24)]\n",
    "dest_valid_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.validation.csv\") for c in range(24)]\n",
    "dest_test_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.test.csv\") for c in range(24)]\n",
    "\n",
    "for src, dest_train, dest_valid, dest_test in zip(src_files, dest_train_files, dest_valid_files, dest_test_files):\n",
    "    split_and_store_csv(src, fractions, [\n",
    "        dest_train, dest_valid, dest_test\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle.sample.csv from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "chr_bundle_sample_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.sample.csv\") for c in range(24)]\n",
    "merged_bundle_sample = os.path.join(\"workspace\", \"seq2seq-stride.384\", \"bundle.sample.csv\")\n",
    "merge_csv(chr_bundle_sample_csvs, merged_bundle_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "chr_bundle_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.csv\") for c in range(24)]\n",
    "merged_bundle = os.path.join(\"workspace\", \"seq2seq-stride.384\", \"bundle.csv\")\n",
    "merge_csv(chr_bundle_csvs, merged_bundle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle train, validation, and test from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "\n",
    "for t in [\"train\", \"validation\", \"test\"]:\n",
    "    chr_bundle_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", f\"bundle.{t}.csv\") for c in range(24)]\n",
    "    merged_bundle = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"bundle.{t}.csv\")\n",
    "    merge_csv(chr_bundle_csvs, merged_bundle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle training, validation, and test from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "for t in [\"train\", \"validation\", \"test\"]:\n",
    "    chr_bundle_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", f\"bundle.{t}.csv\") for c in range(24)]\n",
    "    merged_bundle = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"bundle.{t}.csv\")\n",
    "    merge_csv(chr_bundle_csvs, merged_bundle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_preparation import kmer\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "_cols = ['sequence', 'label']\n",
    "_stride = 256\n",
    "for c in ['chr{}'.format(i+1) for i in range(24)]:\n",
    "    genes_dir_by_chr = os.path.join('data', 'genome', 'sequential-labelling-positive-strand', c)\n",
    "    genes_expansion_dir_by_chr = os.path.join('workspace', 'sequential-labelling-stride.256', c)\n",
    "    if not os.path.exists(genes_expansion_dir_by_chr):\n",
    "        os.makedirs(genes_expansion_dir_by_chr, exist_ok=True)\n",
    "    for gene_file in os.listdir(genes_dir_by_chr):\n",
    "        gene_file_path = os.path.join(genes_dir_by_chr, gene_file)\n",
    "        gene_expansion_file_path = os.path.join(genes_expansion_dir_by_chr, f\"{gene_file.split('.')[0]}.expanded.csv\")\n",
    "        if os.path.exists(gene_expansion_file_path):\n",
    "            os.remove(gene_expansion_file_path)\n",
    "        target_file = open(gene_expansion_file_path, 'x')\n",
    "        target_file.write(f\"sequence,label\\n\")\n",
    "        df = pd.read_csv(gene_file_path)\n",
    "        print(f\"Working on {c} {gene_file_path}                                 \", end='\\r')\n",
    "        for _, row in df.iterrows():\n",
    "            seq_chunks = kmer(row['sequence'].strip(), _stride)\n",
    "            label_chunks = kmer(row['label'].strip(), _stride)\n",
    "            for seq, label in zip(seq_chunks, label_chunks):\n",
    "                target_file.write(f\"{seq},{label}\\n\")\n",
    "        target_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Filter index based on gene name.\n",
    "\"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "_chr_indices = [\n",
    "    'NC_000001.11.csv',\n",
    "    'NC_000002.12.csv',\n",
    "    'NC_000003.12.csv',\n",
    "    'NC_000004.12.csv',\n",
    "    'NC_000005.10.csv',\n",
    "    'NC_000006.12.csv',\n",
    "    'NC_000007.14.csv',\n",
    "    'NC_000008.11.csv',\n",
    "    'NC_000009.12.csv',\n",
    "    'NC_000010.11.csv',\n",
    "    'NC_000011.10.csv',\n",
    "    'NC_000012.12.csv',\n",
    "    'NC_000013.11.csv',\n",
    "    'NC_000014.9.csv',\n",
    "    'NC_000015.10.csv',\n",
    "    'NC_000016.10.csv',\n",
    "    'NC_000017.11.csv',\n",
    "    'NC_000018.10.csv',\n",
    "    'NC_000019.10.csv',\n",
    "    'NC_000020.11.csv',\n",
    "    'NC_000021.9.csv',\n",
    "    'NC_000022.11.csv',\n",
    "    'NC_000023.11.csv',\n",
    "    'NC_000024.10.csv']\n",
    "_chr_dir = [\"chr{}\".format(i+1) for i in range(len(_chr_indices))]\n",
    "\n",
    "for chr, chr_number in zip(_chr_indices, _chr_dir):\n",
    "    path = os.path.join('data', 'genome', 'grch38', 'csvs_strand', chr)\n",
    "    df = pd.read_csv(path)\n",
    "    genes = list(df['gene'].unique())\n",
    "    genes = [a for a in genes if not pd.isnull(a)]\n",
    "    for g in genes:\n",
    "        ndf = df[df['gene'] == g]\n",
    "        path = os.path.join('data', 'genome', 'grch38', 'genes', chr_number,\"{}.csv\".format(g))\n",
    "        if not os.path.exists(os.path.dirname(path)):\n",
    "            os.mkdir(os.path.dirname(path))\n",
    "        ndf.to_csv(path, index=False)\n",
    "        print(\"Success: {}                                                      \".format(path), end=\"\\r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Generate sequence from fasta based on gene indices.\n",
    "\"\"\"\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from data_dir import chr_fasta_mapname\n",
    "\n",
    "strand = '+'\n",
    "chrs = ['chr{}'.format(i+1) for i in range(24)] # Folder name.\n",
    "# chrs = ['chr{}'.format(i+1) for i in range(2)] # Folder name.\n",
    "\n",
    "for c in chrs:\n",
    "    chr_genes_indices_dir = os.path.join(\"data\", \"genome\", \"grch38\", \"genes\", c)\n",
    "    chr_genes_sequence_csv_dir = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\",c)\n",
    "    chr_fasta = os.path.join(\"data\", \"chr\", chr_fasta_mapname[c])\n",
    "\n",
    "#chr1_genes_indices_dir = os.path.join(\"data\", \"genome\", \"grch38\", \"genes\", \"chr1\")\n",
    "#chr1_genes_sequence_csv_dir = os.path.join(\"data\", \"genome\", \"sequential-labelling\", \"chr1\")\n",
    "#chr1_fasta = os.path.join(\"data\", \"chr\", \"NC_000001.11.fasta\")\n",
    "\n",
    "    if not os.path.exists(chr_fasta):\n",
    "        raise FileNotFoundError(\"Fasta {} not found.\".format(chr_fasta))\n",
    "\n",
    "    \"\"\"\n",
    "    Read whole chromosome here and return its complete sequence.\n",
    "    Yes, it's long.\n",
    "    \"\"\"\n",
    "    records = SeqIO.parse(chr_fasta, \"fasta\")\n",
    "    chr_records = next(records)\n",
    "    chr_sequence = str(chr_records.seq)\n",
    "    genome_sequence = chr_sequence # Use genome sequence reader here.\n",
    "\n",
    "    _columns = ['sequence', 'label']\n",
    "    for fname in os.listdir(chr_genes_indices_dir):\n",
    "        print(\"Working on chr {} gene {}                                    \".format(c, fname), end='\\r')\n",
    "        fpath = os.path.join(chr_genes_indices_dir, fname)\n",
    "        if os.path.isfile(fpath):\n",
    "            \"\"\"\n",
    "            Gene index found. Read the index and cross-reference with genome sequence.\n",
    "            \"\"\"\n",
    "            index_df = pd.read_csv(fpath)\n",
    "            gene_region = index_df[index_df['region'] == \"gene\"]\n",
    "            if strand != None:\n",
    "                gene_region = gene_region[gene_region['strand'] == strand]\n",
    "\n",
    "            if len(gene_region) > 0:\n",
    "                gene_df = pd.DataFrame(columns=_columns)\n",
    "                for i, g in gene_region.iterrows():\n",
    "                    gene_start_index = int(g['start_index'])\n",
    "                    gene_end_index = int(g['end_index'])\n",
    "                    gene_sequence = genome_sequence[gene_start_index:gene_end_index + 1]\n",
    "                    # print(\"Gene sequence: {}\".format(gene_sequence))\n",
    "                    # If gene sequence isn't None then this gene is available in chromosome sequence.\n",
    "                    if gene_sequence != None:\n",
    "                        gene_sequential_labelling = ['i' for a in gene_sequence]\n",
    "                        exons = index_df[index_df['region'] == \"exon\"]\n",
    "                        #if strand != None:\n",
    "                        #    exons = index_df[index_df['strand'] == strand]\n",
    "                        for j, r in exons.iterrows():\n",
    "                            start_index = int(r['start_index'])\n",
    "                            end_index = int(r['end_index'])\n",
    "                            if (start_index >= gene_start_index and end_index <= gene_end_index):\n",
    "                                start_index = int(r['start_index']) - gene_start_index\n",
    "                                end_index = int(r['end_index']) - gene_start_index\n",
    "                                for k in range(start_index, end_index + 1):\n",
    "                                    gene_sequential_labelling[k] = 'E'\n",
    "                        gene_sequential_labelling = ''.join(gene_sequential_labelling)\n",
    "                        #endfor\n",
    "                        gene_df = pd.concat([gene_df, pd.DataFrame([[gene_sequence, gene_sequential_labelling]], columns=_columns)])\n",
    "\n",
    "                target_path = os.path.join(chr_genes_sequence_csv_dir, fname)\n",
    "                # data\\genome\\sequential-labelling\\chr1\n",
    "                # Only write of dataframe is not empty.\n",
    "                if gene_df.shape[0] > 0:\n",
    "                    if not os.path.exists(chr_genes_sequence_csv_dir):\n",
    "                        os.makedirs(chr_genes_sequence_csv_dir, exist_ok=True)\n",
    "                    if gene_df.shape[0] > 0:    \n",
    "                        gene_df.to_csv(target_path, index=False)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Chunk each gene into 512 characters, for each chromosome.\"\"\"\n",
    "import pandas as pd\n",
    "from data_preparation import kmer\n",
    "import os\n",
    "\n",
    "chrs = [f\"chr{i + 1}\" for i in range(5, 24, 1)] # Test one chromosome.\n",
    "chr_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", chr) for chr in chrs]\n",
    "for cp in chr_paths:\n",
    "    genes = os.listdir(cp)\n",
    "    genes = [g for g in genes if \"expanded\" not in g.split('.')]\n",
    "    for gene in genes:\n",
    "        gene_path = os.path.join(cp, gene)\n",
    "        gene_df = pd.read_csv(gene_path)\n",
    "        print(f\"Processing {os.path.basename(cp)} {gene}                    \", end=\"\\r\")\n",
    "        for i, r in gene_df.iterrows():\n",
    "            sequence = r[\"sequence\"]\n",
    "            label = r[\"label\"]\n",
    "            seq_chunks = kmer(sequence, 512, 256)\n",
    "            label_chunks = kmer(label, 512, 256)\n",
    "            dest_path = os.path.join(cp, f\"{gene.split('.')[0]}.expanded.csv\")\n",
    "            if os.path.exists(dest_path):\n",
    "                os.remove(dest_path)\n",
    "            dest = open(dest_path, \"x\")\n",
    "            dest.write(\"sequence,label\\n\")\n",
    "            for c, l in zip(seq_chunks, label_chunks):\n",
    "                dest.write(f\"{c},{l}\\n\")\n",
    "            dest.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "for i in range(24):\n",
    "    k = i + 1\n",
    "    src_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{k}\")\n",
    "    dest_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{k}.expanded\")\n",
    "    if not os.path.exists(dest_path):\n",
    "        os.makedirs(dest_path, exist_ok=True)\n",
    "    files = os.listdir(src_path)\n",
    "    files = [f for f in files if \"expanded\" in f.split('.')]\n",
    "    for f in files:\n",
    "        oldpath = os.path.join(src_path, f)\n",
    "        newpath = os.path.join(dest_path, f)\n",
    "        if not os.path.exists(newpath):\n",
    "            os.rename(oldpath, newpath)\n",
    "        else:\n",
    "            print(f\"Skip {oldpath}                                                    \", end=\"\\r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Rename files in expanded folder\n",
    "filename.expanded.csv => filename.csv\n",
    "\"\"\"\n",
    "src_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i + 1}.expanded\") for i in range(24)]\n",
    "for srcdir in src_paths:\n",
    "    files = os.listdir(srcdir)\n",
    "    for f in files:\n",
    "        oldpath = os.path.join(srcdir, f)\n",
    "        newpath = os.path.join(srcdir, f\"{f.split('.')[0]}.csv\")\n",
    "        os.rename(oldpath, newpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Generate kmer version from expanded sequence.\"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "from data_preparation import str_kmer\n",
    "\n",
    "src_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i + 1}.expanded\") for i in range(24)]\n",
    "dest_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i + 1}.expanded.kmer\") for i in range(24)]\n",
    "\n",
    "for srcdir, destdir in zip(src_paths, dest_paths):\n",
    "    files = os.listdir(srcdir)\n",
    "    for f in files:\n",
    "        src = os.path.join(srcdir, f)\n",
    "        dest = os.path.join(destdir, f)\n",
    "        if not os.path.exists(destdir):\n",
    "            os.makedirs(destdir, exist_ok=True)\n",
    "        if os.path.exists(dest):\n",
    "            os.remove(dest)\n",
    "        dest = open(dest, \"x\")\n",
    "        dest.write(\"sequence,label\\n\")\n",
    "        df = pd.read_csv(src)\n",
    "        for i, r in df.iterrows():\n",
    "            sequence = r[\"sequence\"]\n",
    "            label = r[\"label\"]\n",
    "            dest.write(f\"{str_kmer(sequence, 3)},{str_kmer(label, 3)}\\n\")\n",
    "        dest.close()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Generate gene index in CSV ('chr', 'gene') to see if dataset split can be done by genes.\n",
    "\"\"\"\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "strand = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\")\n",
    "chrs = [os.path.join(strand, f\"chr{i+1}\") for i in range(24)]\n",
    "_cols = [\"chr\", \"gene\"]\n",
    "df = pd.DataFrame(columns=_cols)\n",
    "for chrdir in chrs:\n",
    "    files = os.listdir(chrdir)\n",
    "    for f in files:\n",
    "        df = pd.concat([df, \n",
    "            pd.DataFrame([[f\"{os.path.basename(chrdir)}.expanded\", f]], columns=_cols)\n",
    "        ])\n",
    "gene_index = os.path.join(strand, f\"gene_index.csv\")\n",
    "if os.path.exists(gene_index):\n",
    "    os.remove(gene_index)\n",
    "df.to_csv(gene_index, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Split gene_index for training, validation, and testing.\n",
    "\"\"\"\n",
    "import os\n",
    "strand = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\")\n",
    "gene_index = os.path.join(strand, f\"gene_index.csv\")\n",
    "train_index = os.path.join(strand, \"gene_train_index.csv\")\n",
    "val_index = os.path.join(strand, \"gene_validation_index.csv\")\n",
    "test_index = os.path.join(strand, \"gene_test_index.csv\")\n",
    "\n",
    "import pandas as pd\n",
    "df = pd.read_csv(gene_index)\n",
    "train_df = df.sample(frac=0.8)\n",
    "test_df = df.drop(train_df.index)\n",
    "val_df = train_df.sample(frac=0.8)\n",
    "train_df = train_df.drop(val_df.index)\n",
    "train_df.to_csv(train_index, index=False)\n",
    "val_df.to_csv(val_index, index=False)\n",
    "test_df.to_csv(test_index, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "_cols = ['chr', 'gene']\n",
    "df_path = os.path.join('workspace', 'sequential-labelling', 'gene_index.csv')\n",
    "_dir = os.path.join('workspace', 'sequential-labelling', 'duplicate genes')\n",
    "os.makedirs(_dir, exist_ok=True)\n",
    "df = pd.read_csv(df_path)\n",
    "train_df = pd.DataFrame(columns=_cols)\n",
    "valid_df = pd.DataFrame(columns=_cols)\n",
    "test_df = pd.DataFrame(columns=_cols)\n",
    "genes_unique = df['gene'].unique()\n",
    "for g in genes_unique:\n",
    "    filtered_df = df[df['gene'] == g]\n",
    "    if filtered_df.shape[0] > 1:\n",
    "        _g = \"{}.csv\".format(g)\n",
    "        filtered_df.to_csv(os.path.join(_dir, _g), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_preparation import gff_to_csvs, gff_to_csv\n",
    "from data_dir import (annotated_grch38_gff, annotated_grch38_gff_dir, annotated_grch38_gff_csv)\n",
    "\n",
    "print(annotated_grch38_gff)\n",
    "print(annotated_grch38_gff_csv)\n",
    "print(annotated_grch38_gff_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_dir import chr21_fasta, chr21_index_csv, data_genome_grch38_labels_dir\n",
    "from data_preparation import generate_sequence_labelling\n",
    "import os\n",
    "\n",
    "target_path = os.path.join(data_genome_grch38_labels_dir, 'chr21.csv')\n",
    "print(\"Generate sequential labelling {} => {}: {}\".format(chr21_index_csv, target_path, generate_sequence_labelling(chr21_index_csv, chr21_fasta, target_path)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_dir import (\n",
    "    chr1_index_csv, chr2_index_csv, chr3_index_csv, chr4_index_csv, chr5_index_csv, chr6_index_csv, chr7_index_csv, chr8_index_csv, chr9_index_csv, chr10_index_csv,\n",
    "    chr11_index_csv, chr12_index_csv, chr13_index_csv, chr14_index_csv, chr15_index_csv, chr16_index_csv, chr17_index_csv, chr18_index_csv, chr19_index_csv, chr20_index_csv,\n",
    "    chr21_index_csv, chr22_index_csv, chr23_index_csv, chr24_index_csv\n",
    ")\n",
    "from data_dir import (\n",
    "    chr1_fasta, chr2_fasta, chr3_fasta, chr4_fasta, chr5_fasta, chr6_fasta, chr7_fasta, chr8_fasta, chr9_fasta, chr10_fasta, \n",
    "\tchr11_fasta, chr12_fasta, chr13_fasta, chr14_fasta, chr15_fasta, chr16_fasta, chr17_fasta, chr18_fasta, chr19_fasta, chr20_fasta, \n",
    "\tchr21_fasta, chr22_fasta, chr23_fasta, chr24_fasta,\n",
    ")\n",
    "chr_fastas = [\n",
    "    chr1_fasta,\n",
    "\t#chr2_fasta, chr3_fasta, chr4_fasta, chr5_fasta, chr6_fasta, chr7_fasta, chr8_fasta, chr9_fasta, chr10_fasta,\n",
    "\t#chr11_fasta, chr12_fasta, chr13_fasta, chr14_fasta, chr15_fasta, chr16_fasta, chr17_fasta, chr18_fasta, chr19_fasta, chr20_fasta,\n",
    "\t#chr21_fasta, chr22_fasta, chr23_fasta, chr24_fasta\n",
    "]\n",
    "from data_dir import labseq_dir, labseq_names\n",
    "from data_preparation import generate_sequence_labelling\n",
    "chr_indices = [\n",
    "    chr1_index_csv, \n",
    "\t#chr2_index_csv, chr3_index_csv, chr4_index_csv, chr5_index_csv, chr6_index_csv, chr7_index_csv, chr8_index_csv, chr9_index_csv, chr10_index_csv,\n",
    "    #chr11_index_csv, chr12_index_csv, chr13_index_csv, chr14_index_csv, chr15_index_csv, chr16_index_csv, chr17_index_csv, chr18_index_csv, chr19_index_csv, chr20_index_csv,\n",
    "    #chr21_index_csv, chr22_index_csv, chr23_index_csv, chr24_index_csv\n",
    "]\n",
    "chr_labseq_path = [os.path.join(labseq_dir, fname) for fname in labseq_names[0:1]]\n",
    "for src, fasta, target in zip(chr_indices, chr_fastas, chr_labseq_path):\n",
    "    print(src, fasta, target)\n",
    "    #print(\"Generating sequential labelling for index {}, from fasta {}, to {}: {}\".format(src, fasta, target, generate_sequence_labelling(src, fasta, target, do_expand=True, expand_size=512)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sequential_labelling import Label_Dictionary\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "\n",
    "def prepare_sequence_from_csv(src_csv, label_dictionary=Label_Dictionary):\n",
    "    \"\"\"\n",
    "    Convert sequence into tokenized DNA sequence and label sequence.\n",
    "    CSV source has columns `sequence` and `label`. \n",
    "    `sequence` contains tokenized DNA sequence and `label` contains sequence of labels.\n",
    "    @param      src_csv (string): path to CSV source.\n",
    "    @param      label_dictionary (dict): dictionary to convert label into number.\n",
    "    @return     sequence, labels\n",
    "    \"\"\"\n",
    "    if not os.path.exists(src_csv):\n",
    "        raise FileNotFoundError(src_csv)\n",
    "    if label_dictionary == None:\n",
    "        raise Exception(\"Argument `label_dictionary` cannot be empty!\")\n",
    "    \n",
    "    df = pd.read_csv(src_csv)\n",
    "    return list(df['sequence']), list(df['label'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Create bundle from all expanded sequence from all genes from all chromosomes.\"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "src_dirs = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i+1}.expanded\") for i in range(24)]\n",
    "bundle_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", \"bundle.csv\")\n",
    "if os.path.exists(bundle_path):\n",
    "    os.remove(bundle_path)\n",
    "bundle = open(bundle_path, \"x\")\n",
    "bundle.write(\"sequence,label\\n\")\n",
    "for srcdir in tqdm(src_dirs, total=24):\n",
    "    files = [os.path.join(srcdir, f) for f in os.listdir(srcdir)]\n",
    "    for f in files:\n",
    "        df = pd.read_csv(f)\n",
    "        for i, r in df.iterrows():\n",
    "            bundle.write(f\"{r['sequence']},{r['label']}\\n\")\n",
    "bundle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Split the bundle into training, validation, and testing bundle.\"\"\"\n",
    "import pandas as pd\n",
    "import os\n",
    "bundle_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", \"bundle.csv\")\n",
    "train_bundle_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", \"train_bundle.csv\")\n",
    "val_bundle_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", \"val_bundle.csv\")\n",
    "test_bundle_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", \"test_bundle.csv\")\n",
    "bundle_df = pd.read_csv(bundle_path)\n",
    "train_bundle_df = bundle_df.sample(frac=0.8)\n",
    "test_bundle_df = bundle_df.drop(train_bundle_df.index)\n",
    "val_bundle_df = train_bundle_df.sample(frac=0.2)\n",
    "train_bundle_df = train_bundle_df.drop(val_bundle_df.index)\n",
    "train_bundle_df.to_csv(train_bundle_path)\n",
    "val_bundle_df.to_csv(val_bundle_path)\n",
    "test_bundle_df.to_csv(test_bundle_path)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "473c7453bcb969eece5b07ef8b7f234e7c84010927f6bebce35f0aeb1f8c121e"
  },
  "kernelspec": {
   "display_name": "Python 3.9.1 64-bit ('sequence-processing-py39': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
