{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating index: 100%|██████████| 24/24 [06:53<00:00, 17.21s/it]\n"
     ]
    }
   ],
   "source": [
    "# Generate gene index.\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "def generate_gene_index(gene_dir, index_dir):\n",
    "    chr_dirs = [f\"chr{i+1}\" for i in range(0, 24)]\n",
    "    chrs_path = [os.path.join(gene_dir, a) for a in chr_dirs]\n",
    "    \n",
    "    index_path = os.path.join(index_dir, \"gene_index.csv\")\n",
    "    os.makedirs(os.path.dirname(index_path), exist_ok=True)\n",
    "    if os.path.exists(index_path):\n",
    "        os.remove(index_path)\n",
    "    index = open(index_path, \"x\")\n",
    "    index.write(\"chr,gene,size\\n\")\n",
    "    for chr_dir in tqdm(chrs_path, total=len(chrs_path), desc=\"Creating index\"):\n",
    "        file_names = [a for a in os.listdir(chr_dir) if os.path.isfile(os.path.join(chr_dir, a))]\n",
    "        for fname in file_names:\n",
    "            # Count gene length.\n",
    "            fpath = os.path.join(chr_dir, fname)\n",
    "            df = pd.read_csv(fpath)\n",
    "            len_sequence = 0\n",
    "            for i, r in df.iterrows():                \n",
    "                len_sequence += len(r[\"sequence\"])\n",
    "\n",
    "            index.write(f\"{os.path.basename(chr_dir)},{fname},{len_sequence}\\n\")\n",
    "    index.close()\n",
    "\n",
    "gene_dir = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\")\n",
    "index_dir = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\")\n",
    "generate_gene_index(gene_dir, index_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge genes.\n",
    "\"\"\"\n",
    "import os\n",
    "from data_preparation import merge_csv\n",
    "for c in range(4, 24):\n",
    "    chr = f\"chr{c+1}\"\n",
    "    src_dir = os.path.join(\"workspace\", \"seq2seq-stride.384\", chr)\n",
    "    dest_file = os.path.join(\"workspace\", \"seq2seq-stride.384\", chr, \"bundle.csv\")\n",
    "    src_files = [os.path.join(src_dir, fname) for fname in os.listdir(src_dir)] \n",
    "    status = merge_csv(src_files, dest_file)\n",
    "    if not status:\n",
    "        print(f\"Something wrong with merging files in directory {src_dir}.\")\n",
    "        raise Exception(\"something wrong.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Make sample from bundle.csv.\n",
    "\"\"\"\n",
    "from utils.utils import create_n_sample\n",
    "import os\n",
    "\n",
    "n_sample = 100\n",
    "for c in range(24):\n",
    "    bundle_csv = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.csv\")\n",
    "    bundle_sample_csv = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.sample.csv\")\n",
    "    create_n_sample(bundle_csv, n_sample, bundle_sample_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Make train, validation, and test bundle for each chr.\n",
    "\"\"\"\n",
    "from utils.utils import split_and_store_csv\n",
    "import os\n",
    "\n",
    "fractions = [0.7, 0.2, 0.1]\n",
    "file_types = [\"train\", \"validation\", \"test\"]\n",
    "dest_dir = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\") for c in range(24)]\n",
    "src_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.csv\") for c in range(24)]\n",
    "dest_train_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.train.csv\") for c in range(24)]\n",
    "dest_valid_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.validation.csv\") for c in range(24)]\n",
    "dest_test_files = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.test.csv\") for c in range(24)]\n",
    "\n",
    "for src, dest_train, dest_valid, dest_test in zip(src_files, dest_train_files, dest_valid_files, dest_test_files):\n",
    "    split_and_store_csv(src, fractions, [\n",
    "        dest_train, dest_valid, dest_test\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle.sample.csv from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "chr_bundle_sample_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.sample.csv\") for c in range(24)]\n",
    "merged_bundle_sample = os.path.join(\"workspace\", \"seq2seq-stride.384\", \"bundle.sample.csv\")\n",
    "merge_csv(chr_bundle_sample_csvs, merged_bundle_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "chr_bundle_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", \"bundle.csv\") for c in range(24)]\n",
    "merged_bundle = os.path.join(\"workspace\", \"seq2seq-stride.384\", \"bundle.csv\")\n",
    "merge_csv(chr_bundle_csvs, merged_bundle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle train, validation, and test from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "\n",
    "for t in [\"train\", \"validation\", \"test\"]:\n",
    "    chr_bundle_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", f\"bundle.{t}.csv\") for c in range(24)]\n",
    "    merged_bundle = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"bundle.{t}.csv\")\n",
    "    merge_csv(chr_bundle_csvs, merged_bundle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Merge bundle training, validation, and test from every chr.\n",
    "\"\"\"\n",
    "from data_preparation import merge_csv\n",
    "import os\n",
    "for t in [\"train\", \"validation\", \"test\"]:\n",
    "    chr_bundle_csvs = [os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"chr{c+1}\", f\"bundle.{t}.csv\") for c in range(24)]\n",
    "    merged_bundle = os.path.join(\"workspace\", \"seq2seq-stride.384\", f\"bundle.{t}.csv\")\n",
    "    merge_csv(chr_bundle_csvs, merged_bundle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_preparation import kmer\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "_cols = ['sequence', 'label']\n",
    "_stride = 256\n",
    "_length = 512\n",
    "for c in ['chr{}'.format(i+1) for i in range(24)]:\n",
    "    genes_dir_by_chr = os.path.join('data', 'genome', 'sequential-labelling-positive-strand', c)\n",
    "    genes_expansion_dir_by_chr = os.path.join('workspace', f\"sequential-labelling-stride.{_stride}\", c)\n",
    "    if not os.path.exists(genes_expansion_dir_by_chr):\n",
    "        os.makedirs(genes_expansion_dir_by_chr, exist_ok=True)\n",
    "    for gene_file in os.listdir(genes_dir_by_chr):\n",
    "        gene_file_path = os.path.join(genes_dir_by_chr, gene_file)\n",
    "        gene_expansion_file_path = os.path.join(genes_expansion_dir_by_chr, f\"{gene_file.split('.')[0]}.expanded.csv\")\n",
    "        if os.path.exists(gene_expansion_file_path):\n",
    "            os.remove(gene_expansion_file_path)\n",
    "        target_file = open(gene_expansion_file_path, 'x')\n",
    "        target_file.write(f\"sequence,label\\n\")\n",
    "        df = pd.read_csv(gene_file_path)\n",
    "        print(f\"Working on {c} {gene_file_path}                                 \", end='\\r')\n",
    "        for _, row in df.iterrows():\n",
    "            seq_chunks = kmer(row['sequence'].strip(), _length, _stride)\n",
    "            label_chunks = kmer(row['label'].strip(), _length, _stride)\n",
    "            for seq, label in zip(seq_chunks, label_chunks):\n",
    "                target_file.write(f\"{seq},{label}\\n\")\n",
    "        target_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Filter index based on gene name.\n",
    "\"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "_chr_indices = [\n",
    "    'NC_000001.11.csv',\n",
    "    'NC_000002.12.csv',\n",
    "    'NC_000003.12.csv',\n",
    "    'NC_000004.12.csv',\n",
    "    'NC_000005.10.csv',\n",
    "    'NC_000006.12.csv',\n",
    "    'NC_000007.14.csv',\n",
    "    'NC_000008.11.csv',\n",
    "    'NC_000009.12.csv',\n",
    "    'NC_000010.11.csv',\n",
    "    'NC_000011.10.csv',\n",
    "    'NC_000012.12.csv',\n",
    "    'NC_000013.11.csv',\n",
    "    'NC_000014.9.csv',\n",
    "    'NC_000015.10.csv',\n",
    "    'NC_000016.10.csv',\n",
    "    'NC_000017.11.csv',\n",
    "    'NC_000018.10.csv',\n",
    "    'NC_000019.10.csv',\n",
    "    'NC_000020.11.csv',\n",
    "    'NC_000021.9.csv',\n",
    "    'NC_000022.11.csv',\n",
    "    'NC_000023.11.csv',\n",
    "    'NC_000024.10.csv']\n",
    "_chr_dir = [\"chr{}\".format(i+1) for i in range(len(_chr_indices))]\n",
    "\n",
    "for chr, chr_number in zip(_chr_indices, _chr_dir):\n",
    "    path = os.path.join('data', 'genome', 'grch38', 'csvs_strand', chr)\n",
    "    df = pd.read_csv(path)\n",
    "    genes = list(df['gene'].unique())\n",
    "    genes = [a for a in genes if not pd.isnull(a)]\n",
    "    for g in genes:\n",
    "        ndf = df[df['gene'] == g]\n",
    "        path = os.path.join('data', 'genome', 'grch38', 'genes', chr_number,\"{}.csv\".format(g))\n",
    "        if not os.path.exists(os.path.dirname(path)):\n",
    "            os.mkdir(os.path.dirname(path))\n",
    "        ndf.to_csv(path, index=False)\n",
    "        print(\"Success: {}                                                      \".format(path), end=\"\\r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on chr chr24 gene ZNF92P1Y.csv                                                     \r"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Generate sequence from fasta based on gene indices.\n",
    "\"\"\"\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from data_dir import chr_fasta_mapname\n",
    "\n",
    "strand = '+'\n",
    "chrs = ['chr{}'.format(i+1) for i in range(24)] # Folder name.\n",
    "# chrs = ['chr{}'.format(i+1) for i in range(2)] # Folder name.\n",
    "\n",
    "for c in chrs:\n",
    "    chr_genes_indices_dir = os.path.join(\"data\", \"genome\", \"grch38\", \"genes\", c)\n",
    "    chr_genes_sequence_csv_dir = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\",c)\n",
    "    chr_fasta = os.path.join(\"data\", \"chr\", chr_fasta_mapname[c])\n",
    "\n",
    "    if not os.path.exists(chr_fasta):\n",
    "        raise FileNotFoundError(\"Fasta {} not found.\".format(chr_fasta))\n",
    "\n",
    "    \"\"\"\n",
    "    Read whole chromosome here and return its complete sequence.\n",
    "    Yes, it's long.\n",
    "    \"\"\"\n",
    "    records = SeqIO.parse(chr_fasta, \"fasta\")\n",
    "    chr_records = next(records)\n",
    "    chr_sequence = str(chr_records.seq)\n",
    "    genome_sequence = chr_sequence # Use genome sequence reader here.\n",
    "\n",
    "    _columns = ['sequence', 'label']\n",
    "    for fname in os.listdir(chr_genes_indices_dir):\n",
    "        print(\"Working on chr {} gene {}                                    \".format(c, fname), end='\\r')\n",
    "        fpath = os.path.join(chr_genes_indices_dir, fname)\n",
    "        if os.path.isfile(fpath):\n",
    "            \"\"\"\n",
    "            Gene index found. Read the index and cross-reference with genome sequence.\n",
    "            \"\"\"\n",
    "            index_df = pd.read_csv(fpath)\n",
    "            gene_region = index_df[index_df['region'] == \"gene\"]\n",
    "            if strand != None:\n",
    "                gene_region = gene_region[gene_region['strand'] == strand]\n",
    "\n",
    "            if len(gene_region) > 0:\n",
    "                gene_df = pd.DataFrame(columns=_columns)\n",
    "                for i, g in gene_region.iterrows():\n",
    "                    gene_start_index = int(g['start_index'])\n",
    "                    gene_end_index = int(g['end_index'])\n",
    "                    gene_sequence = genome_sequence[gene_start_index:gene_end_index + 1]\n",
    "                    # print(\"Gene sequence: {}\".format(gene_sequence))\n",
    "                    # If gene sequence isn't None then this gene is available in chromosome sequence.\n",
    "                    if gene_sequence != None:\n",
    "                        gene_sequential_labelling = ['i' for a in gene_sequence]\n",
    "                        exons = index_df[index_df['region'] == \"exon\"]\n",
    "                        #if strand != None:\n",
    "                        #    exons = index_df[index_df['strand'] == strand]\n",
    "                        for j, r in exons.iterrows():\n",
    "                            start_index = int(r['start_index'])\n",
    "                            end_index = int(r['end_index'])\n",
    "                            if (start_index >= gene_start_index and end_index <= gene_end_index):\n",
    "                                start_index = int(r['start_index']) - gene_start_index\n",
    "                                end_index = int(r['end_index']) - gene_start_index\n",
    "                                for k in range(start_index, end_index + 1):\n",
    "                                    gene_sequential_labelling[k] = 'E'\n",
    "                        gene_sequential_labelling = ''.join(gene_sequential_labelling)\n",
    "                        #endfor\n",
    "                        gene_df = pd.concat([gene_df, pd.DataFrame([[gene_sequence, gene_sequential_labelling]], columns=_columns)])\n",
    "\n",
    "                target_path = os.path.join(chr_genes_sequence_csv_dir, fname)\n",
    "                # data\\genome\\sequential-labelling\\chr1\n",
    "                # Only write if dataframe is not empty.\n",
    "                if gene_df.shape[0] > 0:\n",
    "                    if not os.path.exists(chr_genes_sequence_csv_dir):\n",
    "                        os.makedirs(chr_genes_sequence_csv_dir, exist_ok=True)\n",
    "                    if gene_df.shape[0] > 0:    \n",
    "                        gene_df.to_csv(target_path, index=False)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chr24 ZFY.csv                                       \r"
     ]
    }
   ],
   "source": [
    "\"\"\"Chunk each gene into 512 characters, for each chromosome.\"\"\"\n",
    "import pandas as pd\n",
    "from data_preparation import kmer\n",
    "import os\n",
    "\n",
    "chrs = [f\"chr{i + 1}\" for i in range(24)] # Test one chromosome.\n",
    "chr_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"{chr}\") for chr in chrs]\n",
    "dest_paths = [os.path.join(\"data\", \"genome\", \"seqlab.strand-positive.stride-512\", f\"{chr}\") for chr in chrs]\n",
    "for cp, dp in zip(chr_paths, dest_paths):\n",
    "    genes = os.listdir(cp)\n",
    "    genes = [g for g in genes if \"expanded\" not in g.split('.')]\n",
    "    for gene in genes:\n",
    "        gene_path = os.path.join(cp, gene)\n",
    "        gene_df = pd.read_csv(gene_path)\n",
    "        print(f\"Processing {os.path.basename(cp)} {gene}                    \", end=\"\\r\")\n",
    "        for i, r in gene_df.iterrows():\n",
    "            sequence = r[\"sequence\"]\n",
    "            label = r[\"label\"]\n",
    "            seq_chunks = kmer(sequence, 512, 512)\n",
    "            label_chunks = kmer(label, 512, 512)\n",
    "            dest_path = os.path.join(dp, f\"{gene.split('.')[0]}.csv\")\n",
    "            if os.path.exists(dest_path):\n",
    "                os.remove(dest_path)\n",
    "            if not os.path.exists(os.path.dirname(dest_path)):\n",
    "                os.makedirs(os.path.dirname(dest_path), exist_ok=True)\n",
    "            dest = open(dest_path, \"x\")\n",
    "            dest.write(\"sequence,label\\n\")\n",
    "            for c, l in zip(seq_chunks, label_chunks):\n",
    "                dest.write(f\"{c},{l}\\n\")\n",
    "            dest.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Move file around\"\"\"\n",
    "\n",
    "import os\n",
    "for i in range(24):\n",
    "    k = i + 1\n",
    "    src_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{k}\")\n",
    "    dest_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{k}.expanded\")\n",
    "    if not os.path.exists(dest_path):\n",
    "        os.makedirs(dest_path, exist_ok=True)\n",
    "    files = os.listdir(src_path)\n",
    "    files = [f for f in files if \"expanded\" in f.split('.')]\n",
    "    for f in files:\n",
    "        oldpath = os.path.join(src_path, f)\n",
    "        newpath = os.path.join(dest_path, f)\n",
    "        if not os.path.exists(newpath):\n",
    "            os.rename(oldpath, newpath)\n",
    "        else:\n",
    "            print(f\"Skip {oldpath}                                                    \", end=\"\\r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Rename files in expanded folder\n",
    "filename.expanded.csv => filename.csv\n",
    "\"\"\"\n",
    "src_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i + 1}.expanded\") for i in range(24)]\n",
    "for srcdir in src_paths:\n",
    "    files = os.listdir(srcdir)\n",
    "    for f in files:\n",
    "        oldpath = os.path.join(srcdir, f)\n",
    "        newpath = os.path.join(srcdir, f\"{f.split('.')[0]}.csv\")\n",
    "        os.rename(oldpath, newpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Generate kmer version from sequence.\"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "from data_preparation import str_kmer\n",
    "\n",
    "# src_paths = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i + 1}\") for i in range(24)]\n",
    "src_paths = [os.path.join(\"data\", \"genome\", \"seqlab.strand-positive.stride-512\", f\"chr{i + 1}\") for i in range(24)]\n",
    "# dest_paths = [os.path.join(\"data\", \"genome\", \"seqlab.strand-positive.kmer\", f\"chr{i + 1}\") for i in range(24)]\n",
    "dest_paths = [os.path.join(\"data\", \"genome\", \"seqlab.strand-positive.kmer.stride-512\", f\"chr{i + 1}\") for i in range(24)]\n",
    "for srcdir, destdir in zip(src_paths, dest_paths):\n",
    "    files = os.listdir(srcdir)\n",
    "    for f in files:\n",
    "        src = os.path.join(srcdir, f)\n",
    "        dest = os.path.join(destdir, f)\n",
    "        if not os.path.exists(destdir):\n",
    "            os.makedirs(destdir, exist_ok=True)\n",
    "        if os.path.exists(dest):\n",
    "            os.remove(dest)\n",
    "        dest = open(dest, \"x\")\n",
    "        dest.write(\"sequence,label\\n\")\n",
    "        df = pd.read_csv(src)\n",
    "        for i, r in df.iterrows():\n",
    "            sequence = r[\"sequence\"]\n",
    "            label = r[\"label\"]\n",
    "            dest.write(f\"{str_kmer(sequence, 3)},{str_kmer(label, 3)}\\n\")\n",
    "        dest.close()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chr24 ZFY.csv                                       \r"
     ]
    }
   ],
   "source": [
    "\"\"\"Chunk each gene into 510 token. Since sequence is already in token form, ``kmer`` function cannot be used, had to create another script.\"\"\"\n",
    "import pandas as pd\n",
    "from data_preparation import kmer\n",
    "import os\n",
    "\n",
    "def chunk_kmer_sequence(chunk: str, size: int, stride: int) -> str:\n",
    "    arr = chunk.split(' ')\n",
    "    arr = kmer(arr, size, window_size=stride)\n",
    "    return arr\n",
    "\n",
    "chrs = [f\"chr{i + 1}\" for i in range(24)] # Test one chromosome.\n",
    "chr_paths = [os.path.join(\"data\", \"genome\", \"seqlab.strand-positive.kmer\", f\"{chr}\") for chr in chrs]\n",
    "dest_paths = [os.path.join(\"data\", \"genome\", \"seqlab.strand-positive.kmer.stride-205\", f\"{chr}\") for chr in chrs]\n",
    "for cp, dp in zip(chr_paths, dest_paths):\n",
    "    genes = os.listdir(cp)\n",
    "    genes = [g for g in genes if \"expanded\" not in g.split('.')]\n",
    "    for gene in genes:\n",
    "        gene_path = os.path.join(cp, gene)\n",
    "        gene_df = pd.read_csv(gene_path)\n",
    "        print(f\"Processing {os.path.basename(cp)} {gene}                    \", end=\"\\r\")\n",
    "        for i, r in gene_df.iterrows():\n",
    "            sequence = r[\"sequence\"]\n",
    "            label = r[\"label\"]\n",
    "            seq_chunks = chunk_kmer_sequence(sequence, 510, 205)\n",
    "            label_chunks = chunk_kmer_sequence(label, 510, 205)    \n",
    "            dest_path = os.path.join(dp, f\"{gene.split('.')[0]}.csv\")\n",
    "            if os.path.exists(dest_path):\n",
    "                os.remove(dest_path)\n",
    "            if not os.path.exists(os.path.dirname(dest_path)):\n",
    "                os.makedirs(os.path.dirname(dest_path), exist_ok=True)\n",
    "            dest = open(dest_path, \"x\")\n",
    "            dest.write(\"sequence,label\\n\")\n",
    "            for c, l in zip(seq_chunks, label_chunks):\n",
    "                dest.write(f\"{' '.join(c)},{' '.join(l)}\\n\")\n",
    "            dest.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Split gene_index for training, validation, and testing.\n",
    "\"\"\"\n",
    "import os\n",
    "index_dir = os.path.join(\"index\")\n",
    "gene_index = os.path.join(index_dir, f\"gene_index.csv\")\n",
    "train_index = os.path.join(index_dir, \"gene_train_index.csv\")\n",
    "val_index = os.path.join(index_dir, \"gene_validation_index.csv\")\n",
    "test_index = os.path.join(index_dir, \"gene_test_index.csv\")\n",
    "\n",
    "import pandas as pd\n",
    "df = pd.read_csv(gene_index)\n",
    "train_df = df.sample(frac=0.8)\n",
    "test_df = df.drop(train_df.index)\n",
    "val_df = test_df.sample(frac=0.5)\n",
    "test_df = test_df.drop(val_df.index)\n",
    "train_df.to_csv(train_index, index=False)\n",
    "val_df.to_csv(val_index, index=False)\n",
    "test_df.to_csv(test_index, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   0%|          | 21/19235 [00:00<01:32, 207.83it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 AADACL3.csv\n",
      "chr1 AADACL4.csv\n",
      "chr1 ABCD3.csv\n",
      "chr1 ACADM.csv\n",
      "chr1 ACBD3-AS1.csv\n",
      "chr1 ACKR1.csv\n",
      "chr1 ACOT11.csv\n",
      "chr1 ACTL8.csv\n",
      "chr1 ACTN2.csv\n",
      "chr1 ACTRT2.csv\n",
      "chr1 ADAM15.csv\n",
      "chr1 ADAMTSL4.csv\n",
      "chr1 ADGRL2.csv\n",
      "chr1 ADORA1.csv\n",
      "chr1 ADPRS.csv\n",
      "chr1 AGBL4-AS1.csv\n",
      "chr1 AGL.csv\n",
      "chr1 AGO1.csv\n",
      "chr1 AGO3.csv\n",
      "chr1 AGO4.csv\n",
      "chr1 AGRN.csv\n",
      "chr1 AGTRAP.csv\n",
      "chr1 AHCYL1.csv\n",
      "chr1 AJAP1.csv\n",
      "chr1 AK4.csv\n",
      "chr1 AK5.csv\n",
      "chr1 AKIRIN1.csv\n",
      "chr1 AKR1A1.csv\n",
      "chr1 ALG14-AS1.csv\n",
      "chr1 ALG6.csv\n",
      "chr1 ALPL.csv\n",
      "chr1 AMPD2.csv\n",
      "chr1 AMY1A.csv\n",
      "chr1 AMY1C.csv\n",
      "chr1 AMY2A.csv\n",
      "chr1 AMY2B.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   0%|          | 66/19235 [00:00<01:27, 220.14it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 ANGPTL3.csv\n",
      "chr1 ANGPTL7.csv\n",
      "chr1 ANKRD65-AS1.csv\n",
      "chr1 ANXA9.csv\n",
      "chr1 AP4B1-AS1.csv\n",
      "chr1 APCS.csv\n",
      "chr1 AQP10.csv\n",
      "chr1 ARF1.csv\n",
      "chr1 ARHGAP29-AS1.csv\n",
      "chr1 ARHGEF10L.csv\n",
      "chr1 ARHGEF16.csv\n",
      "chr1 ARHGEF2-AS1.csv\n",
      "chr1 ARHGEF2-AS2.csv\n",
      "chr1 ARID1A.csv\n",
      "chr1 ARMH1.csv\n",
      "chr1 ARTN.csv\n",
      "chr1 ARV1.csv\n",
      "chr1 ASH1L-AS1.csv\n",
      "chr1 ATAD3A.csv\n",
      "chr1 ATAD3B.csv\n",
      "chr1 ATAD3C.csv\n",
      "chr1 ATF3.csv\n",
      "chr1 ATF6.csv\n",
      "chr1 ATG4C.csv\n",
      "chr1 ATP1A1.csv\n",
      "chr1 ATP1A2.csv\n",
      "chr1 ATP1A4.csv\n",
      "chr1 ATP1B1.csv\n",
      "chr1 ATP2B4.csv\n",
      "chr1 ATP5IF1.csv\n",
      "chr1 ATP5PB.csv\n",
      "chr1 ATP6V0B.csv\n",
      "chr1 ATP8B2.csv\n",
      "chr1 ATXN7L2.csv\n",
      "chr1 AXDND1.csv\n",
      "chr1 AZIN2.csv\n",
      "chr1 B3GALT6.csv\n",
      "chr1 B4GALT2.csv\n",
      "chr1 BCAN.csv\n",
      "chr1 BCAR3-AS1.csv\n",
      "chr1 BCL10-AS1.csv\n",
      "chr1 BCL9.csv\n",
      "chr1 BECN2.csv\n",
      "chr1 BGLAP.csv\n",
      "chr1 BLZF1.csv\n",
      "chr1 BMP8A.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   1%|          | 116/19235 [00:00<01:23, 229.71it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 BMP8B-AS1.csv\n",
      "chr1 BNIPL.csv\n",
      "chr1 BOLA1.csv\n",
      "chr1 BRDT.csv\n",
      "chr1 BRINP2.csv\n",
      "chr1 BRINP3-DT.csv\n",
      "chr1 BROX.csv\n",
      "chr1 BSND.csv\n",
      "chr1 BTBD19.csv\n",
      "chr1 BTBD8.csv\n",
      "chr1 BTF3L4.csv\n",
      "chr1 BTG2.csv\n",
      "chr1 C1orf100.csv\n",
      "chr1 C1orf105.csv\n",
      "chr1 C1orf112.csv\n",
      "chr1 C1orf115.csv\n",
      "chr1 C1orf122.csv\n",
      "chr1 C1orf146.csv\n",
      "chr1 C1orf162.csv\n",
      "chr1 C1orf167.csv\n",
      "chr1 C1orf185.csv\n",
      "chr1 C1orf21.csv\n",
      "chr1 C1orf220.csv\n",
      "chr1 C1orf226.csv\n",
      "chr1 C1orf50.csv\n",
      "chr1 C1orf53.csv\n",
      "chr1 C1orf54.csv\n",
      "chr1 C1orf56.csv\n",
      "chr1 C1orf68.csv\n",
      "chr1 C1orf94.csv\n",
      "chr1 C1QA.csv\n",
      "chr1 C1QB.csv\n",
      "chr1 C1QC.csv\n",
      "chr1 C2CD4D-AS1.csv\n",
      "chr1 C4BPA.csv\n",
      "chr1 C4BPB.csv\n",
      "chr1 C8A.csv\n",
      "chr1 CA14.csv\n",
      "chr1 CA6.csv\n",
      "chr1 CACHD1.csv\n",
      "chr1 CACNA1E.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   1%|          | 139/19235 [00:00<01:40, 189.54it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 CACYBP.csv\n",
      "chr1 CADM3.csv\n",
      "chr1 CALML6.csv\n",
      "chr1 CAMK1G.csv\n",
      "chr1 CAMSAP2.csv\n",
      "chr1 CAMTA1.csv\n",
      "chr1 CAP1.csv\n",
      "chr1 CAPN2.csv\n",
      "chr1 CAPN9.csv\n",
      "chr1 CAPZA1.csv\n",
      "chr1 CASQ1.csv\n",
      "chr1 CATSPER4.csv\n",
      "chr1 CATSPERE.csv\n",
      "chr1 CCDC18.csv\n",
      "chr1 CCDC185.csv\n",
      "chr1 CCDC24.csv\n",
      "chr1 CCDC27.csv\n",
      "chr1 CCDC28B.csv\n",
      "chr1 CCDC30.csv\n",
      "chr1 CCN1.csv\n",
      "chr1 CD101.csv\n",
      "chr1 CD160.csv\n",
      "chr1 CD1A.csv\n",
      "chr1 CD1C.csv\n",
      "chr1 CD1D.csv\n",
      "chr1 CD1E.csv\n",
      "chr1 CD2.csv\n",
      "chr1 CD46.csv\n",
      "chr1 CD52.csv\n",
      "chr1 CD53.csv\n",
      "chr1 CD55.csv\n",
      "chr1 CDA.csv\n",
      "chr1 CDC14A.csv\n",
      "chr1 CDC20.csv\n",
      "chr1 CDC42.csv\n",
      "chr1 CDC7.csv\n",
      "chr1 CDC73.csv"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   1%|          | 198/19235 [00:00<01:20, 237.49it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "chr1 CDCA8.csv\n",
      "chr1 CDK18.csv\n",
      "chr1 CDKN2C.csv\n",
      "chr1 CELA2A.csv\n",
      "chr1 CELA2B.csv\n",
      "chr1 CELA3A.csv\n",
      "chr1 CELA3B.csv\n",
      "chr1 CELSR2.csv\n",
      "chr1 CENPF.csv\n",
      "chr1 CENPS-CORT.csv\n",
      "chr1 CENPS.csv\n",
      "chr1 CEP350.csv\n",
      "chr1 CEP85.csv\n",
      "chr1 CEPT1.csv\n",
      "chr1 CFAP107.csv\n",
      "chr1 CFAP57.csv\n",
      "chr1 CFH.csv\n",
      "chr1 CFHR1.csv\n",
      "chr1 CFHR2.csv\n",
      "chr1 CFHR3.csv\n",
      "chr1 CFHR4.csv\n",
      "chr1 CFHR5.csv\n",
      "chr1 CGN.csv\n",
      "chr1 CH17-125A10.2.csv\n",
      "chr1 CHD1L.csv\n",
      "chr1 CHI3L2.csv\n",
      "chr1 CHIA.csv\n",
      "chr1 CHRM3.csv\n",
      "chr1 CHRNB2.csv\n",
      "chr1 CHTOP.csv\n",
      "chr1 CIART.csv\n",
      "chr1 CKS1B.csv\n",
      "chr1 CLCA1.csv\n",
      "chr1 CLCA2.csv\n",
      "chr1 CLCA4.csv\n",
      "chr1 CLCN6.csv\n",
      "chr1 CLCNKA.csv\n",
      "chr1 CLCNKB.csv\n",
      "chr1 CLIC4.csv\n",
      "chr1 CMPK1.csv\n",
      "chr1 CNIH3.csv\n",
      "chr1 CNIH4.csv\n",
      "chr1 CNKSR1.csv\n",
      "chr1 CNN3-DT.csv\n",
      "chr1 CNST.csv\n",
      "chr1 CNTN2.csv\n",
      "chr1 COA6.csv\n",
      "chr1 COG2.csv\n",
      "chr1 COQ8A.csv\n",
      "chr1 CORT.csv\n",
      "chr1 COX20.csv\n",
      "chr1 CPT2.csv\n",
      "chr1 CPTP.csv\n",
      "chr1 CR1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   1%|▏         | 252/19235 [00:01<01:15, 252.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 CR1L.csv\n",
      "chr1 CR2.csv\n",
      "chr1 CRB1.csv\n",
      "chr1 CRCT1.csv\n",
      "chr1 CREB3L4.csv\n",
      "chr1 CROCC.csv\n",
      "chr1 CSF1.csv\n",
      "chr1 CSMD2-AS1.csv\n",
      "chr1 CSRP1-AS1.csv\n",
      "chr1 CTH.csv\n",
      "chr1 CTPS1.csv\n",
      "chr1 CTRC.csv\n",
      "chr1 CTTNBP2NL.csv\n",
      "chr1 CTXND2.csv\n",
      "chr1 CYB561D1.csv\n",
      "chr1 CYP4A22.csv\n",
      "chr1 CYP4B1.csv\n",
      "chr1 CYP4X1.csv\n",
      "chr1 CYP4Z1.csv\n",
      "chr1 DAB1-AS1.csv\n",
      "chr1 DAP3.csv\n",
      "chr1 DARS2.csv\n",
      "chr1 DCAF6.csv\n",
      "chr1 DCAF8-DT.csv\n",
      "chr1 DCDC2B.csv\n",
      "chr1 DCLRE1B.csv\n",
      "chr1 DCST1.csv\n",
      "chr1 DDI2.csv\n",
      "chr1 DDR2.csv\n",
      "chr1 DDX20.csv\n",
      "chr1 DDX59-AS1.csv\n",
      "chr1 DEGS1.csv\n",
      "chr1 DEPDC1-AS1.csv\n",
      "chr1 DESI2.csv\n",
      "chr1 DFFB.csv\n",
      "chr1 DHCR24-DT.csv\n",
      "chr1 DHDDS.csv\n",
      "chr1 DHX9.csv\n",
      "chr1 DIO1.csv\n",
      "chr1 DISC1-IT1.csv\n",
      "chr1 DISC1.csv\n",
      "chr1 DISP1.csv\n",
      "chr1 DISP3.csv\n",
      "chr1 DMAP1.csv\n",
      "chr1 DMBX1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   1%|▏         | 278/19235 [00:01<01:26, 218.86it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 DMRTB1.csv\n",
      "chr1 DNAH14.csv\n",
      "chr1 DNAI3.csv\n",
      "chr1 DNAJB4.csv\n",
      "chr1 DNAJC16.csv\n",
      "chr1 DNAJC6.csv\n",
      "chr1 DNALI1.csv\n",
      "chr1 DNASE2B.csv\n",
      "chr1 DNM3-IT1.csv\n",
      "chr1 DNM3.csv\n",
      "chr1 DPH2.csv\n",
      "chr1 DPYD-AS1.csv\n",
      "chr1 DPYD-AS2.csv\n",
      "chr1 DR1.csv\n",
      "chr1 DRAXIN.csv\n",
      "chr1 DTL.csv\n",
      "chr1 DUSP12.csv\n",
      "chr1 DUSP23.csv\n",
      "chr1 DYNLT5.csv\n",
      "chr1 DYRK3.csv\n",
      "chr1 ECE1-AS1.csv\n",
      "chr1 ECM1.csv\n",
      "chr1 EDARADD.csv\n",
      "chr1 EFCAB14-AS1.csv\n",
      "chr1 EFCAB2.csv\n",
      "chr1 EFCAB7.csv\n",
      "chr1 EFHD2.csv\n",
      "chr1 EFNA1.csv\n",
      "chr1 EFNA3.csv\n",
      "chr1 EFNA4.csv\n",
      "chr1 EIF3I.csv\n",
      "chr1 ELAPOR1.csv\n",
      "chr1 ELAVL4.csv\n",
      "chr1 ELF3.csv\n",
      "chr1 ELOA.csv\n",
      "chr1 EMC1-AS1.csv\n",
      "chr1 ENO1-AS1.csv\n",
      "chr1 EPB41.csv\n",
      "chr1 EPHA2-AS1.csv\n",
      "chr1 EPHA8.csv\n",
      "chr1 EPHB2.csv\n",
      "chr1 EPHX1.csv\n",
      "chr1 EPHX4.csv\n",
      "chr1 EPS15-AS1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   2%|▏         | 342/19235 [00:01<01:11, 262.98it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 ERICH3-AS1.csv\n",
      "chr1 ERLNC1.csv\n",
      "chr1 ERMAP.csv\n",
      "chr1 ESPN.csv\n",
      "chr1 EXO1.csv\n",
      "chr1 EXO5.csv\n",
      "chr1 EXOSC10-AS1.csv\n",
      "chr1 EXTL1.csv\n",
      "chr1 FAAH.csv\n",
      "chr1 FAF1-AS1.csv\n",
      "chr1 FALEC.csv\n",
      "chr1 FAM102B.csv\n",
      "chr1 FAM110D.csv\n",
      "chr1 FAM163A.csv\n",
      "chr1 FAM167B.csv\n",
      "chr1 FAM177B.csv\n",
      "chr1 FAM183A.csv\n",
      "chr1 FAM20B.csv\n",
      "chr1 FAM43B.csv\n",
      "chr1 FAM71A.csv\n",
      "chr1 FAM72D.csv\n",
      "chr1 FAM76A.csv\n",
      "chr1 FAM78B-AS1.csv\n",
      "chr1 FAM87B.csv\n",
      "chr1 FASLG.csv\n",
      "chr1 FBLIM1.csv\n",
      "chr1 FBXO28.csv\n",
      "chr1 FBXO44.csv\n",
      "chr1 FBXO6.csv\n",
      "chr1 FCER1A.csv\n",
      "chr1 FCER1G.csv\n",
      "chr1 FCGR1A.csv\n",
      "chr1 FCGR2A.csv\n",
      "chr1 FCGR2B.csv\n",
      "chr1 FCRL6.csv\n",
      "chr1 FCRLA.csv\n",
      "chr1 FCRLB.csv\n",
      "chr1 FDPS.csv\n",
      "chr1 FGGY.csv\n",
      "chr1 FHAD1.csv\n",
      "chr1 FLAD1.csv\n",
      "chr1 FLG-AS1.csv\n",
      "chr1 FLJ39095.csv\n",
      "chr1 FLVCR1.csv\n",
      "chr1 FMN2.csv\n",
      "chr1 FMO1.csv\n",
      "chr1 FMO2.csv\n",
      "chr1 FMO3.csv\n",
      "chr1 FMO4.csv\n",
      "chr1 FNBP1L.csv\n",
      "chr1 FNDC7.csv\n",
      "chr1 FOXD2.csv\n",
      "chr1 FOXD3.csv\n",
      "chr1 FOXE3.csv\n",
      "chr1 FOXO6.csv\n",
      "chr1 FPGT-TNNI3K.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   2%|▏         | 400/19235 [00:01<01:10, 266.26it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 FPGT.csv\n",
      "chr1 G0S2.csv\n",
      "chr1 GABPB2.csv\n",
      "chr1 GABRD.csv\n",
      "chr1 GADD45A.csv\n",
      "chr1 GALNT2.csv\n",
      "chr1 GAS5-AS1.csv\n",
      "chr1 GBP6.csv\n",
      "chr1 GCSAML.csv\n",
      "chr1 GGPS1.csv\n",
      "chr1 GIPC2.csv\n",
      "chr1 GJA4.csv\n",
      "chr1 GJA8.csv\n",
      "chr1 GJB3.csv\n",
      "chr1 GJB4.csv\n",
      "chr1 GJB5.csv\n",
      "chr1 GJC2.csv\n",
      "chr1 GMEB1.csv\n",
      "chr1 GNAI3.csv\n",
      "chr1 GNB1-DT.csv\n",
      "chr1 GNG12-AS1.csv\n",
      "chr1 GNPAT.csv\n",
      "chr1 GORAB.csv\n",
      "chr1 GPR137B.csv\n",
      "chr1 GPR25.csv\n",
      "chr1 GPR3.csv\n",
      "chr1 GPR37L1.csv\n",
      "chr1 GPR52.csv\n",
      "chr1 GPR61.csv\n",
      "chr1 GPR88.csv\n",
      "chr1 GPR89A.csv\n",
      "chr1 GPR89B.csv\n",
      "chr1 GPSM2.csv\n",
      "chr1 GPX7.csv\n",
      "chr1 GRHL3.csv\n",
      "chr1 GS1-204I12.4.csv\n",
      "chr1 GSTM1.csv\n",
      "chr1 GSTM2.csv\n",
      "chr1 GSTM4.csv\n",
      "chr1 GSTM5.csv\n",
      "chr1 GUCA2B.csv\n",
      "chr1 GUK1.csv\n",
      "chr1 H2AC19.csv\n",
      "chr1 H2AC20.csv\n",
      "chr1 H2BU1.csv\n",
      "chr1 H3-3A.csv\n",
      "chr1 H3C15.csv\n",
      "chr1 H4C14.csv\n",
      "chr1 H6PD.csv\n",
      "chr1 HAO2-IT1.csv\n",
      "chr1 HAO2.csv\n",
      "chr1 HAPLN2.csv\n",
      "chr1 HAX1.csv\n",
      "chr1 HCN3.csv\n",
      "chr1 HCRTR1.csv\n",
      "chr1 HDAC1.csv\n",
      "chr1 HES3.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   2%|▏         | 459/19235 [00:01<01:10, 265.30it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 HHAT.csv\n",
      "chr1 HHLA3.csv\n",
      "chr1 HIPK1.csv\n",
      "chr1 HLX.csv\n",
      "chr1 HMCN1.csv\n",
      "chr1 HMGB4.csv\n",
      "chr1 HMGN2.csv\n",
      "chr1 HNRNPCL3.csv\n",
      "chr1 HOOK1.csv\n",
      "chr1 HPCA.csv\n",
      "chr1 HPDL.csv\n",
      "chr1 HS2ST1.csv\n",
      "chr1 HSD11B1.csv\n",
      "chr1 HSD17B7.csv\n",
      "chr1 HSD3B1.csv\n",
      "chr1 HSD3B2.csv\n",
      "chr1 HSPA6.csv\n",
      "chr1 HTR6.csv\n",
      "chr1 IARS2.csv\n",
      "chr1 IBA57.csv\n",
      "chr1 ICMT-DT.csv\n",
      "chr1 IER5.csv\n",
      "chr1 IFI16.csv\n",
      "chr1 IFI44.csv\n",
      "chr1 IFI44L.csv\n",
      "chr1 IGFN1.csv\n",
      "chr1 IGSF21.csv\n",
      "chr1 IKBKE.csv\n",
      "chr1 IL12RB2.csv\n",
      "chr1 IL19.csv\n",
      "chr1 IL20.csv\n",
      "chr1 IL23R.csv\n",
      "chr1 IL24.csv\n",
      "chr1 IL6R.csv\n",
      "chr1 INAVA.csv\n",
      "chr1 INKA2-AS1.csv\n",
      "chr1 INTS3.csv\n",
      "chr1 IPO13.csv\n",
      "chr1 IPO9.csv\n",
      "chr1 IQCC.csv\n",
      "chr1 ISG15.csv\n",
      "chr1 IVL.csv\n",
      "chr1 KAZN.csv\n",
      "chr1 KCNAB2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   3%|▎         | 486/19235 [00:02<01:12, 257.29it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 KCNC4.csv\n",
      "chr1 KCND3-AS1.csv\n",
      "chr1 KCNJ9.csv\n",
      "chr1 KCNK1.csv\n",
      "chr1 KCNK2.csv\n",
      "chr1 KCNQ4.csv\n",
      "chr1 KCTD3.csv\n",
      "chr1 KDM1A.csv\n",
      "chr1 KDM4A.csv\n",
      "chr1 KHDRBS1.csv\n",
      "chr1 KIAA1522.csv\n",
      "chr1 KIAA1614.csv\n",
      "chr1 KIF1B.csv\n",
      "chr1 KIF26B.csv\n",
      "chr1 KIF2C.csv\n",
      "chr1 KIRREL1-IT1.csv\n",
      "chr1 KIRREL1.csv\n",
      "chr1 KLF17.csv\n",
      "chr1 KLHDC7A.csv\n",
      "chr1 KLHDC9.csv\n",
      "chr1 KLHL17.csv\n",
      "chr1 KLHL20.csv\n",
      "chr1 KMO.csv\n",
      "chr1 KPNA6.csv\n",
      "chr1 KPRP.csv\n",
      "chr1 L1TD1.csv\n",
      "chr1 LAMC1.csv\n",
      "chr1 LAMC2.csv\n",
      "chr1 LAMTOR2.csv\n",
      "chr1 LAMTOR5-AS1.csv\n",
      "chr1 LAX1.csv\n",
      "chr1 LCE1A.csv\n",
      "chr1 LCE1B.csv\n",
      "chr1 LCE1D.csv\n",
      "chr1 LCE1E.csv\n",
      "chr1 LCE1F.csv\n",
      "chr1 LCE2A.csv\n",
      "chr1 LCE2B.csv\n",
      "chr1 LCE2C.csv\n",
      "chr1 LCE2D.csv\n",
      "chr1 LCE3B.csv\n",
      "chr1 LCE3C.csv\n",
      "chr1 LCE4A.csv\n",
      "chr1 LCE5A.csv\n",
      "chr1 LCE6A.csv\n",
      "chr1 LCK.csv\n",
      "chr1 LDLRAD2.csv\n",
      "chr1 LDLRAP1.csv\n",
      "chr1 LELP1.csv\n",
      "chr1 LEMD1-AS1.csv\n",
      "chr1 LEMD1-DT.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   3%|▎         | 544/19235 [00:02<01:08, 271.69it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LENEP.csv\n",
      "chr1 LEPR.csv\n",
      "chr1 LEPROT.csv\n",
      "chr1 LEXM.csv\n",
      "chr1 LGALS8.csv\n",
      "chr1 LGR6.csv\n",
      "chr1 LHX4.csv\n",
      "chr1 LHX8.csv\n",
      "chr1 LHX9.csv\n",
      "chr1 LIN28A.csv\n",
      "chr1 LINC00184.csv\n",
      "chr1 LINC00210.csv\n",
      "chr1 LINC00260.csv\n",
      "chr1 LINC00272.csv\n",
      "chr1 LINC00302.csv\n",
      "chr1 LINC00339.csv\n",
      "chr1 LINC00467.csv\n",
      "chr1 LINC00538.csv\n",
      "chr1 LINC00623.csv\n",
      "chr1 LINC00626.csv\n",
      "chr1 LINC00853.csv\n",
      "chr1 LINC00869.csv\n",
      "chr1 LINC01031.csv\n",
      "chr1 LINC01036.csv\n",
      "chr1 LINC01037.csv\n",
      "chr1 LINC01128.csv\n",
      "chr1 LINC01132.csv\n",
      "chr1 LINC01133.csv\n",
      "chr1 LINC01134.csv\n",
      "chr1 LINC01135.csv\n",
      "chr1 LINC01140.csv\n",
      "chr1 LINC01144.csv\n",
      "chr1 LINC01221.csv\n",
      "chr1 LINC01226.csv\n",
      "chr1 LINC01307.csv\n",
      "chr1 LINC01341.csv\n",
      "chr1 LINC01342.csv\n",
      "chr1 LINC01346.csv\n",
      "chr1 LINC01352.csv\n",
      "chr1 LINC01353.csv\n",
      "chr1 LINC01357.csv\n",
      "chr1 LINC01358.csv\n",
      "chr1 LINC01360.csv\n",
      "chr1 LINC01362.csv\n",
      "chr1 LINC01397.csv\n",
      "chr1 LINC01409.csv\n",
      "chr1 LINC01525.csv\n",
      "chr1 LINC01633.csv\n",
      "chr1 LINC01641.csv\n",
      "chr1 LINC01645.csv\n",
      "chr1 LINC01646.csv\n",
      "chr1 LINC01647.csv\n",
      "chr1 LINC01649.csv\n",
      "chr1 LINC01653.csv\n",
      "chr1 LINC01654.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   3%|▎         | 611/19235 [00:02<01:03, 294.69it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LINC01657.csv\n",
      "chr1 LINC01661.csv\n",
      "chr1 LINC01672.csv\n",
      "chr1 LINC01681.csv\n",
      "chr1 LINC01685.csv\n",
      "chr1 LINC01686.csv\n",
      "chr1 LINC01688.csv\n",
      "chr1 LINC01693.csv\n",
      "chr1 LINC01696.csv\n",
      "chr1 LINC01698.csv\n",
      "chr1 LINC01699.csv\n",
      "chr1 LINC01701.csv\n",
      "chr1 LINC01702.csv\n",
      "chr1 LINC01703.csv\n",
      "chr1 LINC01707.csv\n",
      "chr1 LINC01709.csv\n",
      "chr1 LINC01712.csv\n",
      "chr1 LINC01714.csv\n",
      "chr1 LINC01717.csv\n",
      "chr1 LINC01719.csv\n",
      "chr1 LINC01720.csv\n",
      "chr1 LINC01732.csv\n",
      "chr1 LINC01736.csv\n",
      "chr1 LINC01738.csv\n",
      "chr1 LINC01740.csv\n",
      "chr1 LINC01741.csv\n",
      "chr1 LINC01745.csv\n",
      "chr1 LINC01750.csv\n",
      "chr1 LINC01755.csv\n",
      "chr1 LINC01757.csv\n",
      "chr1 LINC01758.csv\n",
      "chr1 LINC01761.csv\n",
      "chr1 LINC01763.csv\n",
      "chr1 LINC01765.csv\n",
      "chr1 LINC01767.csv\n",
      "chr1 LINC01768.csv\n",
      "chr1 LINC01772.csv\n",
      "chr1 LINC01774.csv\n",
      "chr1 LINC01776.csv\n",
      "chr1 LINC01777.csv\n",
      "chr1 LINC01778.csv\n",
      "chr1 LINC01779.csv\n",
      "chr1 LINC01780.csv\n",
      "chr1 LINC01781.csv\n",
      "chr1 LINC01786.csv\n",
      "chr1 LINC01788.csv\n",
      "chr1 LINC02238.csv\n",
      "chr1 LINC02474.csv\n",
      "chr1 LINC02591.csv\n",
      "chr1 LINC02596.csv\n",
      "chr1 LINC02607.csv\n",
      "chr1 LINC02766.csv\n",
      "chr1 LINC02767.csv\n",
      "chr1 LINC02770.csv\n",
      "chr1 LINC02772.csv\n",
      "chr1 LINC02774.csv\n",
      "chr1 LINC02778.csv\n",
      "chr1 LINC02780.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   4%|▎         | 675/19235 [00:02<01:01, 301.02it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LINC02781.csv\n",
      "chr1 LINC02783.csv\n",
      "chr1 LINC02785.csv\n",
      "chr1 LINC02788.csv\n",
      "chr1 LINC02789.csv\n",
      "chr1 LINC02790.csv\n",
      "chr1 LINC02792.csv\n",
      "chr1 LINC02793.csv\n",
      "chr1 LINC02794.csv\n",
      "chr1 LINC02796.csv\n",
      "chr1 LINC02797.csv\n",
      "chr1 LINC02798.csv\n",
      "chr1 LINC02800.csv\n",
      "chr1 LINC02801.csv\n",
      "chr1 LINC02804.csv\n",
      "chr1 LINC02805.csv\n",
      "chr1 LINC02806.csv\n",
      "chr1 LINC02810.csv\n",
      "chr1 LINC02813.csv\n",
      "chr1 LINC02816.csv\n",
      "chr1 LINC02819.csv\n",
      "chr1 LINC02868.csv\n",
      "chr1 LINC02869.csv\n",
      "chr1 LIX1L-AS1.csv\n",
      "chr1 LMNA.csv\n",
      "chr1 LMO4.csv\n",
      "chr1 LMX1A-AS1.csv\n",
      "chr1 LMX1A-AS2.csv\n",
      "chr1 LNCTAM34A.csv\n",
      "chr1 LOC100288175.csv\n",
      "chr1 LOC100507634.csv\n",
      "chr1 LOC100996583.csv\n",
      "chr1 LOC100996740.csv\n",
      "chr1 LOC100996756.csv\n",
      "chr1 LOC101926907.csv\n",
      "chr1 LOC101926944.csv\n",
      "chr1 LOC101927164.csv\n",
      "chr1 LOC101927247.csv\n",
      "chr1 LOC101927342.csv\n",
      "chr1 LOC101927452.csv\n",
      "chr1 LOC101927604.csv\n",
      "chr1 LOC101927711.csv\n",
      "chr1 LOC101927787.csv\n",
      "chr1 LOC101927871.csv\n",
      "chr1 LOC101927975.csv\n",
      "chr1 LOC101928009.csv\n",
      "chr1 LOC101928034.csv\n",
      "chr1 LOC101928059.csv\n",
      "chr1 LOC101928120.csv\n",
      "chr1 LOC101928163.csv\n",
      "chr1 LOC101928219.csv\n",
      "chr1 LOC101928334.csv\n",
      "chr1 LOC101928372.csv\n",
      "chr1 LOC101928391.csv\n",
      "chr1 LOC101928512.csv\n",
      "chr1 LOC101928596.csv\n",
      "chr1 LOC101929178.csv\n",
      "chr1 LOC101929305.csv\n",
      "chr1 LOC101929343.csv\n",
      "chr1 LOC101929444.csv\n",
      "chr1 LOC101929536.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   4%|▍         | 745/19235 [00:02<00:58, 318.39it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LOC101929626.csv\n",
      "chr1 LOC101929750.csv\n",
      "chr1 LOC101929788.csv\n",
      "chr1 LOC101929805.csv\n",
      "chr1 LOC102606465.csv\n",
      "chr1 LOC102723321.csv\n",
      "chr1 LOC102723348.csv\n",
      "chr1 LOC102723480.csv\n",
      "chr1 LOC102723529.csv\n",
      "chr1 LOC102723760.csv\n",
      "chr1 LOC102723834.csv\n",
      "chr1 LOC102724319.csv\n",
      "chr1 LOC102724382.csv\n",
      "chr1 LOC102724429.csv\n",
      "chr1 LOC102724572.csv\n",
      "chr1 LOC102724830.csv\n",
      "chr1 LOC102724856.csv\n",
      "chr1 LOC102724857.csv\n",
      "chr1 LOC103021295.csv\n",
      "chr1 LOC105371175.csv\n",
      "chr1 LOC105371206.csv\n",
      "chr1 LOC105371209.csv\n",
      "chr1 LOC105371214.csv\n",
      "chr1 LOC105371215.csv\n",
      "chr1 LOC105371217.csv\n",
      "chr1 LOC105371223.csv\n",
      "chr1 LOC105371227.csv\n",
      "chr1 LOC105371233.csv\n",
      "chr1 LOC105371254.csv\n",
      "chr1 LOC105371288.csv\n",
      "chr1 LOC105371406.csv\n",
      "chr1 LOC105371433.csv\n",
      "chr1 LOC105371441.csv\n",
      "chr1 LOC105371446.csv\n",
      "chr1 LOC105371458.csv\n",
      "chr1 LOC105371459.csv\n",
      "chr1 LOC105371460.csv\n",
      "chr1 LOC105371468.csv\n",
      "chr1 LOC105371473.csv\n",
      "chr1 LOC105371600.csv\n",
      "chr1 LOC105371601.csv\n",
      "chr1 LOC105371606.csv\n",
      "chr1 LOC105371608.csv\n",
      "chr1 LOC105371610.csv\n",
      "chr1 LOC105371614.csv\n",
      "chr1 LOC105371619.csv\n",
      "chr1 LOC105371623.csv\n",
      "chr1 LOC105371630.csv\n",
      "chr1 LOC105371642.csv\n",
      "chr1 LOC105371646.csv\n",
      "chr1 LOC105371648.csv\n",
      "chr1 LOC105371649.csv\n",
      "chr1 LOC105371655.csv\n",
      "chr1 LOC105371657.csv\n",
      "chr1 LOC105371658.csv\n",
      "chr1 LOC105371659.csv\n",
      "chr1 LOC105371661.csv\n",
      "chr1 LOC105371668.csv\n",
      "chr1 LOC105371679.csv\n",
      "chr1 LOC105371680.csv\n",
      "chr1 LOC105371686.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   4%|▍         | 777/19235 [00:02<00:58, 314.28it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LOC105371689.csv\n",
      "chr1 LOC105371692.csv\n",
      "chr1 LOC105371694.csv\n",
      "chr1 LOC105371701.csv\n",
      "chr1 LOC105371702.csv\n",
      "chr1 LOC105371729.csv\n",
      "chr1 LOC105372873.csv\n",
      "chr1 LOC105372880.csv\n",
      "chr1 LOC105372884.csv\n",
      "chr1 LOC105372887.csv\n",
      "chr1 LOC105372888.csv\n",
      "chr1 LOC105372889.csv\n",
      "chr1 LOC105372895.csv\n",
      "chr1 LOC105372901.csv\n",
      "chr1 LOC105372912.csv\n",
      "chr1 LOC105372915.csv\n",
      "chr1 LOC105372916.csv\n",
      "chr1 LOC105372923.csv\n",
      "chr1 LOC105372928.csv\n",
      "chr1 LOC105372930.csv\n",
      "chr1 LOC105372932.csv\n",
      "chr1 LOC105372945.csv\n",
      "chr1 LOC105373034.csv\n",
      "chr1 LOC105373061.csv\n",
      "chr1 LOC105373124.csv\n",
      "chr1 LOC105373143.csv\n",
      "chr1 LOC105373164.csv\n",
      "chr1 LOC105373167.csv\n",
      "chr1 LOC105373171.csv\n",
      "chr1 LOC105373172.csv\n",
      "chr1 LOC105373196.csv\n",
      "chr1 LOC105373201.csv\n",
      "chr1 LOC105373206.csv\n",
      "chr1 LOC105373207.csv\n",
      "chr1 LOC105373209.csv\n",
      "chr1 LOC105373211.csv\n",
      "chr1 LOC105373215.csv\n",
      "chr1 LOC105373218.csv\n",
      "chr1 LOC105373220.csv\n",
      "chr1 LOC105373222.csv\n",
      "chr1 LOC105373230.csv\n",
      "chr1 LOC105373231.csv\n",
      "chr1 LOC105373235.csv\n",
      "chr1 LOC105373260.csv\n",
      "chr1 LOC105373269.csv\n",
      "chr1 LOC105373270.csv\n",
      "chr1 LOC105373273.csv\n",
      "chr1 LOC105373289.csv\n",
      "chr1 LOC105373467.csv\n",
      "chr1 LOC105376674.csv\n",
      "chr1 LOC105376682.csv\n",
      "chr1 LOC105376687.csv\n",
      "chr1 LOC105376691.csv\n",
      "chr1 LOC105376695.csv\n",
      "chr1 LOC105376717.csv\n",
      "chr1 LOC105376739.csv\n",
      "chr1 LOC105376740.csv\n",
      "chr1 LOC105376794.csv\n",
      "chr1 LOC105376806.csv\n",
      "chr1 LOC105376808.csv\n",
      "chr1 LOC105376812.csv\n",
      "chr1 LOC105376819.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   4%|▍         | 854/19235 [00:03<00:52, 348.45it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LOC105376825.csv\n",
      "chr1 LOC105376828.csv\n",
      "chr1 LOC105376829.csv\n",
      "chr1 LOC105376830.csv\n",
      "chr1 LOC105376845.csv\n",
      "chr1 LOC105376850.csv\n",
      "chr1 LOC105376863.csv\n",
      "chr1 LOC105376864.csv\n",
      "chr1 LOC105376876.csv\n",
      "chr1 LOC105376885.csv\n",
      "chr1 LOC105376888.csv\n",
      "chr1 LOC105376892.csv\n",
      "chr1 LOC105376910.csv\n",
      "chr1 LOC105378586.csv\n",
      "chr1 LOC105378593.csv\n",
      "chr1 LOC105378598.csv\n",
      "chr1 LOC105378599.csv\n",
      "chr1 LOC105378600.csv\n",
      "chr1 LOC105378601.csv\n",
      "chr1 LOC105378602.csv\n",
      "chr1 LOC105378607.csv\n",
      "chr1 LOC105378608.csv\n",
      "chr1 LOC105378612.csv\n",
      "chr1 LOC105378618.csv\n",
      "chr1 LOC105378625.csv\n",
      "chr1 LOC105378631.csv\n",
      "chr1 LOC105378638.csv\n",
      "chr1 LOC105378639.csv\n",
      "chr1 LOC105378640.csv\n",
      "chr1 LOC105378646.csv\n",
      "chr1 LOC105378649.csv\n",
      "chr1 LOC105378651.csv\n",
      "chr1 LOC105378654.csv\n",
      "chr1 LOC105378658.csv\n",
      "chr1 LOC105378662.csv\n",
      "chr1 LOC105378663.csv\n",
      "chr1 LOC105378665.csv\n",
      "chr1 LOC105378667.csv\n",
      "chr1 LOC105378673.csv\n",
      "chr1 LOC105378675.csv\n",
      "chr1 LOC105378678.csv\n",
      "chr1 LOC105378693.csv\n",
      "chr1 LOC105378699.csv\n",
      "chr1 LOC105378706.csv\n",
      "chr1 LOC105378708.csv\n",
      "chr1 LOC105378710.csv\n",
      "chr1 LOC105378714.csv\n",
      "chr1 LOC105378717.csv\n",
      "chr1 LOC105378721.csv\n",
      "chr1 LOC105378723.csv\n",
      "chr1 LOC105378724.csv\n",
      "chr1 LOC105378726.csv\n",
      "chr1 LOC105378728.csv\n",
      "chr1 LOC105378730.csv\n",
      "chr1 LOC105378733.csv\n",
      "chr1 LOC105378735.csv\n",
      "chr1 LOC105378737.csv\n",
      "chr1 LOC105378745.csv\n",
      "chr1 LOC105378746.csv\n",
      "chr1 LOC105378747.csv\n",
      "chr1 LOC105378748.csv\n",
      "chr1 LOC105378751.csv\n",
      "chr1 LOC105378768.csv\n",
      "chr1 LOC105378770.csv\n",
      "chr1 LOC105378771.csv\n",
      "chr1 LOC105378774.csv\n",
      "chr1 LOC105378776.csv\n",
      "chr1 LOC105378780.csv\n",
      "chr1 LOC105378781.csv\n",
      "chr1 LOC105378782.csv\n",
      "chr1 LOC105378787.csv\n",
      "chr1 LOC105378789.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   5%|▍         | 923/19235 [00:03<00:54, 335.95it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LOC105378790.csv\n",
      "chr1 LOC105378793.csv\n",
      "chr1 LOC105378797.csv\n",
      "chr1 LOC105378805.csv\n",
      "chr1 LOC105378808.csv\n",
      "chr1 LOC105378810.csv\n",
      "chr1 LOC105378814.csv\n",
      "chr1 LOC105378816.csv\n",
      "chr1 LOC105378819.csv\n",
      "chr1 LOC105378821.csv\n",
      "chr1 LOC105378823.csv\n",
      "chr1 LOC105378826.csv\n",
      "chr1 LOC105378833.csv\n",
      "chr1 LOC105378834.csv\n",
      "chr1 LOC105378838.csv\n",
      "chr1 LOC105378839.csv\n",
      "chr1 LOC105378841.csv\n",
      "chr1 LOC105378842.csv\n",
      "chr1 LOC105378847.csv\n",
      "chr1 LOC105378848.csv\n",
      "chr1 LOC105378849.csv\n",
      "chr1 LOC105378860.csv\n",
      "chr1 LOC105378861.csv\n",
      "chr1 LOC105378863.csv\n",
      "chr1 LOC105378866.csv\n",
      "chr1 LOC105378867.csv\n",
      "chr1 LOC105378879.csv\n",
      "chr1 LOC105378883.csv\n",
      "chr1 LOC105378886.csv\n",
      "chr1 LOC105378887.csv\n",
      "chr1 LOC105378890.csv\n",
      "chr1 LOC105378891.csv\n",
      "chr1 LOC105378898.csv\n",
      "chr1 LOC105378906.csv\n",
      "chr1 LOC105378914.csv\n",
      "chr1 LOC105378924.csv\n",
      "chr1 LOC105378925.csv\n",
      "chr1 LOC105378927.csv\n",
      "chr1 LOC105378930.csv\n",
      "chr1 LOC105378936.csv\n",
      "chr1 LOC105378938.csv\n",
      "chr1 LOC105378940.csv\n",
      "chr1 LOC105378941.csv\n",
      "chr1 LOC105378943.csv\n",
      "chr1 LOC105378951.csv\n",
      "chr1 LOC105379574.csv\n",
      "chr1 LOC105379772.csv\n",
      "chr1 LOC105379825.csv\n",
      "chr1 LOC107983960.csv\n",
      "chr1 LOC107984109.csv\n",
      "chr1 LOC107984850.csv\n",
      "chr1 LOC107984872.csv\n",
      "chr1 LOC107984904.csv\n",
      "chr1 LOC107984909.csv\n",
      "chr1 LOC107984917.csv\n",
      "chr1 LOC107984918.csv\n",
      "chr1 LOC107984919.csv\n",
      "chr1 LOC107984929.csv\n",
      "chr1 LOC107984931.csv\n",
      "chr1 LOC107984932.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   5%|▌         | 994/19235 [00:03<00:53, 342.28it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LOC107984933.csv\n",
      "chr1 LOC107984934.csv\n",
      "chr1 LOC107984935.csv\n",
      "chr1 LOC107984940.csv\n",
      "chr1 LOC107984945.csv\n",
      "chr1 LOC107984948.csv\n",
      "chr1 LOC107984951.csv\n",
      "chr1 LOC107984952.csv\n",
      "chr1 LOC107984954.csv\n",
      "chr1 LOC107984956.csv\n",
      "chr1 LOC107984962.csv\n",
      "chr1 LOC107984963.csv\n",
      "chr1 LOC107984964.csv\n",
      "chr1 LOC107985041.csv\n",
      "chr1 LOC107985043.csv\n",
      "chr1 LOC107985076.csv\n",
      "chr1 LOC107985095.csv\n",
      "chr1 LOC107985096.csv\n",
      "chr1 LOC107985115.csv\n",
      "chr1 LOC107985116.csv\n",
      "chr1 LOC107985174.csv\n",
      "chr1 LOC107985180.csv\n",
      "chr1 LOC107985191.csv\n",
      "chr1 LOC107985193.csv\n",
      "chr1 LOC107985203.csv\n",
      "chr1 LOC107985207.csv\n",
      "chr1 LOC107985211.csv\n",
      "chr1 LOC107985216.csv\n",
      "chr1 LOC107985219.csv\n",
      "chr1 LOC107985220.csv\n",
      "chr1 LOC107985227.csv\n",
      "chr1 LOC107985230.csv\n",
      "chr1 LOC107985232.csv\n",
      "chr1 LOC107985233.csv\n",
      "chr1 LOC107985238.csv\n",
      "chr1 LOC107985239.csv\n",
      "chr1 LOC107985241.csv\n",
      "chr1 LOC107985250.csv\n",
      "chr1 LOC107985253.csv\n",
      "chr1 LOC107985262.csv\n",
      "chr1 LOC107985272.csv\n",
      "chr1 LOC107985281.csv\n",
      "chr1 LOC107985300.csv\n",
      "chr1 LOC107985318.csv\n",
      "chr1 LOC107985352.csv\n",
      "chr1 LOC107985354.csv\n",
      "chr1 LOC107985355.csv\n",
      "chr1 LOC107985360.csv\n",
      "chr1 LOC107985363.csv\n",
      "chr1 LOC107985364.csv\n",
      "chr1 LOC107985368.csv\n",
      "chr1 LOC107985371.csv\n",
      "chr1 LOC107985373.csv\n",
      "chr1 LOC107985375.csv\n",
      "chr1 LOC107985376.csv\n",
      "chr1 LOC107985377.csv\n",
      "chr1 LOC107985379.csv\n",
      "chr1 LOC107985391.csv\n",
      "chr1 LOC107985396.csv\n",
      "chr1 LOC107985408.csv\n",
      "chr1 LOC107985417.csv\n",
      "chr1 LOC107985420.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   5%|▌         | 1029/19235 [00:03<00:53, 340.58it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LOC107985454.csv\n",
      "chr1 LOC107985458.csv\n",
      "chr1 LOC107985460.csv\n",
      "chr1 LOC107985462.csv\n",
      "chr1 LOC107985467.csv\n",
      "chr1 LOC107985470.csv\n",
      "chr1 LOC107985517.csv\n",
      "chr1 LOC107985523.csv\n",
      "chr1 LOC107985524.csv\n",
      "chr1 LOC107985591.csv\n",
      "chr1 LOC107985593.csv\n",
      "chr1 LOC107985594.csv\n",
      "chr1 LOC107985721.csv\n",
      "chr1 LOC107985725.csv\n",
      "chr1 LOC107985729.csv\n",
      "chr1 LOC107985732.csv\n",
      "chr1 LOC107985743.csv\n",
      "chr1 LOC107985744.csv\n",
      "chr1 LOC107985745.csv\n",
      "chr1 LOC107985747.csv\n",
      "chr1 LOC112267871.csv\n",
      "chr1 LOC112267900.csv\n",
      "chr1 LOC112268218.csv\n",
      "chr1 LOC112268219.csv\n",
      "chr1 LOC112268220.csv\n",
      "chr1 LOC112268225.csv\n",
      "chr1 LOC112268227.csv\n",
      "chr1 LOC112268231.csv\n",
      "chr1 LOC112268235.csv\n",
      "chr1 LOC112268239.csv\n",
      "chr1 LOC112268240.csv\n",
      "chr1 LOC112268254.csv\n",
      "chr1 LOC112268256.csv\n",
      "chr1 LOC112268258.csv\n",
      "chr1 LOC112268263.csv\n",
      "chr1 LOC112268267.csv\n",
      "chr1 LOC112268273.csv\n",
      "chr1 LOC112268276.csv\n",
      "chr1 LOC112268287.csv\n",
      "chr1 LOC112543491.csv\n",
      "chr1 LOC117779438.csv\n",
      "chr1 LOC122526777.csv\n",
      "chr1 LOC148696.csv\n",
      "chr1 LOC284581.csv\n",
      "chr1 LOC284600.csv\n",
      "chr1 LOC339539.csv\n",
      "chr1 LOC400748.csv\n",
      "chr1 LOC440742.csv\n",
      "chr1 LOC644634.csv\n",
      "chr1 LORICRIN.csv\n",
      "chr1 LPGAT1-AS1.csv\n",
      "chr1 LRIG2.csv\n",
      "chr1 LRP8-DT.csv\n",
      "chr1 LRRC42.csv\n",
      "chr1 LRRC52.csv\n",
      "chr1 LRRC7.csv\n",
      "chr1 LRRC71.csv\n",
      "chr1 LRRC8B.csv\n",
      "chr1 LRRC8C.csv\n",
      "chr1 LRRC8D.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   6%|▌         | 1097/19235 [00:03<01:02, 289.27it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 LURAP1.csv\n",
      "chr1 LY9.csv\n",
      "chr1 LYPLA2.csv\n",
      "chr1 LYPLAL1.csv\n",
      "chr1 MAB21L3.csv\n",
      "chr1 MACF1.csv\n",
      "chr1 MACO1.csv\n",
      "chr1 MAEL.csv\n",
      "chr1 MAGI3.csv\n",
      "chr1 MAGOH-DT.csv\n",
      "chr1 MAN1A2.csv\n",
      "chr1 MAN1C1.csv\n",
      "chr1 MANEAL.csv\n",
      "chr1 MAP10.csv\n",
      "chr1 MAP3K21.csv\n",
      "chr1 MAP7D1.csv\n",
      "chr1 MAPKAPK2.csv\n",
      "chr1 MARK1.csv\n",
      "chr1 MAST2.csv\n",
      "chr1 MATN1-AS1.csv\n",
      "chr1 MDM4.csv\n",
      "chr1 MDS2.csv\n",
      "chr1 MED18.csv\n",
      "chr1 MED8-AS1.csv\n",
      "chr1 METTL13.csv\n",
      "chr1 METTL25B.csv\n",
      "chr1 MFN2.csv\n",
      "chr1 MFSD14A.csv\n",
      "chr1 MFSD2A.csv\n",
      "chr1 MFSD4A.csv\n",
      "chr1 MGC27382.csv\n",
      "chr1 MGST3.csv\n",
      "chr1 MIA3.csv\n",
      "chr1 MIB2.csv\n",
      "chr1 MICOS10-NBL1.csv\n",
      "chr1 MICOS10.csv\n",
      "chr1 MIER1.csv\n",
      "chr1 MIGA1.csv\n",
      "chr1 MIIP.csv\n",
      "chr1 MIR11399.csv\n",
      "chr1 MIR12116.csv\n",
      "chr1 MIR12132.csv\n",
      "chr1 MIR12133.csv\n",
      "chr1 MIR1231.csv\n",
      "chr1 MIR1255B2.csv\n",
      "chr1 MIR1278.csv\n",
      "chr1 MIR1295B.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   6%|▌         | 1180/19235 [00:04<00:51, 348.27it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 MIR1302-2.csv\n",
      "chr1 MIR1302-2HG.csv\n",
      "chr1 MIR197.csv\n",
      "chr1 MIR1976.csv\n",
      "chr1 MIR200A.csv\n",
      "chr1 MIR200B.csv\n",
      "chr1 MIR205.csv\n",
      "chr1 MIR205HG.csv\n",
      "chr1 MIR30C1.csv\n",
      "chr1 MIR30E.csv\n",
      "chr1 MIR3115.csv\n",
      "chr1 MIR3116-1.csv\n",
      "chr1 MIR3117.csv\n",
      "chr1 MIR3119-2.csv\n",
      "chr1 MIR3120.csv\n",
      "chr1 MIR3122.csv\n",
      "chr1 MIR3123.csv\n",
      "chr1 MIR3124.csv\n",
      "chr1 MIR320B1.csv\n",
      "chr1 MIR3620.csv\n",
      "chr1 MIR3658.csv\n",
      "chr1 MIR3659.csv\n",
      "chr1 MIR3659HG.csv\n",
      "chr1 MIR378F.csv\n",
      "chr1 MIR3972.csv\n",
      "chr1 MIR4251.csv\n",
      "chr1 MIR4255.csv\n",
      "chr1 MIR4257.csv\n",
      "chr1 MIR4258.csv\n",
      "chr1 MIR429.csv\n",
      "chr1 MIR4418.csv\n",
      "chr1 MIR4421.csv\n",
      "chr1 MIR4422.csv\n",
      "chr1 MIR4422HG.csv\n",
      "chr1 MIR4423.csv\n",
      "chr1 MIR4424.csv\n",
      "chr1 MIR4425.csv\n",
      "chr1 MIR4426.csv\n",
      "chr1 MIR4427.csv\n",
      "chr1 MIR4428.csv\n",
      "chr1 MIR4632.csv\n",
      "chr1 MIR4654.csv\n",
      "chr1 MIR4666A.csv\n",
      "chr1 MIR4671.csv\n",
      "chr1 MIR4677.csv\n",
      "chr1 MIR4684.csv\n",
      "chr1 MIR4781.csv\n",
      "chr1 MIR4794.csv\n",
      "chr1 MIR5187.csv\n",
      "chr1 MIR5191.csv\n",
      "chr1 MIR553.csv\n",
      "chr1 MIR554.csv\n",
      "chr1 MIR556.csv\n",
      "chr1 MIR557.csv\n",
      "chr1 MIR5584.csv\n",
      "chr1 MIR5585.csv\n",
      "chr1 MIR5697.csv\n",
      "chr1 MIR6077.csv\n",
      "chr1 MIR6079.csv\n",
      "chr1 MIR6084.csv\n",
      "chr1 MIR6500.csv\n",
      "chr1 MIR6729.csv\n",
      "chr1 MIR6732.csv\n",
      "chr1 MIR6735.csv\n",
      "chr1 MIR6739.csv\n",
      "chr1 MIR6740.csv\n",
      "chr1 MIR6769B.csv\n",
      "chr1 MIR6878.csv\n",
      "chr1 MIR7156.csv\n",
      "chr1 MIR760.csv\n",
      "chr1 MIR7846.csv\n",
      "chr1 MIR7852.csv\n",
      "chr1 MIR92B.csv\n",
      "chr1 MIR942.csv\n",
      "chr1 MIXL1.csv\n",
      "chr1 MKNK1-AS1.csv\n",
      "chr1 MLLT11.csv\n",
      "chr1 MMACHC.csv\n",
      "chr1 MMEL1-AS1.csv\n",
      "chr1 MMP23B.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   6%|▋         | 1216/19235 [00:04<00:53, 338.29it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 MNDA.csv\n",
      "chr1 MOV10.csv\n",
      "chr1 MPL.csv\n",
      "chr1 MPZL1.csv\n",
      "chr1 MR1.csv\n",
      "chr1 MROH7-TTC4.csv\n",
      "chr1 MROH7.csv\n",
      "chr1 MROH9.csv\n",
      "chr1 MRPL20-AS1.csv\n",
      "chr1 MRPL20-DT.csv\n",
      "chr1 MRPL37.csv\n",
      "chr1 MRPS21.csv\n",
      "chr1 MRTO4.csv\n",
      "chr1 MSH4.csv\n",
      "chr1 MSTO1.csv\n",
      "chr1 MTARC1.csv\n",
      "chr1 MTARC2.csv\n",
      "chr1 MTF2.csv\n",
      "chr1 MTFR1L.csv\n",
      "chr1 MTOR-AS1.csv\n",
      "chr1 MTR.csv\n",
      "chr1 MTX1.csv\n",
      "chr1 MYCL-AS1.csv\n",
      "chr1 MYOCOS.csv\n",
      "chr1 MYOM3-AS1.csv\n",
      "chr1 MYOPARR.csv\n",
      "chr1 NASP.csv\n",
      "chr1 NAV1.csv\n",
      "chr1 NAXE.csv\n",
      "chr1 NBL1.csv\n",
      "chr1 NBPF12.csv\n",
      "chr1 NBPF19.csv\n",
      "chr1 NBPF26.csv\n",
      "chr1 NBPF3.csv\n",
      "chr1 NBPF6.csv\n",
      "chr1 NBPF8.csv\n",
      "chr1 NCDN.csv\n",
      "chr1 NCMAP.csv\n",
      "chr1 NCSTN.csv\n",
      "chr1 NDUFS2.csv\n",
      "chr1 NDUFS5.csv\n",
      "chr1 NECAP2.csv\n",
      "chr1 NECTIN4-AS1.csv\n",
      "chr1 NEK2-DT.csv\n",
      "chr1 NEK7.csv\n",
      "chr1 NENF.csv\n",
      "chr1 NEXN.csv\n",
      "chr1 NFASC.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   7%|▋         | 1282/19235 [00:04<01:01, 291.40it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 NFIA.csv\n",
      "chr1 NFYC.csv\n",
      "chr1 NGF-AS1.csv\n",
      "chr1 NHLH1.csv\n",
      "chr1 NIPAL3.csv\n",
      "chr1 NIT1.csv\n",
      "chr1 NLRP3.csv\n",
      "chr1 NMNAT1.csv\n",
      "chr1 NOS1AP.csv\n",
      "chr1 NOTCH2NLC.csv\n",
      "chr1 NOTCH2NLR.csv\n",
      "chr1 NPL.csv\n",
      "chr1 NPPA-AS1.csv\n",
      "chr1 NPR1.csv\n",
      "chr1 NR5A2.csv\n",
      "chr1 NSUN4.csv\n",
      "chr1 NTMT2.csv\n",
      "chr1 NTNG1.csv\n",
      "chr1 NTPCR.csv\n",
      "chr1 NTRK1.csv\n",
      "chr1 NUDC.csv\n",
      "chr1 NUDT17.csv\n",
      "chr1 NUDT4B.csv\n",
      "chr1 NUF2.csv\n",
      "chr1 NUP133-DT.csv\n",
      "chr1 OAZ3.csv\n",
      "chr1 OBSCN.csv\n",
      "chr1 ODR4.csv\n",
      "chr1 OLFML3.csv\n",
      "chr1 OPRD1.csv\n",
      "chr1 OPTC.csv\n",
      "chr1 OR10J1.csv\n",
      "chr1 OR10K1.csv\n",
      "chr1 OR10R2.csv\n",
      "chr1 OR10Z1.csv\n",
      "chr1 OR14C36.csv\n",
      "chr1 OR14K1.csv\n",
      "chr1 OR14L1P.csv\n",
      "chr1 OR2AJ1.csv\n",
      "chr1 OR2AK2.csv\n",
      "chr1 OR2G2.csv\n",
      "chr1 OR2G3.csv\n",
      "chr1 OR2G6.csv\n",
      "chr1 OR2L13.csv\n",
      "chr1 OR2L2.csv\n",
      "chr1 OR2L3.csv\n",
      "chr1 OR2L5.csv\n",
      "chr1 OR2L8.csv\n",
      "chr1 OR2M2.csv\n",
      "chr1 OR2M3.csv\n",
      "chr1 OR2M4.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   7%|▋         | 1313/19235 [00:04<01:00, 296.25it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 OR2M5.csv\n",
      "chr1 OR2T1.csv\n",
      "chr1 OR2T2.csv\n",
      "chr1 OR2T3.csv\n",
      "chr1 OR2T4.csv\n",
      "chr1 OR2T5.csv\n",
      "chr1 OR2T6.csv\n",
      "chr1 OR2T7.csv\n",
      "chr1 OR2T8.csv\n",
      "chr1 OR2W3.csv\n",
      "chr1 OR4F5.csv\n",
      "chr1 OR6K6.csv\n",
      "chr1 OR9H1P.csv\n",
      "chr1 OSBPL9.csv\n",
      "chr1 OTUD3.csv\n",
      "chr1 OVAAL.csv\n",
      "chr1 PABPC4-AS1.csv\n",
      "chr1 PACERR.csv\n",
      "chr1 PADI1.csv\n",
      "chr1 PADI3.csv\n",
      "chr1 PADI4.csv\n",
      "chr1 PADI6.csv\n",
      "chr1 PALMD.csv\n",
      "chr1 PAPPA2.csv\n",
      "chr1 PARK7.csv\n",
      "chr1 PATJ.csv\n",
      "chr1 PAX7.csv\n",
      "chr1 PBX1.csv\n",
      "chr1 PCAT6.csv\n",
      "chr1 PCP4L1.csv\n",
      "chr1 PCSK9.csv\n",
      "chr1 PDC-AS1.csv\n",
      "chr1 PDE4B.csv\n",
      "chr1 PDE4DIP.csv\n",
      "chr1 PDIK1L.csv\n",
      "chr1 PDPN.csv\n",
      "chr1 PEA15.csv\n",
      "chr1 PEAR1.csv\n",
      "chr1 PEF1-AS1.csv\n",
      "chr1 PER3.csv\n",
      "chr1 PEX14.csv\n",
      "chr1 PFKFB2.csv\n",
      "chr1 PGBD2.csv\n",
      "chr1 PGD.csv\n",
      "chr1 PGM1.csv\n",
      "chr1 PHACTR4.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   7%|▋         | 1372/19235 [00:04<01:05, 270.85it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 PHC2-AS1.csv\n",
      "chr1 PHF13.csv\n",
      "chr1 PHGDH.csv\n",
      "chr1 PIFO.csv\n",
      "chr1 PIGV.csv\n",
      "chr1 PIK3CD.csv\n",
      "chr1 PINK1.csv\n",
      "chr1 PIP5K1A.csv\n",
      "chr1 PITHD1.csv\n",
      "chr1 PKN2.csv\n",
      "chr1 PKP1.csv\n",
      "chr1 PLA2G2F.csv\n",
      "chr1 PLA2G4A.csv\n",
      "chr1 PLA2G5.csv\n",
      "chr1 PLCH2.csv\n",
      "chr1 PLEKHM2.csv\n",
      "chr1 PLEKHN1.csv\n",
      "chr1 PLEKHO1.csv\n",
      "chr1 PLK3.csv\n",
      "chr1 PLOD1.csv\n",
      "chr1 PLPPR4.csv\n",
      "chr1 PLPPR5-AS1.csv\n",
      "chr1 PMF1-BGLAP.csv\n",
      "chr1 PMF1.csv\n",
      "chr1 PNRC2.csv\n",
      "chr1 PODN.csv\n",
      "chr1 POGK.csv\n",
      "chr1 POLR3C.csv\n",
      "chr1 POLR3GL.csv\n",
      "chr1 POU2F1.csv\n",
      "chr1 PPCS.csv\n",
      "chr1 PPFIA4.csv\n",
      "chr1 PPIAL4A.csv\n",
      "chr1 PPIAL4C.csv\n",
      "chr1 PPIAL4F.csv\n",
      "chr1 PPIE.csv\n",
      "chr1 PPIH.csv\n",
      "chr1 PPM1J-DT.csv\n",
      "chr1 PPOX.csv\n",
      "chr1 PPP1R12B.csv\n",
      "chr1 PPP1R8.csv\n",
      "chr1 PPP2R5A.csv\n",
      "chr1 PRAMEF1.csv\n",
      "chr1 PRAMEF12.csv\n",
      "chr1 PRAMEF15.csv\n",
      "chr1 PRAMEF17.csv\n",
      "chr1 PRAMEF2.csv\n",
      "chr1 PRAMEF20.csv\n",
      "chr1 PRAMEF25.csv\n",
      "chr1 PRAMEF33.csv\n",
      "chr1 PRAMEF5.csv\n",
      "chr1 PRAMEF7.csv\n",
      "chr1 PRAMEF9.csv\n",
      "chr1 PRCC.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   7%|▋         | 1428/19235 [00:05<01:10, 252.85it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 PRDM16.csv\n",
      "chr1 PRDM2.csv\n",
      "chr1 PRDX6.csv\n",
      "chr1 PRELP.csv\n",
      "chr1 PRG4.csv\n",
      "chr1 PRKAA2.csv\n",
      "chr1 PRKACB.csv\n",
      "chr1 PRKCZ.csv\n",
      "chr1 PRMT6.csv\n",
      "chr1 PROK1.csv\n",
      "chr1 PROX1.csv\n",
      "chr1 PRPF3.csv\n",
      "chr1 PRPF38A.csv\n",
      "chr1 PRPF38B.csv\n",
      "chr1 PRR9.csv\n",
      "chr1 PRRC2C.csv\n",
      "chr1 PRRX1.csv\n",
      "chr1 PRSS38.csv\n",
      "chr1 PRUNE1.csv\n",
      "chr1 PRXL2B.csv\n",
      "chr1 PSEN2.csv\n",
      "chr1 PSMB4.csv\n",
      "chr1 PSMD4.csv\n",
      "chr1 PTBP2.csv\n",
      "chr1 PTGFR.csv\n",
      "chr1 PTGFRN.csv\n",
      "chr1 PTPRC.csv\n",
      "chr1 PTPRF.csv\n",
      "chr1 PTPRU.csv\n",
      "chr1 PUSL1.csv\n",
      "chr1 PYHIN1.csv\n",
      "chr1 QSOX1.csv\n",
      "chr1 RAB25.csv\n",
      "chr1 RAB42.csv\n",
      "chr1 RAB4A.csv\n",
      "chr1 RABGAP1L.csv\n",
      "chr1 RABGGTB.csv\n",
      "chr1 RAD54L.csv\n",
      "chr1 RALGPS2.csv\n",
      "chr1 RAP1A.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   8%|▊         | 1454/19235 [00:05<01:10, 253.40it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 RASAL2.csv\n",
      "chr1 RASSF5.csv\n",
      "chr1 RAVER2.csv\n",
      "chr1 RBBP4.csv\n",
      "chr1 RBM15.csv\n",
      "chr1 RBP7.csv\n",
      "chr1 RC3H1-DT.csv\n",
      "chr1 RCAN3.csv\n",
      "chr1 RCC1.csv\n",
      "chr1 RCC2-AS1.csv\n",
      "chr1 RCOR3.csv\n",
      "chr1 RCSD1.csv\n",
      "chr1 RER1.csv\n",
      "chr1 RERE-AS1.csv\n",
      "chr1 RFX5-AS1.csv\n",
      "chr1 RGL1.csv\n",
      "chr1 RGS1.csv\n",
      "chr1 RGS13.csv\n",
      "chr1 RGS18.csv\n",
      "chr1 RGS2.csv\n",
      "chr1 RGS21.csv\n",
      "chr1 RGS4.csv\n",
      "chr1 RGS5-AS1.csv\n",
      "chr1 RGSL1.csv\n",
      "chr1 RHBG.csv\n",
      "chr1 RHD.csv\n",
      "chr1 RHEX.csv\n",
      "chr1 RHOU.csv\n",
      "chr1 RIIAD1.csv\n",
      "chr1 RIMKLA.csv\n",
      "chr1 RLF.csv\n",
      "chr1 RNF11.csv\n",
      "chr1 RNF187.csv\n",
      "chr1 RNF2.csv\n",
      "chr1 RNF207.csv\n",
      "chr1 RNF220.csv\n",
      "chr1 RNPC3.csv\n",
      "chr1 RNPEP.csv\n",
      "chr1 RNU1-2.csv\n",
      "chr1 RNU1-4.csv\n",
      "chr1 RNU11.csv\n",
      "chr1 RNU5E-1.csv\n",
      "chr1 RNVU1-14.csv\n",
      "chr1 RNVU1-15.csv\n",
      "chr1 RNVU1-3.csv\n",
      "chr1 RNVU1-4.csv\n",
      "chr1 RO60.csv\n",
      "chr1 ROR1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   8%|▊         | 1506/19235 [00:05<01:09, 255.11it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 RPAP2.csv\n",
      "chr1 RPF1.csv\n",
      "chr1 RPL11.csv\n",
      "chr1 RPL5.csv\n",
      "chr1 RPRD2.csv\n",
      "chr1 RPS27.csv\n",
      "chr1 RPS6KA1.csv\n",
      "chr1 RPS6KC1.csv\n",
      "chr1 RPS8.csv\n",
      "chr1 RRP15.csv\n",
      "chr1 RSC1A1.csv\n",
      "chr1 RTCA.csv\n",
      "chr1 RUNX3-AS1.csv\n",
      "chr1 RUSC1.csv\n",
      "chr1 RWDD3.csv\n",
      "chr1 RXFP4.csv\n",
      "chr1 RYR2.csv\n",
      "chr1 S100A1.csv\n",
      "chr1 S100A7A.csv\n",
      "chr1 S100A9.csv\n",
      "chr1 S100PBP.csv\n",
      "chr1 S1PR1.csv\n",
      "chr1 SAMD11.csv\n",
      "chr1 SAMD13.csv\n",
      "chr1 SARS1.csv\n",
      "chr1 SCARNA1.csv\n",
      "chr1 SCARNA2.csv\n",
      "chr1 SCARNA21B.csv\n",
      "chr1 SCCPDH.csv\n",
      "chr1 SCNM1.csv\n",
      "chr1 SCNN1D.csv\n",
      "chr1 SCP2.csv\n",
      "chr1 SDCCAG8.csv\n",
      "chr1 SDHC.csv\n",
      "chr1 SELENON.csv\n",
      "chr1 SEMA4A.csv\n",
      "chr1 SERINC2.csv\n",
      "chr1 SERTAD4.csv\n",
      "chr1 SESN2.csv\n",
      "chr1 SETDB1.csv\n",
      "chr1 SFN.csv\n",
      "chr1 SFT2D2.csv\n",
      "chr1 SGIP1.csv\n",
      "chr1 SH3BGRL3.csv\n",
      "chr1 SH3D21.csv\n",
      "chr1 SH3GLB1.csv\n",
      "chr1 SHISA4.csv\n",
      "chr1 SHISAL2A.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   8%|▊         | 1561/19235 [00:05<01:14, 236.32it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 SKI.csv\n",
      "chr1 SLAMF7.csv\n",
      "chr1 SLAMF8.csv\n",
      "chr1 SLC16A1-AS1.csv\n",
      "chr1 SLC22A15.csv\n",
      "chr1 SLC25A33.csv\n",
      "chr1 SLC25A34.csv\n",
      "chr1 SLC25A44.csv\n",
      "chr1 SLC26A9-AS1.csv\n",
      "chr1 SLC27A3.csv\n",
      "chr1 SLC2A1-DT.csv\n",
      "chr1 SLC30A7.csv\n",
      "chr1 SLC35A3.csv\n",
      "chr1 SLC35F3.csv\n",
      "chr1 SLC44A3.csv\n",
      "chr1 SLC45A1.csv\n",
      "chr1 SLC50A1.csv\n",
      "chr1 SLC5A9.csv\n",
      "chr1 SLC66A1.csv\n",
      "chr1 SLC6A17.csv\n",
      "chr1 SLFNL1-AS1.csv\n",
      "chr1 SMAP2.csv\n",
      "chr1 SMCP.csv\n",
      "chr1 SMG7.csv\n",
      "chr1 SMIM1.csv\n",
      "chr1 SMPDL3B.csv\n",
      "chr1 SMYD2.csv\n",
      "chr1 SMYD3-AS1.csv\n",
      "chr1 SNAP47.csv\n",
      "chr1 SNAPIN.csv\n",
      "chr1 SNHG3.csv\n",
      "chr1 SNORA16B.csv\n",
      "chr1 SNORA58B.csv\n",
      "chr1 SNORA59A.csv\n",
      "chr1 SNORA66.csv\n",
      "chr1 SNORA70H.csv\n",
      "chr1 SNORA73A.csv\n",
      "chr1 SNORA73B.csv\n",
      "chr1 SNORA77.csv\n",
      "chr1 SNORD13C.csv\n",
      "chr1 SNORD160.csv\n",
      "chr1 SNORD21.csv\n",
      "chr1 SNORD38A.csv\n",
      "chr1 SNORD38B.csv\n",
      "chr1 SNORD3G.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   8%|▊         | 1592/19235 [00:05<01:09, 254.15it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 SNORD45A.csv\n",
      "chr1 SNORD45B.csv\n",
      "chr1 SNORD45C.csv\n",
      "chr1 SNORD46.csv\n",
      "chr1 SNORD55.csv\n",
      "chr1 SNRPE.csv\n",
      "chr1 SNX27.csv\n",
      "chr1 SNX7.csv\n",
      "chr1 SOAT1.csv\n",
      "chr1 SOX13.csv\n",
      "chr1 SPATA1.csv\n",
      "chr1 SPATA17.csv\n",
      "chr1 SPATA42.csv\n",
      "chr1 SPEN.csv\n",
      "chr1 SPRR1A.csv\n",
      "chr1 SPRR1B.csv\n",
      "chr1 SPRR3.csv\n",
      "chr1 SPRR4.csv\n",
      "chr1 SPRR5.csv\n",
      "chr1 SPRTN.csv\n",
      "chr1 SPSB1.csv\n",
      "chr1 SRARP.csv\n",
      "chr1 SRGAP2.csv\n",
      "chr1 SRGAP2C.csv\n",
      "chr1 SRP9.csv\n",
      "chr1 SRRM1.csv\n",
      "chr1 SRSF11.csv\n",
      "chr1 SSBP3-AS1.csv\n",
      "chr1 ST3GAL3.csv\n",
      "chr1 ST6GALNAC3.csv\n",
      "chr1 ST6GALNAC5.csv\n",
      "chr1 STRIP1.csv\n",
      "chr1 STUM.csv\n",
      "chr1 STX12.csv\n",
      "chr1 STXBP3.csv\n",
      "chr1 STYXL2.csv\n",
      "chr1 SUCO.csv\n",
      "chr1 SWT1.csv\n",
      "chr1 SYCP1.csv\n",
      "chr1 SYPL2.csv\n",
      "chr1 SYT11.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   9%|▊         | 1646/19235 [00:05<01:11, 247.03it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 SYT14.csv\n",
      "chr1 SYTL1.csv\n",
      "chr1 SZRD1.csv\n",
      "chr1 SZT2.csv\n",
      "chr1 TAF12-DT.csv\n",
      "chr1 TAF1A-AS1.csv\n",
      "chr1 TAFA3.csv\n",
      "chr1 TARDBP.csv\n",
      "chr1 TARS2.csv\n",
      "chr1 TAS1R1.csv\n",
      "chr1 TAS1R3.csv\n",
      "chr1 TATDN3.csv\n",
      "chr1 TBCE.csv\n",
      "chr1 TBX19.csv\n",
      "chr1 TCEANC2.csv\n",
      "chr1 TDRD10.csv\n",
      "chr1 TDRD5.csv\n",
      "chr1 TDRKH-AS1.csv\n",
      "chr1 TEKT2.csv\n",
      "chr1 TENT5C.csv\n",
      "chr1 TEX35.csv\n",
      "chr1 TEX38.csv\n",
      "chr1 TEX50.csv\n",
      "chr1 TFAP2E.csv\n",
      "chr1 TGFB2-OT1.csv\n",
      "chr1 TGFB2.csv\n",
      "chr1 THAP3.csv\n",
      "chr1 THBS3-AS1.csv\n",
      "chr1 THEMIS2.csv\n",
      "chr1 THRAP3.csv\n",
      "chr1 TIE1.csv\n",
      "chr1 TIMM17A.csv\n",
      "chr1 TINAGL1.csv\n",
      "chr1 TIPRL.csv\n",
      "chr1 TLCD4-RWDD3.csv\n",
      "chr1 TLCD4.csv\n",
      "chr1 TMCC2.csv\n",
      "chr1 TMCO1-AS1.csv\n",
      "chr1 TMCO2.csv\n",
      "chr1 TMEM125.csv\n",
      "chr1 TMEM167B.csv\n",
      "chr1 TMEM183A.csv\n",
      "chr1 TMEM201.csv\n",
      "chr1 TMEM222.csv\n",
      "chr1 TMEM269.csv\n",
      "chr1 TMEM39B.csv\n",
      "chr1 TMEM50A.csv\n",
      "chr1 TMEM51.csv\n",
      "chr1 TMEM61.csv\n",
      "chr1 TMEM69.csv\n",
      "chr1 TMEM78.csv\n",
      "chr1 TMEM79.csv\n",
      "chr1 TMEM82.csv\n",
      "chr1 TMEM88B.csv\n",
      "chr1 TNFAIP8L2-SCNM1.csv\n",
      "chr1 TNFAIP8L2.csv\n",
      "chr1 TNFRSF14.csv\n",
      "chr1 TNFRSF1B.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   9%|▉         | 1720/19235 [00:06<00:57, 306.57it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 TNFRSF8.csv\n",
      "chr1 TNN.csv\n",
      "chr1 TNNI3K.csv\n",
      "chr1 TOE1.csv\n",
      "chr1 TOMM40L.csv\n",
      "chr1 TOR1AIP1.csv\n",
      "chr1 TOR3A.csv\n",
      "chr1 TP73.csv\n",
      "chr1 TPRG1L.csv\n",
      "chr1 TRAF3IP3.csv\n",
      "chr1 TRAF5.csv\n",
      "chr1 TRE-CTC1-1.csv\n",
      "chr1 TRE-CTC2-1.csv\n",
      "chr1 TRE-TTC3-1.csv\n",
      "chr1 TRG-CCC1-2.csv\n",
      "chr1 TRG-CCC5-1.csv\n",
      "chr1 TRG-GCC1-1.csv\n",
      "chr1 TRG-GCC1-2.csv\n",
      "chr1 TRG-GCC1-3.csv\n",
      "chr1 TRG-GCC1-4.csv\n",
      "chr1 TRG-GCC4-1.csv\n",
      "chr1 TRG-TCC2-1.csv\n",
      "chr1 TRG-TCC2-6.csv\n",
      "chr1 TRH-GTG1-1.csv\n",
      "chr1 TRH-GTG1-2.csv\n",
      "chr1 TRH-GTG1-3.csv\n",
      "chr1 TRIM46.csv\n",
      "chr1 TRIM58.csv\n",
      "chr1 TRIM67.csv\n",
      "chr1 TRK-CTT2-1.csv\n",
      "chr1 TRK-TTT3-1.csv\n",
      "chr1 TRK-TTT8-1.csv\n",
      "chr1 TRL-CAA4-1.csv\n",
      "chr1 TRL-CAG1-1.csv\n",
      "chr1 TRL-CAG1-2.csv\n",
      "chr1 TRL-CAG1-3.csv\n",
      "chr1 TRL-CAG1-4.csv\n",
      "chr1 TRL-CAG1-5.csv\n",
      "chr1 TRMT13.csv\n",
      "chr1 TRN-GTT1-1.csv\n",
      "chr1 TRN-GTT10-1.csv\n",
      "chr1 TRN-GTT13-1.csv\n",
      "chr1 TRN-GTT2-1.csv\n",
      "chr1 TRN-GTT2-7.csv\n",
      "chr1 TRN-GTT24-1.csv\n",
      "chr1 TRN-GTT3-1.csv\n",
      "chr1 TRN-GTT4-1.csv\n",
      "chr1 TRNAU1AP.csv\n",
      "chr1 TRNAV-CAC.csv\n",
      "chr1 TRNP1.csv\n",
      "chr1 TRP-CGG1-1.csv\n",
      "chr1 TRQ-CTG3-1.csv\n",
      "chr1 TRQ-CTG3-2.csv\n",
      "chr1 TRQ-CTG4-2.csv\n",
      "chr1 TRQ-CTG7-1.csv\n",
      "chr1 TRR-TCT1-1.csv\n",
      "chr1 TRT-TGT2-1.csv\n",
      "chr1 TRX-CAT1-1.csv\n",
      "chr1 TSACC.csv\n",
      "chr1 TSEN15.csv\n",
      "chr1 TSHB.csv\n",
      "chr1 TSNAX-DISC1.csv\n",
      "chr1 TSNAX.csv\n",
      "chr1 TSPAN1.csv\n",
      "chr1 TSSK3.csv\n",
      "chr1 TTC24.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   9%|▉         | 1784/19235 [00:06<00:59, 294.66it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 TTC39A-AS1.csv\n",
      "chr1 TTC4.csv\n",
      "chr1 TTF2.csv\n",
      "chr1 TTLL10.csv\n",
      "chr1 TUFT1.csv\n",
      "chr1 TXLNA.csv\n",
      "chr1 TXNDC12-AS1.csv\n",
      "chr1 TYW3.csv\n",
      "chr1 UAP1.csv\n",
      "chr1 UBAP2L.csv\n",
      "chr1 UBE2Q1-AS1.csv\n",
      "chr1 UBE2U.csv\n",
      "chr1 UBE4B.csv\n",
      "chr1 UBIAD1.csv\n",
      "chr1 UBL4B.csv\n",
      "chr1 UBXN10.csv\n",
      "chr1 UCK2.csv\n",
      "chr1 UFC1.csv\n",
      "chr1 UHMK1.csv\n",
      "chr1 UQCRH.csv\n",
      "chr1 URB2.csv\n",
      "chr1 UROD.csv\n",
      "chr1 USH2A-AS1.csv\n",
      "chr1 USH2A-AS2.csv\n",
      "chr1 USP1.csv\n",
      "chr1 USP21.csv\n",
      "chr1 UTP11.csv\n",
      "chr1 UTP25.csv\n",
      "chr1 VAMP3.csv\n",
      "chr1 VANGL1.csv\n",
      "chr1 VANGL2.csv\n",
      "chr1 VASH2.csv\n",
      "chr1 VAV3-AS1.csv\n",
      "chr1 VCAM1.csv\n",
      "chr1 VPS13D.csv\n",
      "chr1 VPS45.csv\n",
      "chr1 VWA1.csv\n",
      "chr1 VWA5B1.csv\n",
      "chr1 WARS2-AS1.csv\n",
      "chr1 WARS2-IT1.csv\n",
      "chr1 WDR3.csv\n",
      "chr1 WDR64.csv\n",
      "chr1 WDTC1.csv\n",
      "chr1 WNT2B.csv\n",
      "chr1 WNT3A.csv\n",
      "chr1 XCL1.csv\n",
      "chr1 XKR8.csv\n",
      "chr1 XPR1.csv\n",
      "chr1 YBX1.csv\n",
      "chr1 YTHDF2.csv\n",
      "chr1 ZBED6.csv\n",
      "chr1 ZBTB18.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:   9%|▉         | 1814/19235 [00:06<01:00, 288.84it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 ZBTB37.csv\n",
      "chr1 ZBTB40.csv\n",
      "chr1 ZBTB48.csv\n",
      "chr1 ZBTB7B.csv\n",
      "chr1 ZBTB8A.csv\n",
      "chr1 ZBTB8B.csv\n",
      "chr1 ZC3H11A.csv\n",
      "chr1 ZC3H12A.csv\n",
      "chr1 ZCCHC17.csv\n",
      "chr1 ZDHHC18.csv\n",
      "chr1 ZFP69.csv\n",
      "chr1 ZFP69B.csv\n",
      "chr1 ZFYVE9.csv\n",
      "chr1 ZMPSTE24.csv\n",
      "chr1 ZMYM1.csv\n",
      "chr1 ZMYM4.csv\n",
      "chr1 ZNF326.csv\n",
      "chr1 ZNF362.csv\n",
      "chr1 ZNF436-AS1.csv\n",
      "chr1 ZNF593.csv\n",
      "chr1 ZNF672.csv\n",
      "chr1 ZNF678.csv\n",
      "chr1 ZNF684.csv\n",
      "chr1 ZNF687.csv\n",
      "chr1 ZNF691.csv\n",
      "chr1 ZPLD2P.csv\n",
      "chr1 ZRANB2-AS1.csv\n",
      "chr1 ZRANB2-DT.csv\n",
      "chr1 ZSCAN20.csv\n",
      "chr1 ZYG11A.csv\n",
      "chr1 ZYG11B.csv\n",
      "chr2 ABCG8.csv\n",
      "chr2 ABHD1.csv\n",
      "chr2 ABI2.csv\n",
      "chr2 ACKR3.csv\n",
      "chr2 ACMSD.csv\n",
      "chr2 ACOXL.csv\n",
      "chr2 ACP1.csv\n",
      "chr2 ACSL3.csv\n",
      "chr2 ACTG2.csv\n",
      "chr2 ACTR2.csv\n",
      "chr2 ACTR3.csv\n",
      "chr2 ACVR2A.csv\n",
      "chr2 ACYP2.csv\n",
      "chr2 ADAM23.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  10%|▉         | 1844/19235 [00:06<01:09, 251.66it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 AFTPH.csv\n",
      "chr2 AGAP1-IT1.csv\n",
      "chr2 AGAP1.csv\n",
      "chr2 AGBL5.csv\n",
      "chr2 AGFG1.csv\n",
      "chr2 AGPS.csv\n",
      "chr2 AGXT.csv\n",
      "chr2 ALLC.csv\n",
      "chr2 ALMS1-IT1.csv\n",
      "chr2 ALMS1.csv\n",
      "chr2 ALPG.csv\n",
      "chr2 ALPI.csv\n",
      "chr2 ALPP.csv\n",
      "chr2 AMER3.csv\n",
      "chr2 ANKAR.csv\n",
      "chr2 ANKRD36.csv\n",
      "chr2 ANKRD44-AS1.csv\n",
      "chr2 ANKRD44-DT.csv\n",
      "chr2 ANKRD53.csv\n",
      "chr2 ANKZF1.csv\n",
      "chr2 ANO7.csv\n",
      "chr2 ANTXR1.csv\n",
      "chr2 ANXA4.csv\n",
      "chr2 AOX1.csv\n",
      "chr2 APLF.csv\n",
      "chr2 AQP12A.csv\n",
      "chr2 AQP7B.csv\n",
      "chr2 ARHGAP15.csv\n",
      "chr2 ARHGAP25.csv\n",
      "chr2 ARHGEF33.csv\n",
      "chr2 ARHGEF4.csv\n",
      "chr2 ARID5A.csv\n",
      "chr2 ARL6IP6.csv\n",
      "chr2 ARMC9.csv\n",
      "chr2 ARPC2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  10%|▉         | 1896/19235 [00:06<01:14, 232.09it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 ASAP2.csv\n",
      "chr2 ASB1.csv\n",
      "chr2 ASDURF.csv\n",
      "chr2 ASIC4.csv\n",
      "chr2 ASNSD1.csv\n",
      "chr2 ATG16L1.csv\n",
      "chr2 ATG4B.csv\n",
      "chr2 ATIC.csv\n",
      "chr2 ATOH8.csv\n",
      "chr2 ATP6V1B1.csv\n",
      "chr2 ATP6V1C2.csv\n",
      "chr2 ATRAID.csv\n",
      "chr2 B3GALT1.csv\n",
      "chr2 B3GNT2.csv\n",
      "chr2 B3GNT7.csv\n",
      "chr2 BABAM2.csv\n",
      "chr2 BAZ2B-AS1.csv\n",
      "chr2 BBS5.csv\n",
      "chr2 BCL2L11.csv\n",
      "chr2 BCS1L.csv\n",
      "chr2 BCYRN1.csv\n",
      "chr2 BIRC6.csv\n",
      "chr2 BMPR2.csv\n",
      "chr2 BOK.csv\n",
      "chr2 BOLA3-DT.csv\n",
      "chr2 BZW1.csv\n",
      "chr2 C2orf15.csv\n",
      "chr2 C2orf16.csv\n",
      "chr2 C2orf27A.csv\n",
      "chr2 C2orf49.csv\n",
      "chr2 C2orf50.csv\n",
      "chr2 C2orf69.csv\n",
      "chr2 C2orf72.csv\n",
      "chr2 C2orf73.csv\n",
      "chr2 C2orf74.csv\n",
      "chr2 C2orf78.csv\n",
      "chr2 C2orf88.csv\n",
      "chr2 C2orf92.csv\n",
      "chr2 CAB39.csv\n",
      "chr2 CAD.csv\n",
      "chr2 CALCRL-AS1.csv\n",
      "chr2 CAMKMT.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  10%|█         | 1946/19235 [00:07<01:12, 239.57it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 CAPN10.csv\n",
      "chr2 CARF.csv\n",
      "chr2 CASP10.csv\n",
      "chr2 CASP8.csv\n",
      "chr2 CATIP.csv\n",
      "chr2 CAVIN2-AS1.csv\n",
      "chr2 CBWD2.csv\n",
      "chr2 CCDC138.csv\n",
      "chr2 CCDC140.csv\n",
      "chr2 CCDC148-AS1.csv\n",
      "chr2 CCDC150.csv\n",
      "chr2 CCDC74A.csv\n",
      "chr2 CCDC85A.csv\n",
      "chr2 CCL20.csv\n",
      "chr2 CCNT2.csv\n",
      "chr2 CCNYL1.csv\n",
      "chr2 CCT7.csv\n",
      "chr2 CD28.csv\n",
      "chr2 CD8B2.csv\n",
      "chr2 CDCA7.csv\n",
      "chr2 CDK15.csv\n",
      "chr2 CDK5R2.csv\n",
      "chr2 CEBPZOS.csv\n",
      "chr2 CENPA.csv\n",
      "chr2 CENPO.csv\n",
      "chr2 CEP68.csv\n",
      "chr2 CERS6.csv\n",
      "chr2 CFAP221.csv\n",
      "chr2 CFAP36.csv\n",
      "chr2 CFC1B.csv\n",
      "chr2 CFLAR.csv\n",
      "chr2 CHAC2.csv\n",
      "chr2 CHCHD5.csv\n",
      "chr2 CHMP3-AS1.csv\n",
      "chr2 CHRND.csv\n",
      "chr2 CHRNG.csv\n",
      "chr2 CHROMR.csv\n",
      "chr2 CIAO1.csv\n",
      "chr2 CLASP1-AS1.csv\n",
      "chr2 CLIP4.csv\n",
      "chr2 CMKLR2-AS.csv\n",
      "chr2 CNGA3.csv\n",
      "chr2 CNNM3.csv\n",
      "chr2 CNNM4.csv\n",
      "chr2 CNOT11.csv\n",
      "chr2 CNOT9.csv\n",
      "chr2 CNTNAP5.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  10%|█         | 1994/19235 [00:07<01:16, 224.17it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 COL3A1.csv\n",
      "chr2 COL4A3.csv\n",
      "chr2 COLEC11.csv\n",
      "chr2 COMMD1.csv\n",
      "chr2 COPS7B.csv\n",
      "chr2 COPS8.csv\n",
      "chr2 COQ10B.csv\n",
      "chr2 COX5B.csv\n",
      "chr2 CPO.csv\n",
      "chr2 CPS1-IT1.csv\n",
      "chr2 CPS1.csv\n",
      "chr2 CPSF3.csv\n",
      "chr2 CREB1.csv\n",
      "chr2 CRIM1.csv\n",
      "chr2 CRIPT.csv\n",
      "chr2 CROCC2.csv\n",
      "chr2 CSRNP3.csv\n",
      "chr2 CT75.csv\n",
      "chr2 CTDSP1.csv\n",
      "chr2 CTLA4.csv\n",
      "chr2 CTNNA2.csv\n",
      "chr2 CXCR2.csv\n",
      "chr2 CYBRD1.csv\n",
      "chr2 CYP1B1-AS1.csv\n",
      "chr2 CYP20A1.csv\n",
      "chr2 CYP27A1.csv\n",
      "chr2 CYTOR.csv\n",
      "chr2 D2HGDH.csv\n",
      "chr2 DAPL1.csv\n",
      "chr2 DARS1-AS1.csv\n",
      "chr2 DAW1.csv\n",
      "chr2 DBI.csv\n",
      "chr2 DCAF17.csv\n",
      "chr2 DCDC2C.csv\n",
      "chr2 DCTN1-AS1.csv\n",
      "chr2 DDX1.csv\n",
      "chr2 DDX18.csv\n",
      "chr2 DES.csv\n",
      "chr2 DGKD.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  10%|█         | 2017/19235 [00:07<01:21, 210.70it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 DGUOK.csv\n",
      "chr2 DHRS9.csv\n",
      "chr2 DIRC3-AS1.csv\n",
      "chr2 DIS3L2.csv\n",
      "chr2 DLX1.csv\n",
      "chr2 DLX2-DT.csv\n",
      "chr2 DNAH6.csv\n",
      "chr2 DNAJB2.csv\n",
      "chr2 DNAJC10.csv\n",
      "chr2 DNAJC27-AS1.csv\n",
      "chr2 DNAJC5G.csv\n",
      "chr2 DNPEP-AS1.csv\n",
      "chr2 DOK1.csv\n",
      "chr2 DPP10.csv\n",
      "chr2 DPP4-DT.csv\n",
      "chr2 DPYSL5.csv\n",
      "chr2 DRC1.csv\n",
      "chr2 DTNB-AS1.csv\n",
      "chr2 DUSP19.csv\n",
      "chr2 DUSP28.csv\n",
      "chr2 DYNC1I2.csv\n",
      "chr2 DYNC2LI1.csv\n",
      "chr2 DYSF.csv\n",
      "chr2 ECRG4.csv\n",
      "chr2 EEF1B2.csv\n",
      "chr2 EFHD1.csv\n",
      "chr2 EFR3B.csv\n",
      "chr2 EHBP1.csv\n",
      "chr2 EHD3.csv\n",
      "chr2 EIF2AK3-DT.csv\n",
      "chr2 EIF4E2.csv\n",
      "chr2 EIF5B.csv\n",
      "chr2 ELMOD3.csv\n",
      "chr2 EMILIN1.csv\n",
      "chr2 EML4.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  11%|█         | 2060/19235 [00:07<01:24, 203.61it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 EML6.csv\n",
      "chr2 EMX1.csv\n",
      "chr2 EPAS1.csv\n",
      "chr2 EPB41L5.csv\n",
      "chr2 EPC2.csv\n",
      "chr2 EPCAM.csv\n",
      "chr2 ERFE.csv\n",
      "chr2 ERICH2.csv\n",
      "chr2 ERLEC1.csv\n",
      "chr2 ESPNL.csv\n",
      "chr2 ETAA1.csv\n",
      "chr2 EVA1A-AS.csv\n",
      "chr2 FAHD2A.csv\n",
      "chr2 FAM117B.csv\n",
      "chr2 FAM138B.csv\n",
      "chr2 FAM166C.csv\n",
      "chr2 FAM171B.csv\n",
      "chr2 FAM201B.csv\n",
      "chr2 FAM228A.csv\n",
      "chr2 FAM228B.csv\n",
      "chr2 FAM237A.csv\n",
      "chr2 FARP2.csv\n",
      "chr2 FASTKD2.csv\n",
      "chr2 FBLN7.csv\n",
      "chr2 FBXO36.csv\n",
      "chr2 FER1L5.csv\n",
      "chr2 FKBP1B.csv\n",
      "chr2 FMNL2.csv\n",
      "chr2 FN1-DT.csv\n",
      "chr2 FOSL2.csv\n",
      "chr2 FOXD4L1.csv\n",
      "chr2 FOXN2.csv\n",
      "chr2 FSIP2.csv\n",
      "chr2 FZD7.csv\n",
      "chr2 G6PC2.csv\n",
      "chr2 GACAT1.csv\n",
      "chr2 GACAT3.csv\n",
      "chr2 GAD1.csv\n",
      "chr2 GAL3ST2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  11%|█         | 2110/19235 [00:07<01:16, 225.15it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 GALM.csv\n",
      "chr2 GALNT13.csv\n",
      "chr2 GALNT5.csv\n",
      "chr2 GAREM2.csv\n",
      "chr2 GCA.csv\n",
      "chr2 GCC2.csv\n",
      "chr2 GCKR.csv\n",
      "chr2 GCSIR.csv\n",
      "chr2 GDF7.csv\n",
      "chr2 GEMIN6.csv\n",
      "chr2 GEN1.csv\n",
      "chr2 GIGYF2.csv\n",
      "chr2 GKN1.csv\n",
      "chr2 GLI2.csv\n",
      "chr2 GLS.csv\n",
      "chr2 GMCL1.csv\n",
      "chr2 GMPPA.csv\n",
      "chr2 GNLY.csv\n",
      "chr2 GORASP2.csv\n",
      "chr2 GPATCH11.csv\n",
      "chr2 GPBAR1.csv\n",
      "chr2 GPC1.csv\n",
      "chr2 GPD2.csv\n",
      "chr2 GPN1.csv\n",
      "chr2 GPR148.csv\n",
      "chr2 GPR155-DT.csv\n",
      "chr2 GPR17.csv\n",
      "chr2 GPR35.csv\n",
      "chr2 GPR39.csv\n",
      "chr2 GPR45.csv\n",
      "chr2 GREB1.csv\n",
      "chr2 GRHL1.csv\n",
      "chr2 GTF2A1L.csv\n",
      "chr2 GTF3C2-AS1.csv\n",
      "chr2 GTF3C2-AS2.csv\n",
      "chr2 GULP1.csv\n",
      "chr2 GYPC.csv\n",
      "chr2 HADHB.csv\n",
      "chr2 HAGLROS.csv\n",
      "chr2 HAT1.csv\n",
      "chr2 HDAC4-AS1.csv\n",
      "chr2 HECW2-AS1.csv\n",
      "chr2 HK2.csv\n",
      "chr2 HNMT.csv\n",
      "chr2 HNRNPA3.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  11%|█         | 2133/19235 [00:07<01:16, 222.72it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 HOXD1.csv\n",
      "chr2 HOXD10.csv\n",
      "chr2 HOXD11.csv\n",
      "chr2 HOXD12.csv\n",
      "chr2 HOXD13.csv\n",
      "chr2 HOXD3.csv\n",
      "chr2 HOXD4.csv\n",
      "chr2 HOXD8.csv\n",
      "chr2 HOXD9.csv\n",
      "chr2 HPCAL1.csv\n",
      "chr2 HSPE1-MOB4.csv\n",
      "chr2 HSPE1.csv\n",
      "chr2 HTRA2.csv\n",
      "chr2 IAH1.csv\n",
      "chr2 ICOS.csv\n",
      "chr2 ID2.csv\n",
      "chr2 IDH1-AS1.csv\n",
      "chr2 IGFBP-AS1.csv\n",
      "chr2 IGFBP2.csv\n",
      "chr2 IGK.csv\n",
      "chr2 IGKV1D-12.csv\n",
      "chr2 IGKV1D-13.csv\n",
      "chr2 IGKV1D-16.csv\n",
      "chr2 IGKV1D-17.csv\n",
      "chr2 IGKV1D-33.csv\n",
      "chr2 IGKV1D-37.csv\n",
      "chr2 IGKV1D-39.csv\n",
      "chr2 IGKV1D-42.csv\n",
      "chr2 IGKV1D-43.csv\n",
      "chr2 IGKV1D-8.csv\n",
      "chr2 IGKV2D-24.csv\n",
      "chr2 IGKV2D-28.csv\n",
      "chr2 IGKV2D-29.csv\n",
      "chr2 IGKV2D-30.csv\n",
      "chr2 IGKV2D-40.csv\n",
      "chr2 IGKV3D-11.csv\n",
      "chr2 IGKV3D-15.csv\n",
      "chr2 IGKV3D-20.csv\n",
      "chr2 IGKV3D-7.csv\n",
      "chr2 IGKV4-1.csv\n",
      "chr2 IGKV5-2.csv\n",
      "chr2 IGKV6D-21.csv\n",
      "chr2 IGKV6D-41.csv\n",
      "chr2 IL18R1.csv\n",
      "chr2 IL18RAP.csv\n",
      "chr2 IL1F10.csv\n",
      "chr2 IL1R1.csv\n",
      "chr2 IL1R2.csv\n",
      "chr2 IL1RL1.csv\n",
      "chr2 IL1RL2.csv\n",
      "chr2 IL1RN.csv\n",
      "chr2 IL36A.csv\n",
      "chr2 IL36G.csv\n",
      "chr2 IL36RN.csv\n",
      "chr2 IL37.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  11%|█▏        | 2200/19235 [00:08<01:05, 259.24it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 IMP4.csv\n",
      "chr2 ING5.csv\n",
      "chr2 INHA.csv\n",
      "chr2 INHBB.csv\n",
      "chr2 INO80B-WBP1.csv\n",
      "chr2 INO80B.csv\n",
      "chr2 INPP1.csv\n",
      "chr2 INPP4A.csv\n",
      "chr2 INPP5D.csv\n",
      "chr2 INSIG2.csv\n",
      "chr2 ITGA4.csv\n",
      "chr2 ITGA6.csv\n",
      "chr2 ITGAV.csv\n",
      "chr2 ITM2C.csv\n",
      "chr2 ITPRID2.csv\n",
      "chr2 ITPRIPL1.csv\n",
      "chr2 KANSL1L-AS1.csv\n",
      "chr2 KCMF1.csv\n",
      "chr2 KCNE4.csv\n",
      "chr2 KCNF1.csv\n",
      "chr2 KCNH7-AS1.csv\n",
      "chr2 KCNIP3.csv\n",
      "chr2 KCNJ3.csv\n",
      "chr2 KCNK3.csv\n",
      "chr2 KCNS3.csv\n",
      "chr2 KDM3A.csv\n",
      "chr2 KHK.csv\n",
      "chr2 KIAA2012.csv\n",
      "chr2 KIF5C.csv\n",
      "chr2 KLF11.csv\n",
      "chr2 KLHL23.csv\n",
      "chr2 KLHL29.csv\n",
      "chr2 KLHL30.csv\n",
      "chr2 KLHL41.csv\n",
      "chr2 KRTCAP3.csv\n",
      "chr2 KYNU.csv\n",
      "chr2 LANCL1-AS1.csv\n",
      "chr2 LAPTM4A-DT.csv\n",
      "chr2 LBH.csv\n",
      "chr2 LBX2-AS1.csv\n",
      "chr2 LCLAT1.csv\n",
      "chr2 LCT-AS1.csv\n",
      "chr2 LGALSL.csv\n",
      "chr2 LIMS1.csv\n",
      "chr2 LIMS3-LOC440895.csv\n",
      "chr2 "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  12%|█▏        | 2253/19235 [00:08<01:09, 243.50it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LIMS3.csv\n",
      "chr2 LINC00486.csv\n",
      "chr2 LINC00570.csv\n",
      "chr2 LINC00608.csv\n",
      "chr2 LINC00954.csv\n",
      "chr2 LINC01087.csv\n",
      "chr2 LINC01102.csv\n",
      "chr2 LINC01104.csv\n",
      "chr2 LINC01117.csv\n",
      "chr2 LINC01118.csv\n",
      "chr2 LINC01119.csv\n",
      "chr2 LINC01120.csv\n",
      "chr2 LINC01122.csv\n",
      "chr2 LINC01123.csv\n",
      "chr2 LINC01126.csv\n",
      "chr2 LINC01127.csv\n",
      "chr2 LINC01143.csv\n",
      "chr2 LINC01173.csv\n",
      "chr2 LINC01191.csv\n",
      "chr2 LINC01237.csv\n",
      "chr2 LINC01238.csv\n",
      "chr2 LINC01291.csv\n",
      "chr2 LINC01293.csv\n",
      "chr2 LINC01305.csv\n",
      "chr2 LINC01317.csv\n",
      "chr2 LINC01318.csv\n",
      "chr2 LINC01320.csv\n",
      "chr2 LINC01381.csv\n",
      "chr2 LINC01412.csv\n",
      "chr2 LINC01614.csv\n",
      "chr2 LINC01790.csv\n",
      "chr2 LINC01793.csv\n",
      "chr2 LINC01794.csv\n",
      "chr2 LINC01795.csv\n",
      "chr2 LINC01798.csv\n",
      "chr2 LINC01799.csv\n",
      "chr2 LINC01802.csv\n",
      "chr2 LINC01804.csv\n",
      "chr2 LINC01805.csv\n",
      "chr2 LINC01806.csv\n",
      "chr2 LINC01808.csv\n",
      "chr2 LINC01809.csv\n",
      "chr2 LINC01818.csv\n",
      "chr2 LINC01821.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  12%|█▏        | 2285/19235 [00:08<01:04, 263.91it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LINC01823.csv\n",
      "chr2 LINC01824.csv\n",
      "chr2 LINC01828.csv\n",
      "chr2 LINC01832.csv\n",
      "chr2 LINC01849.csv\n",
      "chr2 LINC01850.csv\n",
      "chr2 LINC01851.csv\n",
      "chr2 LINC01856.csv\n",
      "chr2 LINC01857.csv\n",
      "chr2 LINC01865.csv\n",
      "chr2 LINC01866.csv\n",
      "chr2 LINC01867.csv\n",
      "chr2 LINC01870.csv\n",
      "chr2 LINC01871.csv\n",
      "chr2 LINC01873.csv\n",
      "chr2 LINC01875.csv\n",
      "chr2 LINC01877.csv\n",
      "chr2 LINC01878.csv\n",
      "chr2 LINC01881.csv\n",
      "chr2 LINC01883.csv\n",
      "chr2 LINC01884.csv\n",
      "chr2 LINC01886.csv\n",
      "chr2 LINC01888.csv\n",
      "chr2 LINC01889.csv\n",
      "chr2 LINC01911.csv\n",
      "chr2 LINC01913.csv\n",
      "chr2 LINC01918.csv\n",
      "chr2 LINC01921.csv\n",
      "chr2 LINC01934.csv\n",
      "chr2 LINC01936.csv\n",
      "chr2 LINC01937.csv\n",
      "chr2 LINC01954.csv\n",
      "chr2 LINC01958.csv\n",
      "chr2 LINC01960.csv\n",
      "chr2 LINC01961.csv\n",
      "chr2 LINC01963.csv\n",
      "chr2 LINC01965.csv\n",
      "chr2 LINC02576.csv\n",
      "chr2 LINC02579.csv\n",
      "chr2 LINC02583.csv\n",
      "chr2 LINC02612.csv\n",
      "chr2 LINC02831.csv\n",
      "chr2 LINC02832.csv\n",
      "chr2 LINC02850.csv\n",
      "chr2 LINC02934.csv\n",
      "chr2 LIPT1.csv\n",
      "chr2 LOC100129175.csv\n",
      "chr2 LOC100505498.csv\n",
      "chr2 LOC100506076.csv\n",
      "chr2 LOC100506235.csv\n",
      "chr2 LOC100506274.csv\n",
      "chr2 LOC100506405.csv\n",
      "chr2 LOC100507006.csv\n",
      "chr2 LOC100507443.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  12%|█▏        | 2340/19235 [00:08<01:04, 262.36it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC100507562.csv\n",
      "chr2 LOC100996506.csv\n",
      "chr2 LOC100996637.csv\n",
      "chr2 LOC101926959.csv\n",
      "chr2 LOC101926969.csv\n",
      "chr2 LOC101926974.csv\n",
      "chr2 LOC101927053.csv\n",
      "chr2 LOC101927055.csv\n",
      "chr2 LOC101927073.csv\n",
      "chr2 LOC101927089.csv\n",
      "chr2 LOC101927213.csv\n",
      "chr2 LOC101927235.csv\n",
      "chr2 LOC101927262.csv\n",
      "chr2 LOC101927283.csv\n",
      "chr2 LOC101927366.csv\n",
      "chr2 LOC101927383.csv\n",
      "chr2 LOC101927400.csv\n",
      "chr2 LOC101927454.csv\n",
      "chr2 LOC101927509.csv\n",
      "chr2 LOC101927533.csv\n",
      "chr2 LOC101927661.csv\n",
      "chr2 LOC101927881.csv\n",
      "chr2 LOC101927948.csv\n",
      "chr2 LOC101927960.csv\n",
      "chr2 LOC101928156.csv\n",
      "chr2 LOC101928180.csv\n",
      "chr2 LOC101928361.csv\n",
      "chr2 LOC101928371.csv\n",
      "chr2 LOC101928386.csv\n",
      "chr2 LOC101928526.csv\n",
      "chr2 LOC101928881.csv\n",
      "chr2 LOC101929386.csv\n",
      "chr2 LOC101929510.csv\n",
      "chr2 LOC101929532.csv\n",
      "chr2 LOC101929633.csv\n",
      "chr2 LOC101929643.csv\n",
      "chr2 LOC101929667.csv\n",
      "chr2 LOC101929908.csv\n",
      "chr2 LOC102723413.csv\n",
      "chr2 LOC102723730.csv\n",
      "chr2 LOC102724008.csv\n",
      "chr2 LOC102724058.csv\n",
      "chr2 LOC102724072.csv\n",
      "chr2 LOC102724340.csv\n",
      "chr2 LOC102724389.csv\n",
      "chr2 LOC102724497.csv\n",
      "chr2 LOC102724542.csv\n",
      "chr2 LOC102724579.csv\n",
      "chr2 LOC102724744.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  12%|█▏        | 2394/19235 [00:08<01:04, 260.54it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC102724849.csv\n",
      "chr2 LOC102724861.csv\n",
      "chr2 LOC102724875.csv\n",
      "chr2 LOC102724955.csv\n",
      "chr2 LOC105369167.csv\n",
      "chr2 LOC105373351.csv\n",
      "chr2 LOC105373352.csv\n",
      "chr2 LOC105373389.csv\n",
      "chr2 LOC105373392.csv\n",
      "chr2 LOC105373394.csv\n",
      "chr2 LOC105373395.csv\n",
      "chr2 LOC105373397.csv\n",
      "chr2 LOC105373398.csv\n",
      "chr2 LOC105373406.csv\n",
      "chr2 LOC105373407.csv\n",
      "chr2 LOC105373409.csv\n",
      "chr2 LOC105373414.csv\n",
      "chr2 LOC105373416.csv\n",
      "chr2 LOC105373418.csv\n",
      "chr2 LOC105373419.csv\n",
      "chr2 LOC105373429.csv\n",
      "chr2 LOC105373430.csv\n",
      "chr2 LOC105373436.csv\n",
      "chr2 LOC105373437.csv\n",
      "chr2 LOC105373438.csv\n",
      "chr2 LOC105373440.csv\n",
      "chr2 LOC105373447.csv\n",
      "chr2 LOC105373454.csv\n",
      "chr2 LOC105373456.csv\n",
      "chr2 LOC105373461.csv\n",
      "chr2 LOC105373466.csv\n",
      "chr2 LOC105373480.csv\n",
      "chr2 LOC105373481.csv\n",
      "chr2 LOC105373484.csv\n",
      "chr2 LOC105373487.csv\n",
      "chr2 LOC105373488.csv\n",
      "chr2 LOC105373493.csv\n",
      "chr2 LOC105373504.csv\n",
      "chr2 LOC105373508.csv\n",
      "chr2 LOC105373510.csv\n",
      "chr2 LOC105373512.csv\n",
      "chr2 LOC105373525.csv\n",
      "chr2 LOC105373526.csv\n",
      "chr2 LOC105373527.csv\n",
      "chr2 LOC105373531.csv\n",
      "chr2 LOC105373534.csv\n",
      "chr2 LOC105373545.csv\n",
      "chr2 LOC105373551.csv\n",
      "chr2 LOC105373552.csv\n",
      "chr2 LOC105373553.csv\n",
      "chr2 LOC105373554.csv\n",
      "chr2 LOC105373559.csv\n",
      "chr2 LOC105373576.csv\n",
      "chr2 LOC105373587.csv\n",
      "chr2 LOC105373591.csv\n",
      "chr2 LOC105373592.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  13%|█▎        | 2457/19235 [00:09<00:58, 285.41it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC105373595.csv\n",
      "chr2 LOC105373601.csv\n",
      "chr2 LOC105373604.csv\n",
      "chr2 LOC105373605.csv\n",
      "chr2 LOC105373611.csv\n",
      "chr2 LOC105373612.csv\n",
      "chr2 LOC105373613.csv\n",
      "chr2 LOC105373616.csv\n",
      "chr2 LOC105373618.csv\n",
      "chr2 LOC105373622.csv\n",
      "chr2 LOC105373623.csv\n",
      "chr2 LOC105373629.csv\n",
      "chr2 LOC105373640.csv\n",
      "chr2 LOC105373641.csv\n",
      "chr2 LOC105373643.csv\n",
      "chr2 LOC105373645.csv\n",
      "chr2 LOC105373648.csv\n",
      "chr2 LOC105373649.csv\n",
      "chr2 LOC105373651.csv\n",
      "chr2 LOC105373663.csv\n",
      "chr2 LOC105373666.csv\n",
      "chr2 LOC105373668.csv\n",
      "chr2 LOC105373674.csv\n",
      "chr2 LOC105373679.csv\n",
      "chr2 LOC105373685.csv\n",
      "chr2 LOC105373689.csv\n",
      "chr2 LOC105373691.csv\n",
      "chr2 LOC105373693.csv\n",
      "chr2 LOC105373698.csv\n",
      "chr2 LOC105373699.csv\n",
      "chr2 LOC105373700.csv\n",
      "chr2 LOC105373707.csv\n",
      "chr2 LOC105373709.csv\n",
      "chr2 LOC105373714.csv\n",
      "chr2 LOC105373715.csv\n",
      "chr2 LOC105373716.csv\n",
      "chr2 LOC105373717.csv\n",
      "chr2 LOC105373724.csv\n",
      "chr2 LOC105373727.csv\n",
      "chr2 LOC105373728.csv\n",
      "chr2 LOC105373741.csv\n",
      "chr2 LOC105373752.csv\n",
      "chr2 LOC105373756.csv\n",
      "chr2 LOC105373766.csv\n",
      "chr2 LOC105373768.csv\n",
      "chr2 LOC105373774.csv\n",
      "chr2 LOC105373779.csv\n",
      "chr2 LOC105373781.csv\n",
      "chr2 LOC105373788.csv\n",
      "chr2 LOC105373790.csv\n",
      "chr2 LOC105373804.csv\n",
      "chr2 LOC105373812.csv\n",
      "chr2 LOC105373817.csv\n",
      "chr2 LOC105373819.csv\n",
      "chr2 LOC105373821.csv\n",
      "chr2 LOC105373822.csv\n",
      "chr2 LOC105373830.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  13%|█▎        | 2520/19235 [00:09<00:57, 292.17it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC105373835.csv\n",
      "chr2 LOC105373845.csv\n",
      "chr2 LOC105373851.csv\n",
      "chr2 LOC105373854.csv\n",
      "chr2 LOC105373864.csv\n",
      "chr2 LOC105373870.csv\n",
      "chr2 LOC105373874.csv\n",
      "chr2 LOC105373880.csv\n",
      "chr2 LOC105373887.csv\n",
      "chr2 LOC105373888.csv\n",
      "chr2 LOC105373890.csv\n",
      "chr2 LOC105373891.csv\n",
      "chr2 LOC105373893.csv\n",
      "chr2 LOC105373895.csv\n",
      "chr2 LOC105373896.csv\n",
      "chr2 LOC105373899.csv\n",
      "chr2 LOC105373902.csv\n",
      "chr2 LOC105373903.csv\n",
      "chr2 LOC105373907.csv\n",
      "chr2 LOC105373908.csv\n",
      "chr2 LOC105373909.csv\n",
      "chr2 LOC105373910.csv\n",
      "chr2 LOC105373918.csv\n",
      "chr2 LOC105373919.csv\n",
      "chr2 LOC105373920.csv\n",
      "chr2 LOC105373921.csv\n",
      "chr2 LOC105373928.csv\n",
      "chr2 LOC105373932.csv\n",
      "chr2 LOC105373937.csv\n",
      "chr2 LOC105373938.csv\n",
      "chr2 LOC105373940.csv\n",
      "chr2 LOC105373945.csv\n",
      "chr2 LOC105373947.csv\n",
      "chr2 LOC105373949.csv\n",
      "chr2 LOC105373950.csv\n",
      "chr2 LOC105373952.csv\n",
      "chr2 LOC105373953.csv\n",
      "chr2 LOC105373958.csv\n",
      "chr2 LOC105373966.csv\n",
      "chr2 LOC105373976.csv\n",
      "chr2 LOC105373977.csv\n",
      "chr2 LOC105373989.csv\n",
      "chr2 LOC105373991.csv\n",
      "chr2 LOC105374317.csv\n",
      "chr2 LOC105374318.csv\n",
      "chr2 LOC105374320.csv\n",
      "chr2 LOC105374329.csv\n",
      "chr2 LOC105374333.csv\n",
      "chr2 LOC105374334.csv\n",
      "chr2 LOC105374369.csv\n",
      "chr2 LOC105374378.csv\n",
      "chr2 LOC105374381.csv\n",
      "chr2 LOC105374382.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  13%|█▎        | 2550/19235 [00:09<00:57, 288.57it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC105374383.csv\n",
      "chr2 LOC105374435.csv\n",
      "chr2 LOC105374455.csv\n",
      "chr2 LOC105374464.csv\n",
      "chr2 LOC105374468.csv\n",
      "chr2 LOC105374470.csv\n",
      "chr2 LOC105374491.csv\n",
      "chr2 LOC105374567.csv\n",
      "chr2 LOC105374570.csv\n",
      "chr2 LOC105374571.csv\n",
      "chr2 LOC105374573.csv\n",
      "chr2 LOC105374576.csv\n",
      "chr2 LOC105374577.csv\n",
      "chr2 LOC105374584.csv\n",
      "chr2 LOC105374588.csv\n",
      "chr2 LOC105374589.csv\n",
      "chr2 LOC105374590.csv\n",
      "chr2 LOC105374592.csv\n",
      "chr2 LOC105374594.csv\n",
      "chr2 LOC105374595.csv\n",
      "chr2 LOC105374615.csv\n",
      "chr2 LOC105374653.csv\n",
      "chr2 LOC105374662.csv\n",
      "chr2 LOC105374690.csv\n",
      "chr2 LOC105374697.csv\n",
      "chr2 LOC105374760.csv\n",
      "chr2 LOC105374761.csv\n",
      "chr2 LOC105374764.csv\n",
      "chr2 LOC105374775.csv\n",
      "chr2 LOC105374776.csv\n",
      "chr2 LOC105374794.csv\n",
      "chr2 LOC105374796.csv\n",
      "chr2 LOC105374803.csv\n",
      "chr2 LOC105374808.csv\n",
      "chr2 LOC105374810.csv\n",
      "chr2 LOC105374811.csv\n",
      "chr2 LOC105374813.csv\n",
      "chr2 LOC105374814.csv\n",
      "chr2 LOC105374815.csv\n",
      "chr2 LOC105374820.csv\n",
      "chr2 LOC105374821.csv\n",
      "chr2 LOC105374829.csv\n",
      "chr2 LOC105374831.csv\n",
      "chr2 LOC105374834.csv\n",
      "chr2 LOC105374837.csv\n",
      "chr2 LOC105374838.csv\n",
      "chr2 LOC105374839.csv\n",
      "chr2 LOC105374841.csv\n",
      "chr2 LOC105374842.csv\n",
      "chr2 LOC105374846.csv\n",
      "chr2 LOC105374851.csv\n",
      "chr2 LOC105374852.csv\n",
      "chr2 LOC105374854.csv\n",
      "chr2 LOC105374862.csv\n",
      "chr2 LOC105374868.csv\n",
      "chr2 LOC105375311.csv\n",
      "chr2 LOC105375809.csv\n",
      "chr2 LOC105376754.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  14%|█▎        | 2618/19235 [00:09<00:53, 311.35it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC105376787.csv\n",
      "chr2 LOC105376810.csv\n",
      "chr2 LOC105377635.csv\n",
      "chr2 LOC107984020.csv\n",
      "chr2 LOC107984111.csv\n",
      "chr2 LOC107985769.csv\n",
      "chr2 LOC107985771.csv\n",
      "chr2 LOC107985772.csv\n",
      "chr2 LOC107985774.csv\n",
      "chr2 LOC107985776.csv\n",
      "chr2 LOC107985779.csv\n",
      "chr2 LOC107985780.csv\n",
      "chr2 LOC107985782.csv\n",
      "chr2 LOC107985784.csv\n",
      "chr2 LOC107985788.csv\n",
      "chr2 LOC107985795.csv\n",
      "chr2 LOC107985807.csv\n",
      "chr2 LOC107985808.csv\n",
      "chr2 LOC107985811.csv\n",
      "chr2 LOC107985812.csv\n",
      "chr2 LOC107985815.csv\n",
      "chr2 LOC107985817.csv\n",
      "chr2 LOC107985821.csv\n",
      "chr2 LOC107985822.csv\n",
      "chr2 LOC107985830.csv\n",
      "chr2 LOC107985831.csv\n",
      "chr2 LOC107985833.csv\n",
      "chr2 LOC107985838.csv\n",
      "chr2 LOC107985842.csv\n",
      "chr2 LOC107985845.csv\n",
      "chr2 LOC107985847.csv\n",
      "chr2 LOC107985850.csv\n",
      "chr2 LOC107985851.csv\n",
      "chr2 LOC107985852.csv\n",
      "chr2 LOC107985853.csv\n",
      "chr2 LOC107985854.csv\n",
      "chr2 LOC107985855.csv\n",
      "chr2 LOC107985857.csv\n",
      "chr2 LOC107985869.csv\n",
      "chr2 LOC107985877.csv\n",
      "chr2 LOC107985883.csv\n",
      "chr2 LOC107985889.csv\n",
      "chr2 LOC107985892.csv\n",
      "chr2 LOC107985902.csv\n",
      "chr2 LOC107985903.csv\n",
      "chr2 LOC107985904.csv\n",
      "chr2 LOC107985908.csv\n",
      "chr2 LOC107985915.csv\n",
      "chr2 LOC107985916.csv\n",
      "chr2 LOC107985920.csv\n",
      "chr2 LOC107985921.csv\n",
      "chr2 LOC107985922.csv\n",
      "chr2 LOC107985923.csv\n",
      "chr2 LOC107985926.csv\n",
      "chr2 LOC107985929.csv\n",
      "chr2 LOC107985934.csv\n",
      "chr2 LOC107985938.csv\n",
      "chr2 LOC107985939.csv\n",
      "chr2 LOC107985943.csv\n",
      "chr2 LOC107985944.csv\n",
      "chr2 LOC107985947.csv\n",
      "chr2 LOC107985949.csv\n",
      "chr2 LOC107985950.csv\n",
      "chr2 LOC107985952.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  14%|█▍        | 2683/19235 [00:09<00:53, 312.23it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LOC107985953.csv\n",
      "chr2 LOC107985955.csv\n",
      "chr2 LOC107985959.csv\n",
      "chr2 LOC107985960.csv\n",
      "chr2 LOC107985961.csv\n",
      "chr2 LOC107985968.csv\n",
      "chr2 LOC107985972.csv\n",
      "chr2 LOC107985976.csv\n",
      "chr2 LOC107985978.csv\n",
      "chr2 LOC107985980.csv\n",
      "chr2 LOC107985981.csv\n",
      "chr2 LOC107985983.csv\n",
      "chr2 LOC107985984.csv\n",
      "chr2 LOC107985986.csv\n",
      "chr2 LOC107985990.csv\n",
      "chr2 LOC107985991.csv\n",
      "chr2 LOC107985992.csv\n",
      "chr2 LOC107985995.csv\n",
      "chr2 LOC107985996.csv\n",
      "chr2 LOC107985997.csv\n",
      "chr2 LOC107986000.csv\n",
      "chr2 LOC107986001.csv\n",
      "chr2 LOC107986002.csv\n",
      "chr2 LOC112268411.csv\n",
      "chr2 LOC112268412.csv\n",
      "chr2 LOC112268413.csv\n",
      "chr2 LOC112268416.csv\n",
      "chr2 LOC112268418.csv\n",
      "chr2 LOC112268420.csv\n",
      "chr2 LOC112268422.csv\n",
      "chr2 LOC112268426.csv\n",
      "chr2 LOC112268429.csv\n",
      "chr2 LOC112268430.csv\n",
      "chr2 LOC112268433.csv\n",
      "chr2 LOC112268434.csv\n",
      "chr2 LOC112268436.csv\n",
      "chr2 LOC112268438.csv\n",
      "chr2 LOC112268439.csv\n",
      "chr2 LOC150935.csv\n",
      "chr2 LOC284950.csv\n",
      "chr2 LOC285097.csv\n",
      "chr2 LOC285191.csv\n",
      "chr2 LOC344065.csv\n",
      "chr2 LOC400940.csv\n",
      "chr2 LOC401021.csv\n",
      "chr2 LOC401040.csv\n",
      "chr2 LOC440864.csv\n",
      "chr2 LOC440910.csv\n",
      "chr2 LOC646736.csv\n",
      "chr2 LOC729254.csv\n",
      "chr2 LOC730100.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  14%|█▍        | 2715/19235 [00:10<01:03, 260.33it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 LPIN1.csv\n",
      "chr2 LRATD1.csv\n",
      "chr2 LRRFIP1.csv\n",
      "chr2 LRRTM4-AS1.csv\n",
      "chr2 LTBP1.csv\n",
      "chr2 LYPD6.csv\n",
      "chr2 LYPD6B.csv\n",
      "chr2 MAIP1.csv\n",
      "chr2 MAL.csv\n",
      "chr2 MAP2.csv\n",
      "chr2 MAP3K2-DT.csv\n",
      "chr2 MAP3K20.csv\n",
      "chr2 MAP4K3-DT.csv\n",
      "chr2 MAP4K4.csv\n",
      "chr2 MAPRE3.csv\n",
      "chr2 MARCHF7.csv\n",
      "chr2 MARCO.csv\n",
      "chr2 MARS2.csv\n",
      "chr2 MAT2A.csv\n",
      "chr2 MBD5.csv\n",
      "chr2 MDH1.csv\n",
      "chr2 MEIS1.csv\n",
      "chr2 MERTK.csv\n",
      "chr2 METAP1D.csv\n",
      "chr2 MFF.csv\n",
      "chr2 MFSD2B.csv\n",
      "chr2 MFSD6.csv\n",
      "chr2 MGAT5.csv\n",
      "chr2 MIR10B.csv\n",
      "chr2 MIR1244-1.csv\n",
      "chr2 MIR1245A.csv\n",
      "chr2 MIR128-1.csv\n",
      "chr2 MIR149.csv\n",
      "chr2 MIR26B.csv\n",
      "chr2 MIR3125.csv\n",
      "chr2 MIR3126.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  14%|█▍        | 2784/19235 [00:10<00:55, 294.70it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 MIR3127.csv\n",
      "chr2 MIR3130-2.csv\n",
      "chr2 MIR3133.csv\n",
      "chr2 MIR3606.csv\n",
      "chr2 MIR3679.csv\n",
      "chr2 MIR3681.csv\n",
      "chr2 MIR3681HG.csv\n",
      "chr2 MIR4263.csv\n",
      "chr2 MIR4269.csv\n",
      "chr2 MIR4430.csv\n",
      "chr2 MIR4433A.csv\n",
      "chr2 MIR4434.csv\n",
      "chr2 MIR4435-1.csv\n",
      "chr2 MIR4436A.csv\n",
      "chr2 MIR4436B2.csv\n",
      "chr2 MIR4438.csv\n",
      "chr2 MIR4444-1.csv\n",
      "chr2 MIR4757.csv\n",
      "chr2 MIR4765.csv\n",
      "chr2 MIR4771-1.csv\n",
      "chr2 MIR4772.csv\n",
      "chr2 MIR4773-1.csv\n",
      "chr2 MIR4774.csv\n",
      "chr2 MIR4775.csv\n",
      "chr2 MIR4776-1.csv\n",
      "chr2 MIR4777.csv\n",
      "chr2 MIR5000.csv\n",
      "chr2 MIR5192.csv\n",
      "chr2 MIR548AD.csv\n",
      "chr2 MIR548AE1.csv\n",
      "chr2 MIR548BA.csv\n",
      "chr2 MIR548S.csv\n",
      "chr2 MIR558.csv\n",
      "chr2 MIR559.csv\n",
      "chr2 MIR5590.csv\n",
      "chr2 MIR561.csv\n",
      "chr2 MIR562.csv\n",
      "chr2 MIR5696.csv\n",
      "chr2 MIR5703.csv\n",
      "chr2 MIR6810.csv\n",
      "chr2 MIR6811.csv\n",
      "chr2 MIR6888.csv\n",
      "chr2 MIR7158.csv\n",
      "chr2 MIR7515.csv\n",
      "chr2 MIR7704.csv\n",
      "chr2 MIR7845.csv\n",
      "chr2 MIR9500.csv\n",
      "chr2 MIR9986.csv\n",
      "chr2 MLPH.csv\n",
      "chr2 MMADHC-DT.csv\n",
      "chr2 MOB4.csv\n",
      "chr2 MOGAT1.csv\n",
      "chr2 MORN2.csv\n",
      "chr2 MPHOSPH10.csv\n",
      "chr2 MROH2A.csv\n",
      "chr2 MRPL19.csv\n",
      "chr2 MRPL30.csv\n",
      "chr2 MRPL33.csv\n",
      "chr2 MRPL35.csv\n",
      "chr2 MRPL44.csv\n",
      "chr2 MRPS9.csv\n",
      "chr2 MSGN1.csv\n",
      "chr2 MSH2-OT1.csv\n",
      "chr2 MSH2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  15%|█▍        | 2815/19235 [00:10<00:58, 279.86it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 MSH6.csv\n",
      "chr2 MTA3.csv\n",
      "chr2 MTHFD2.csv\n",
      "chr2 MTX2.csv\n",
      "chr2 MXD1.csv\n",
      "chr2 MYCN.csv\n",
      "chr2 MYCNUT.csv\n",
      "chr2 MYO1B.csv\n",
      "chr2 MYO3B.csv\n",
      "chr2 MYO7B.csv\n",
      "chr2 MYOSLID.csv\n",
      "chr2 MYT1L-AS1.csv\n",
      "chr2 MZT2B.csv\n",
      "chr2 NAB1.csv\n",
      "chr2 NABP1-OT1.csv\n",
      "chr2 NABP1.csv\n",
      "chr2 NAGK.csv\n",
      "chr2 NBEAL1.csv\n",
      "chr2 NCAPH.csv\n",
      "chr2 NCK2.csv\n",
      "chr2 NCKAP5-AS1.csv\n",
      "chr2 NCKAP5-AS2.csv\n",
      "chr2 NCOA1.csv\n",
      "chr2 NDUFAF7.csv\n",
      "chr2 NDUFB3.csv\n",
      "chr2 NEMP2-DT.csv\n",
      "chr2 NEU2.csv\n",
      "chr2 NEU4.csv\n",
      "chr2 NIF3L1.csv\n",
      "chr2 NIFK-AS1.csv\n",
      "chr2 NMS.csv\n",
      "chr2 NOP58.csv\n",
      "chr2 NOSTRIN.csv\n",
      "chr2 NOTO.csv\n",
      "chr2 NPAS2.csv\n",
      "chr2 NRBP1.csv\n",
      "chr2 NRP2.csv\n",
      "chr2 NT5DC4.csv\n",
      "chr2 NUP35.csv\n",
      "chr2 NYAP2.csv\n",
      "chr2 ODC1-DT.csv\n",
      "chr2 OSBPL6.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  15%|█▍        | 2872/19235 [00:10<01:13, 222.62it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 OSGEPL1-AS1.csv\n",
      "chr2 OTX1.csv\n",
      "chr2 PAPOLG.csv\n",
      "chr2 PARD3B.csv\n",
      "chr2 PAX8-AS1.csv\n",
      "chr2 PCBP1.csv\n",
      "chr2 PCGEM1.csv\n",
      "chr2 PCYOX1.csv\n",
      "chr2 PDCL3.csv\n",
      "chr2 PDE11A-AS1.csv\n",
      "chr2 PDK1.csv\n",
      "chr2 PEX13.csv\n",
      "chr2 PHOSPHO2-KLHL23.csv\n",
      "chr2 PHOSPHO2.csv\n",
      "chr2 PIKFYVE.csv\n",
      "chr2 PJVK.csv\n",
      "chr2 PKDCC.csv\n",
      "chr2 PKP4.csv\n",
      "chr2 PLB1.csv\n",
      "chr2 PLCD4.csv\n",
      "chr2 PLCL1.csv\n",
      "chr2 PLEK.csv\n",
      "chr2 PLEKHA3.csv\n",
      "chr2 PLEKHB2.csv\n",
      "chr2 PLEKHH2.csv\n",
      "chr2 PLGLB2.csv\n",
      "chr2 PMS1.csv\n",
      "chr2 PNKD.csv\n",
      "chr2 PNO1.csv\n",
      "chr2 POLE4.csv\n",
      "chr2 POLR1B.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  15%|█▌        | 2896/19235 [00:10<01:12, 225.14it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 POTEE.csv\n",
      "chr2 POTEF-AS1.csv\n",
      "chr2 POTEJ.csv\n",
      "chr2 POU3F3.csv\n",
      "chr2 PPIG.csv\n",
      "chr2 PPM1B.csv\n",
      "chr2 PPP1CB.csv\n",
      "chr2 PPP1R1C.csv\n",
      "chr2 PPP1R21.csv\n",
      "chr2 PPP1R7.csv\n",
      "chr2 PPP4R3B-DT.csv\n",
      "chr2 PRKCE.csv\n",
      "chr2 PRKD3-DT.csv\n",
      "chr2 PRLH.csv\n",
      "chr2 PROC.csv\n",
      "chr2 PROKR1.csv\n",
      "chr2 PROM2.csv\n",
      "chr2 PRSS56.csv\n",
      "chr2 PSD4.csv\n",
      "chr2 PSMD1.csv\n",
      "chr2 PSMD14.csv\n",
      "chr2 PTCD3.csv\n",
      "chr2 PTH2R.csv\n",
      "chr2 PTMA.csv\n",
      "chr2 PTPN18.csv\n",
      "chr2 PTPN4.csv\n",
      "chr2 QPCT.csv\n",
      "chr2 R3HDM1.csv\n",
      "chr2 RAB10.csv\n",
      "chr2 RAB3GAP1.csv\n",
      "chr2 RAB6C.csv\n",
      "chr2 RABL2A.csv\n",
      "chr2 RALB.csv\n",
      "chr2 RAMP1.csv\n",
      "chr2 RANBP2.csv\n",
      "chr2 RAPGEF4.csv\n",
      "chr2 RASGRP3.csv\n",
      "chr2 RBM44.csv\n",
      "chr2 RBM45.csv\n",
      "chr2 REG1A.csv\n",
      "chr2 REG3G.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  15%|█▌        | 2945/19235 [00:11<01:11, 228.11it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 REL.csv\n",
      "chr2 RETREG2.csv\n",
      "chr2 RGPD1.csv\n",
      "chr2 RGPD4.csv\n",
      "chr2 RGPD5.csv\n",
      "chr2 RHBDD1.csv\n",
      "chr2 RHOB.csv\n",
      "chr2 RHOQ.csv\n",
      "chr2 RIF1.csv\n",
      "chr2 RMDN2.csv\n",
      "chr2 RMND5A.csv\n",
      "chr2 RNASEH1-DT.csv\n",
      "chr2 RNF144A.csv\n",
      "chr2 RNF181.csv\n",
      "chr2 RNPEPL1.csv\n",
      "chr2 RNU4ATAC.csv\n",
      "chr2 RPE.csv\n",
      "chr2 RPIA.csv\n",
      "chr2 RPL31.csv\n",
      "chr2 RPL37A.csv\n",
      "chr2 RPS27A.csv\n",
      "chr2 RPS7.csv\n",
      "chr2 RRM2.csv\n",
      "chr2 RSAD2.csv\n",
      "chr2 RTP5.csv\n",
      "chr2 RUFY4.csv\n",
      "chr2 SAG.csv\n",
      "chr2 SANBR.csv\n",
      "chr2 SATB2-AS1.csv\n",
      "chr2 SCARNA5.csv\n",
      "chr2 SCARNA6.csv\n",
      "chr2 SCHLAP1.csv\n",
      "chr2 SCLY.csv\n",
      "chr2 SCN1A-AS1.csv\n",
      "chr2 SCN2A.csv\n",
      "chr2 SCRN3.csv\n",
      "chr2 SCTR-AS1.csv\n",
      "chr2 SCYGR2.csv\n",
      "chr2 SCYGR4.csv\n",
      "chr2 SCYGR5.csv\n",
      "chr2 SCYGR7.csv\n",
      "chr2 SCYGR8.csv\n",
      "chr2 SELENOI.csv\n",
      "chr2 SEMA4F.csv\n",
      "chr2 SEPTIN2.csv\n",
      "chr2 SFT2D3.csv\n",
      "chr2 SGO2.csv\n",
      "chr2 SGPP2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  16%|█▌        | 2995/19235 [00:11<01:09, 234.92it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 SH2D6.csv\n",
      "chr2 SH3BP4.csv\n",
      "chr2 SH3RF3.csv\n",
      "chr2 SILC1.csv\n",
      "chr2 SIX3.csv\n",
      "chr2 SLC11A1.csv\n",
      "chr2 SLC1A4.csv\n",
      "chr2 SLC20A1.csv\n",
      "chr2 SLC30A6.csv\n",
      "chr2 SLC35F6.csv\n",
      "chr2 SLC39A10.csv\n",
      "chr2 SLC3A1.csv\n",
      "chr2 SLC4A10.csv\n",
      "chr2 SLC4A1AP.csv\n",
      "chr2 SLC4A3.csv\n",
      "chr2 SLC5A7.csv\n",
      "chr2 SLC66A3.csv\n",
      "chr2 SLC8A1-AS1.csv\n",
      "chr2 SLC9A2.csv\n",
      "chr2 SLC9A4.csv\n",
      "chr2 SMARCAL1.csv\n",
      "chr2 SMIM39.csv\n",
      "chr2 SMYD1.csv\n",
      "chr2 SMYD5.csv\n",
      "chr2 SNED1.csv\n",
      "chr2 SNHG31.csv\n",
      "chr2 SNORA10B.csv\n",
      "chr2 SNORA112.csv\n",
      "chr2 SNORA40B.csv\n",
      "chr2 SNORA41.csv\n",
      "chr2 SNORA70I.csv\n",
      "chr2 SNORC.csv\n",
      "chr2 SNORD11.csv\n",
      "chr2 SNORD11B.csv\n",
      "chr2 SNORD3K.csv\n",
      "chr2 SNORD51.csv\n",
      "chr2 SNORD53.csv\n",
      "chr2 SNORD53B.csv\n",
      "chr2 SNORD70.csv\n",
      "chr2 SNORD70B.csv\n",
      "chr2 SNORD92.csv\n",
      "chr2 SNORD94.csv\n",
      "chr2 SNRNP27.csv\n",
      "chr2 SNTG2.csv\n",
      "chr2 SNX17.csv\n",
      "chr2 SOCAR.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  16%|█▌        | 3043/19235 [00:11<01:10, 230.35it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 SOCS5.csv\n",
      "chr2 SOWAHC.csv\n",
      "chr2 SOX11.csv\n",
      "chr2 SP100.csv\n",
      "chr2 SP140.csv\n",
      "chr2 SP140L.csv\n",
      "chr2 SP5.csv\n",
      "chr2 SP9.csv\n",
      "chr2 SPAG16.csv\n",
      "chr2 SPAST.csv\n",
      "chr2 SPATA3.csv\n",
      "chr2 SPATS2L.csv\n",
      "chr2 SPDYA.csv\n",
      "chr2 SPEG.csv\n",
      "chr2 SPEGNB.csv\n",
      "chr2 SPOPL.csv\n",
      "chr2 SPP2.csv\n",
      "chr2 SPR.csv\n",
      "chr2 SPTBN1.csv\n",
      "chr2 SSB.csv\n",
      "chr2 ST3GAL5-AS1.csv\n",
      "chr2 STAMBP.csv\n",
      "chr2 STARD7-AS1.csv\n",
      "chr2 STAT4-AS1.csv\n",
      "chr2 STEAP3.csv\n",
      "chr2 STK11IP.csv\n",
      "chr2 STK16.csv\n",
      "chr2 STK36.csv\n",
      "chr2 STON1-GTF2A1L.csv\n",
      "chr2 STON1.csv\n",
      "chr2 STRADB.csv\n",
      "chr2 SULT1C2.csv\n",
      "chr2 SULT1C3.csv\n",
      "chr2 SULT1C4.csv\n",
      "chr2 TAF1B.csv\n",
      "chr2 TANC1.csv\n",
      "chr2 TANK.csv\n",
      "chr2 TBC1D8-AS1.csv\n",
      "chr2 TBR1.csv\n",
      "chr2 TCF23.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  16%|█▌        | 3067/19235 [00:11<01:10, 230.45it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 TCF7L1-IT1.csv\n",
      "chr2 TCF7L1.csv\n",
      "chr2 TDRD15.csv\n",
      "chr2 TEKT4.csv\n",
      "chr2 TET3.csv\n",
      "chr2 TEX37.csv\n",
      "chr2 TEX41.csv\n",
      "chr2 TEX44.csv\n",
      "chr2 TEX51.csv\n",
      "chr2 THNSL2.csv\n",
      "chr2 THSD7B.csv\n",
      "chr2 TLX2.csv\n",
      "chr2 TMEM169.csv\n",
      "chr2 TMEM177.csv\n",
      "chr2 TMEM178A.csv\n",
      "chr2 TMEM18-DT.csv\n",
      "chr2 TMEM182.csv\n",
      "chr2 TMEM198.csv\n",
      "chr2 TMEM214.csv\n",
      "chr2 TMEM247.csv\n",
      "chr2 TMEM37.csv\n",
      "chr2 TMEM87B.csv\n",
      "chr2 TMSB10.csv\n",
      "chr2 TNFAIP6.csv\n",
      "chr2 TNS1-AS1.csv\n",
      "chr2 TOGARAM2.csv\n",
      "chr2 TPO.csv\n",
      "chr2 TRA-CGC3-1.csv\n",
      "chr2 TRAF3IP1.csv\n",
      "chr2 TRAPPC12.csv\n",
      "chr2 TRI-TAT2-1.csv\n",
      "chr2 TRIB2.csv\n",
      "chr2 TRIM43.csv\n",
      "chr2 TRIM54.csv\n",
      "chr2 TRPM8.csv\n",
      "chr2 TRY-ATA1-1.csv\n",
      "chr2 TRY-GTA2-1.csv\n",
      "chr2 TSN.csv\n",
      "chr2 TTC21B-AS1.csv\n",
      "chr2 TTC27.csv\n",
      "chr2 TTC31.csv\n",
      "chr2 TTC7A.csv\n",
      "chr2 TTL.csv\n",
      "chr2 TTLL4.csv\n",
      "chr2 TTN-AS1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  16%|█▌        | 3119/19235 [00:11<01:13, 219.83it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 TUBA3D.csv\n",
      "chr2 TUBA4B.csv\n",
      "chr2 TWIST2.csv\n",
      "chr2 UBE2E3.csv\n",
      "chr2 UBE2F-SCLY.csv\n",
      "chr2 UBE2F.csv\n",
      "chr2 UBR3.csv\n",
      "chr2 UBXN2A.csv\n",
      "chr2 UBXN4.csv\n",
      "chr2 UGGT1.csv\n",
      "chr2 UGP2.csv\n",
      "chr2 UGT1A.csv\n",
      "chr2 UGT1A1.csv\n",
      "chr2 UGT1A10.csv\n",
      "chr2 UGT1A3.csv\n",
      "chr2 UGT1A4.csv\n",
      "chr2 UGT1A5.csv\n",
      "chr2 UGT1A6.csv\n",
      "chr2 UGT1A7.csv\n",
      "chr2 UGT1A8.csv\n",
      "chr2 UGT1A9.csv\n",
      "chr2 UNC50.csv\n",
      "chr2 UNC80.csv\n",
      "chr2 UPP2.csv\n",
      "chr2 USP34-DT.csv\n",
      "chr2 USP39.csv\n",
      "chr2 VAMP5.csv\n",
      "chr2 VAMP8.csv\n",
      "chr2 VAX2.csv\n",
      "chr2 VIL1.csv\n",
      "chr2 VIT.csv\n",
      "chr2 VRK2.csv\n",
      "chr2 VSNL1.csv\n",
      "chr2 VWA3B.csv\n",
      "chr2 VWC2L-IT1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  16%|█▋        | 3170/19235 [00:11<01:08, 235.52it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 VWC2L.csv\n",
      "chr2 WBP1.csv\n",
      "chr2 WDR35-DT.csv\n",
      "chr2 WDR43.csv\n",
      "chr2 WDR54.csv\n",
      "chr2 WDR75.csv\n",
      "chr2 WNT10A.csv\n",
      "chr2 WNT6.csv\n",
      "chr2 XIRP2.csv\n",
      "chr2 XRCC5.csv\n",
      "chr2 YIPF4.csv\n",
      "chr2 YPEL5.csv\n",
      "chr2 ZAP70.csv\n",
      "chr2 ZC3H15.csv\n",
      "chr2 ZC3H6.csv\n",
      "chr2 ZDBF2.csv\n",
      "chr2 ZEB2-AS1.csv\n",
      "chr2 ZFAND2B.csv\n",
      "chr2 ZNF2.csv\n",
      "chr2 ZNF512.csv\n",
      "chr2 ZNF638.csv\n",
      "chr2 ZNF804A.csv\n",
      "chr2 ZNF806.csv\n",
      "chr3 AADAC.csv\n",
      "chr3 AADACL2.csv\n",
      "chr3 ABCC5-AS1.csv\n",
      "chr3 ABCF3.csv\n",
      "chr3 ABHD10.csv\n",
      "chr3 ABHD14A-ACY1.csv\n",
      "chr3 ABHD14A.csv\n",
      "chr3 ABHD5.csv\n",
      "chr3 ABHD6.csv\n",
      "chr3 ABTB1.csv\n",
      "chr3 ACAD9.csv\n",
      "chr3 ACKR2.csv\n",
      "chr3 ACKR4.csv\n",
      "chr3 ACP3.csv\n",
      "chr3 ACTL6A.csv\n",
      "chr3 ACVR2B.csv\n",
      "chr3 ACY1.csv\n",
      "chr3 ADAMTS9-AS1.csv\n",
      "chr3 ADAMTS9-AS2.csv\n",
      "chr3 ADGRG7.csv\n",
      "chr3 ADIPOQ.csv\n",
      "chr3 ADPRH.csv\n",
      "chr3 AGTR1.csv\n",
      "chr3 AHSG.csv\n",
      "chr3 ALAS1.csv\n",
      "chr3 ALCAM.csv\n",
      "chr3 ALDH1L1-AS1.csv\n",
      "chr3 ALDH1L1-AS2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  17%|█▋        | 3197/19235 [00:12<01:05, 243.89it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 ALG1L2.csv\n",
      "chr3 AP2M1.csv\n",
      "chr3 APEH.csv\n",
      "chr3 APPL1.csv\n",
      "chr3 APRG1.csv\n",
      "chr3 ARF4-AS1.csv\n",
      "chr3 ARGFX.csv\n",
      "chr3 ARHGAP31.csv\n",
      "chr3 ARHGEF26.csv\n",
      "chr3 ARHGEF3-AS1.csv\n",
      "chr3 ARIH2.csv\n",
      "chr3 ARL13B.csv\n",
      "chr3 ARL14.csv\n",
      "chr3 ARL6.csv\n",
      "chr3 ARL6IP5.csv\n",
      "chr3 ARL8B.csv\n",
      "chr3 ARMC8.csv\n",
      "chr3 ARPC4-TTLL3.csv\n",
      "chr3 ARPC4.csv\n",
      "chr3 ARPP21.csv\n",
      "chr3 ATG7.csv\n",
      "chr3 ATP11B.csv\n",
      "chr3 ATP13A3-DT.csv\n",
      "chr3 ATP13A4-AS1.csv\n",
      "chr3 ATP13A5-AS1.csv\n",
      "chr3 ATP1B3.csv\n",
      "chr3 ATP2C1.csv\n",
      "chr3 ATP6V1A.csv\n",
      "chr3 ATRIP-TREX1.csv\n",
      "chr3 ATRIP.csv\n",
      "chr3 ATXN7.csv\n",
      "chr3 B3GNT5.csv\n",
      "chr3 B4GALT4-AS1.csv\n",
      "chr3 BBX.csv\n",
      "chr3 BFSP2.csv\n",
      "chr3 BHLHE40.csv\n",
      "chr3 BOC.csv\n",
      "chr3 BPESC1.csv\n",
      "chr3 BRK1.csv\n",
      "chr3 BRPF1.csv\n",
      "chr3 BSN.csv\n",
      "chr3 BTD.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  17%|█▋        | 3246/19235 [00:12<01:15, 211.60it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 C3orf14.csv\n",
      "chr3 C3orf20.csv\n",
      "chr3 C3orf38.csv\n",
      "chr3 C3orf49.csv\n",
      "chr3 C3orf52.csv\n",
      "chr3 C3orf56.csv\n",
      "chr3 C3orf80.csv\n",
      "chr3 C3orf85.csv\n",
      "chr3 CACNA1D.csv\n",
      "chr3 CACNA2D3.csv\n",
      "chr3 CADM2.csv\n",
      "chr3 CAMP.csv\n",
      "chr3 CAND2.csv\n",
      "chr3 CAPN7.csv\n",
      "chr3 CASR.csv\n",
      "chr3 CAV3.csv\n",
      "chr3 CCDC13-AS1.csv\n",
      "chr3 CCDC13-AS2.csv\n",
      "chr3 CCDC174.csv\n",
      "chr3 CCDC50.csv\n",
      "chr3 CCDC54.csv\n",
      "chr3 CCDC66.csv\n",
      "chr3 CCR2.csv\n",
      "chr3 CCR3.csv\n",
      "chr3 CCR4.csv\n",
      "chr3 CCR5.csv\n",
      "chr3 CCR8.csv\n",
      "chr3 CCR9.csv\n",
      "chr3 CCRL2.csv\n",
      "chr3 CD200.csv\n",
      "chr3 CD200R1L-AS1.csv\n",
      "chr3 CD86.csv\n",
      "chr3 CD96.csv\n",
      "chr3 CDV3.csv\n",
      "chr3 CEP63.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  17%|█▋        | 3268/19235 [00:12<01:17, 205.13it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 CEP97.csv\n",
      "chr3 CFAP100.csv\n",
      "chr3 CFAP20DC-AS1.csv\n",
      "chr3 CFAP20DC-DT.csv\n",
      "chr3 CFAP44-AS1.csv\n",
      "chr3 CFAP91.csv\n",
      "chr3 CHCHD6.csv\n",
      "chr3 CHL1.csv\n",
      "chr3 CHMP2B.csv\n",
      "chr3 CHRD.csv\n",
      "chr3 CHST13.csv\n",
      "chr3 CHST2.csv\n",
      "chr3 CLDN11.csv\n",
      "chr3 CLDN16.csv\n",
      "chr3 CLDN18.csv\n",
      "chr3 CLEC3B.csv\n",
      "chr3 CLRN1-AS1.csv\n",
      "chr3 CLSTN2.csv\n",
      "chr3 CMC1.csv\n",
      "chr3 CMSS1.csv\n",
      "chr3 CMTM7.csv\n",
      "chr3 CMTM8.csv\n",
      "chr3 CNOT10.csv\n",
      "chr3 CNTN4.csv\n",
      "chr3 CNTN6.csv\n",
      "chr3 COL6A5.csv\n",
      "chr3 COL6A6.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  17%|█▋        | 3308/19235 [00:12<01:28, 179.77it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 COL8A1.csv\n",
      "chr3 COPB2-DT.csv\n",
      "chr3 COPG1.csv\n",
      "chr3 CPA3.csv\n",
      "chr3 CPB1.csv\n",
      "chr3 CPNE9.csv\n",
      "chr3 CRELD1.csv\n",
      "chr3 CRTAP.csv\n",
      "chr3 CRYBG3.csv\n",
      "chr3 CSNKA2IP.csv\n",
      "chr3 CSTA.csv\n",
      "chr3 CTDSPL.csv\n",
      "chr3 CTNNB1.csv\n",
      "chr3 CXCR6.csv\n",
      "chr3 CYB561D2.csv\n",
      "chr3 DAG1.csv\n",
      "chr3 DENND6A-DT.csv\n",
      "chr3 DHX30.csv\n",
      "chr3 DIPK2A.csv\n",
      "chr3 DLEC1.csv\n",
      "chr3 DLG1-AS1.csv\n",
      "chr3 DNAH1.csv\n",
      "chr3 DNAJB11.csv\n",
      "chr3 DNAJB8-AS1.csv\n",
      "chr3 DNAJC13.csv\n",
      "chr3 DOCK3.csv\n",
      "chr3 DTX3L.csv\n",
      "chr3 DUBR.csv\n",
      "chr3 DVL3.csv\n",
      "chr3 DZIP3.csv\n",
      "chr3 EAF1.csv\n",
      "chr3 EAF2.csv\n",
      "chr3 EBLN2.csv\n",
      "chr3 ECE2.csv\n",
      "chr3 ECT2.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  17%|█▋        | 3357/19235 [00:12<01:19, 200.24it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 EDEM1.csv\n",
      "chr3 EEF1AKMT4-ECE2.csv\n",
      "chr3 EEF1AKMT4.csv\n",
      "chr3 EEFSEC.csv\n",
      "chr3 EFCC1.csv\n",
      "chr3 EHHADH-AS1.csv\n",
      "chr3 EIF1B.csv\n",
      "chr3 EIF2A.csv\n",
      "chr3 EIF2B5.csv\n",
      "chr3 EIF4A2.csv\n",
      "chr3 EIF4G1.csv\n",
      "chr3 EMC3-AS1.csv\n",
      "chr3 ENTPD3.csv\n",
      "chr3 EPHA3.csv\n",
      "chr3 EPHA6.csv\n",
      "chr3 EPHB1.csv\n",
      "chr3 EPHB3.csv\n",
      "chr3 ERICH6-AS1.csv\n",
      "chr3 ESYT3.csv\n",
      "chr3 EXOG.csv\n",
      "chr3 EXOSC7.csv\n",
      "chr3 FAIM.csv\n",
      "chr3 FAM131A.csv\n",
      "chr3 FAM157A.csv\n",
      "chr3 FAM162A.csv\n",
      "chr3 FAM240A.csv\n",
      "chr3 FAM3D-AS1.csv\n",
      "chr3 FAM43A.csv\n",
      "chr3 FANCD2.csv\n",
      "chr3 FBLN2.csv\n",
      "chr3 FBXL2.csv\n",
      "chr3 FBXO40.csv\n",
      "chr3 FBXO45.csv\n",
      "chr3 FBXW12.csv\n",
      "chr3 FETUB.csv\n",
      "chr3 FGD5.csv\n",
      "chr3 FGF12-AS1.csv\n",
      "chr3 FGF12-AS2.csv\n",
      "chr3 FLJ42393.csv\n",
      "chr3 FLNB.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  18%|█▊        | 3382/19235 [00:13<01:14, 213.63it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 FNDC3B.csv\n",
      "chr3 FOXL2NB.csv\n",
      "chr3 FOXP1-AS1.csv\n",
      "chr3 FRG2C.csv\n",
      "chr3 FXR1.csv\n",
      "chr3 FYTTD1.csv\n",
      "chr3 GALNT15.csv\n",
      "chr3 GAP43.csv\n",
      "chr3 GASK1A.csv\n",
      "chr3 GATA2-AS1.csv\n",
      "chr3 GFM1.csv\n",
      "chr3 GHRLOS.csv\n",
      "chr3 GLYCTK.csv\n",
      "chr3 GMPS.csv\n",
      "chr3 GNAI2.csv\n",
      "chr3 GNAT1.csv\n",
      "chr3 GNL3.csv\n",
      "chr3 GOLGA4.csv\n",
      "chr3 GP9.csv\n",
      "chr3 GPD1L.csv\n",
      "chr3 GPR15.csv\n",
      "chr3 GPR160.csv\n",
      "chr3 GPR27.csv\n",
      "chr3 GPR62.csv\n",
      "chr3 GRAMD1C.csv\n",
      "chr3 GRK7.csv\n",
      "chr3 GRM2.csv\n",
      "chr3 GRM7.csv\n",
      "chr3 GSK3B-DT.csv\n",
      "chr3 GTF2E1.csv\n",
      "chr3 GTPBP8.csv\n",
      "chr3 GXYLT2.csv\n",
      "chr3 GYG1.csv\n",
      "chr3 H1-10-AS1.csv\n",
      "chr3 H1-8.csv\n",
      "chr3 HDAC11.csv\n",
      "chr3 HEMK1.csv\n",
      "chr3 HES1.csv\n",
      "chr3 HHATL-AS1.csv\n",
      "chr3 HHLA2.csv\n",
      "chr3 HLTF-AS1.csv\n",
      "chr3 HMCES.csv\n",
      "chr3 HPS3.csv\n",
      "chr3 HRG.csv\n",
      "chr3 HRH1.csv\n",
      "chr3 HTD2.csv\n",
      "chr3 HTR1F.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  18%|█▊        | 3431/19235 [00:13<01:13, 214.82it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 HTR3C.csv\n",
      "chr3 HTR3D.csv\n",
      "chr3 HTR3E.csv\n",
      "chr3 IFT122.csv\n",
      "chr3 IGF2BP2-AS1.csv\n",
      "chr3 IGSF11-AS1.csv\n",
      "chr3 IHO1.csv\n",
      "chr3 IL12A.csv\n",
      "chr3 IL17RB.csv\n",
      "chr3 IL17RC.csv\n",
      "chr3 IL17RE.csv\n",
      "chr3 IL1RAP.csv\n",
      "chr3 IL20RB.csv\n",
      "chr3 INKA1.csv\n",
      "chr3 IQCF2.csv\n",
      "chr3 IQCF3.csv\n",
      "chr3 IQCF5-AS1.csv\n",
      "chr3 IQCJ-SCHIP1.csv\n",
      "chr3 IQCJ.csv\n",
      "chr3 IRAK2.csv\n",
      "chr3 ITGA9.csv\n",
      "chr3 ITIH1.csv\n",
      "chr3 ITIH3.csv\n",
      "chr3 ITIH4-AS1.csv\n",
      "chr3 ITPR1.csv\n",
      "chr3 JAGN1.csv\n",
      "chr3 KALRN.csv\n",
      "chr3 KAT2B.csv\n",
      "chr3 KBTBD12.csv\n",
      "chr3 KBTBD8.csv\n",
      "chr3 KCNAB1.csv\n",
      "chr3 KCNH8.csv\n",
      "chr3 KCNMB2.csv\n",
      "chr3 KCTD6.csv\n",
      "chr3 KIF15.csv\n",
      "chr3 KIF9-AS1.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  18%|█▊        | 3477/19235 [00:13<01:13, 215.71it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 KLHDC8B.csv\n",
      "chr3 KLHL18.csv\n",
      "chr3 KLHL24.csv\n",
      "chr3 KLHL40.csv\n",
      "chr3 KLHL6-AS1.csv\n",
      "chr3 KNG1.csv\n",
      "chr3 KRBOX1.csv\n",
      "chr3 LARS2.csv\n",
      "chr3 LEKR1.csv\n",
      "chr3 LIMD1.csv\n",
      "chr3 LINC00312.csv\n",
      "chr3 LINC00488.csv\n",
      "chr3 LINC00501.csv\n",
      "chr3 LINC00506.csv\n",
      "chr3 LINC00578.csv\n",
      "chr3 LINC00620.csv\n",
      "chr3 LINC00636.csv\n",
      "chr3 LINC00690.csv\n",
      "chr3 LINC00693.csv\n",
      "chr3 LINC00698.csv\n",
      "chr3 LINC00852.csv\n",
      "chr3 LINC00870.csv\n",
      "chr3 LINC00879.csv\n",
      "chr3 LINC00881.csv\n",
      "chr3 LINC00885.csv\n",
      "chr3 LINC00888.csv\n",
      "chr3 LINC00901.csv\n",
      "chr3 LINC00903.csv\n",
      "chr3 LINC00960.csv\n",
      "chr3 LINC01014.csv\n",
      "chr3 LINC01100.csv\n",
      "chr3 LINC01205.csv\n",
      "chr3 LINC01206.csv\n",
      "chr3 LINC01210.csv\n",
      "chr3 LINC01213.csv\n",
      "chr3 LINC01215.csv\n",
      "chr3 LINC01266.csv\n",
      "chr3 LINC01322.csv\n",
      "chr3 LINC01327.csv\n",
      "chr3 LINC01487.csv\n",
      "chr3 LINC01811.csv\n",
      "chr3 LINC01839.csv\n",
      "chr3 LINC01968.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  18%|█▊        | 3530/19235 [00:13<01:06, 237.54it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 LINC01980.csv\n",
      "chr3 LINC01985.csv\n",
      "chr3 LINC01986.csv\n",
      "chr3 LINC01988.csv\n",
      "chr3 LINC01990.csv\n",
      "chr3 LINC01997.csv\n",
      "chr3 LINC02004.csv\n",
      "chr3 LINC02005.csv\n",
      "chr3 LINC02008.csv\n",
      "chr3 LINC02010.csv\n",
      "chr3 LINC02015.csv\n",
      "chr3 LINC02017.csv\n",
      "chr3 LINC02018.csv\n",
      "chr3 LINC02019.csv\n",
      "chr3 LINC02020.csv\n",
      "chr3 LINC02021.csv\n",
      "chr3 LINC02025.csv\n",
      "chr3 LINC02027.csv\n",
      "chr3 LINC02029.csv\n",
      "chr3 LINC02030.csv\n",
      "chr3 LINC02031.csv\n",
      "chr3 LINC02032.csv\n",
      "chr3 LINC02034.csv\n",
      "chr3 LINC02035.csv\n",
      "chr3 LINC02037.csv\n",
      "chr3 LINC02040.csv\n",
      "chr3 LINC02041.csv\n",
      "chr3 LINC02046.csv\n",
      "chr3 LINC02047.csv\n",
      "chr3 LINC02048.csv\n",
      "chr3 LINC02050.csv\n",
      "chr3 LINC02051.csv\n",
      "chr3 LINC02053.csv\n",
      "chr3 LINC02054.csv\n",
      "chr3 LINC02066.csv\n",
      "chr3 LINC02067.csv\n",
      "chr3 LINC02070.csv\n",
      "chr3 LINC02082.csv\n",
      "chr3 LINC02083.csv\n",
      "chr3 LINC02084.csv\n",
      "chr3 LINC02085.csv\n",
      "chr3 LINC02585.csv\n",
      "chr3 LINC02618.csv\n",
      "chr3 LINC02877.csv\n",
      "chr3 LINCR-0002.csv\n",
      "chr3 LMCD1.csv\n",
      "chr3 LMLN.csv\n",
      "chr3 LNP1.csv\n",
      "chr3 LOC100131635.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Counting sequence length:  19%|█▊        | 3568/19235 [00:13<01:00, 257.95it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3 LOC100287290.csv\n",
      "chr3 LOC100507389.csv\n",
      "chr3 LOC101926886.csv\n",
      "chr3 LOC101926923.csv\n",
      "chr3 LOC101926953.csv\n",
      "chr3 LOC101926968.csv\n",
      "chr3 LOC101927010.csv\n",
      "chr3 LOC101927296.csv\n",
      "chr3 LOC101927374.csv\n",
      "chr3 LOC101927439.csv\n",
      "chr3 LOC101927467.csv\n",
      "chr3 LOC101927518.csv\n",
      "chr3 LOC101927647.csv\n",
      "chr3 LOC101927829.csv\n",
      "chr3 LOC101927942.csv\n",
      "chr3 LOC101927995.csv\n",
      "chr3 LOC101928236.csv\n",
      "chr3 LOC101928263.csv\n",
      "chr3 LOC101928529.csv\n",
      "chr3 LOC101928636.csv\n",
      "chr3 LOC101928711.csv\n",
      "chr3 LOC101929130.csv\n",
      "chr3 LOC101929411.csv\n",
      "chr3 LOC102723430.csv\n",
      "chr3 LOC102723512.csv\n",
      "chr3 LOC102723596.csv\n",
      "chr3 LOC102724048.csv\n",
      "chr3 LOC102724104.csv\n",
      "chr3 LOC102724120.csv\n",
      "chr3 LOC102724289.csv\n",
      "chr3 LOC102724438.csv\n",
      "chr3 LOC102724479.csv\n",
      "chr3 LOC102724817.csv\n",
      "chr3 LOC102724877.csv\n",
      "chr3 LOC102724949.csv\n",
      "chr3 LOC105369194.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_6224/2610803250.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     18\u001b[0m         \u001b[0mgene_filename\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"gene\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m         \u001b[0mgene_filepath\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgene_dir\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mchr_name\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgene_filename\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 20\u001b[1;33m         \u001b[0mgene_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgene_filepath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     21\u001b[0m         \u001b[0mlength\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     22\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mgene_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\util\\_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    310\u001b[0m                 )\n\u001b[1;32m--> 311\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    312\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    313\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[0;32m    584\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    585\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 586\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    587\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    588\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    486\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    487\u001b[0m     \u001b[1;32mwith\u001b[0m \u001b[0mparser\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 488\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mparser\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    489\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    490\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36m__exit__\u001b[1;34m(self, exc_type, exc_value, traceback)\u001b[0m\n\u001b[0;32m   1078\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1079\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__exit__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexc_type\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexc_value\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtraceback\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1080\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1081\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1082\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36mclose\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    812\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    813\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 814\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    815\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    816\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_get_options_with_defaults\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\parsers\\c_parser_wrapper.py\u001b[0m in \u001b[0;36mclose\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    190\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    191\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m->\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 192\u001b[1;33m         \u001b[0msuper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    193\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    194\u001b[0m         \u001b[1;31m# close additional handles opened by C parser\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\parsers\\base_parser.py\u001b[0m in \u001b[0;36mclose\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    280\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    281\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 282\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    283\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    284\u001b[0m     \u001b[1;33m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\.virtualenv\\sequence-processing\\lib\\site-packages\\pandas\\io\\common.py\u001b[0m in \u001b[0;36mclose\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    110\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    111\u001b[0m             \u001b[1;32mfor\u001b[0m \u001b[0mhandle\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcreated_handles\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 112\u001b[1;33m                 \u001b[0mhandle\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    113\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mOSError\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    114\u001b[0m             \u001b[1;32mpass\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Enrich index with gene length.\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "index_dir = os.path.join(\"index\")\n",
    "gene_index = os.path.join(index_dir, \"gene_index.csv\")\n",
    "gene_train_index = os.path.join(index_dir, \"gene_train_index.csv\")\n",
    "gene_validation_index = os.path.join(index_dir, \"gene_validation_index.csv\")\n",
    "gene_test_index = os.path.join(index_dir, \"gene_test_index.csv\")\n",
    "gene_dir = os.path.join(\"data\", \"gene_dir\")\n",
    "\n",
    "for index in [gene_index, gene_train_index, gene_validation_index, gene_test_index]:\n",
    "    df = pd.read_csv(index)\n",
    "    len_sequences = []\n",
    "    for i, r in tqdm(df.iterrows(), total=df.shape[0], desc=\"Counting sequence length\"):\n",
    "        chr_name = r[\"chr\"]\n",
    "        gene_filename = r[\"gene\"]\n",
    "        gene_filepath = os.path.join(gene_dir, chr_name, gene_filename)\n",
    "        gene_df = pd.read_csv(gene_filepath)\n",
    "        length = 0\n",
    "        if gene_df.shape[0] > 1:\n",
    "            print(chr_name, gene_filename)\n",
    "        for i, j in gene_df.iterrows():\n",
    "            length += len(j[\"sequence\"])\n",
    "        len_sequences.append(length)\n",
    "    df[\"length\"] = len_sequences\n",
    "    df.to_csv(index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "_cols = ['chr', 'gene']\n",
    "df_path = os.path.join('workspace', 'sequential-labelling', 'gene_index.csv')\n",
    "_dir = os.path.join('workspace', 'sequential-labelling', 'duplicate genes')\n",
    "os.makedirs(_dir, exist_ok=True)\n",
    "df = pd.read_csv(df_path)\n",
    "train_df = pd.DataFrame(columns=_cols)\n",
    "valid_df = pd.DataFrame(columns=_cols)\n",
    "test_df = pd.DataFrame(columns=_cols)\n",
    "genes_unique = df['gene'].unique()\n",
    "for g in genes_unique:\n",
    "    filtered_df = df[df['gene'] == g]\n",
    "    if filtered_df.shape[0] > 1:\n",
    "        _g = \"{}.csv\".format(g)\n",
    "        filtered_df.to_csv(os.path.join(_dir, _g), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_preparation import gff_to_csvs, gff_to_csv\n",
    "from data_dir import (annotated_grch38_gff, annotated_grch38_gff_dir, annotated_grch38_gff_csv)\n",
    "\n",
    "print(annotated_grch38_gff)\n",
    "print(annotated_grch38_gff_csv)\n",
    "print(annotated_grch38_gff_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_dir import chr21_fasta, chr21_index_csv, data_genome_grch38_labels_dir\n",
    "from data_preparation import generate_sequence_labelling\n",
    "import os\n",
    "\n",
    "target_path = os.path.join(data_genome_grch38_labels_dir, 'chr21.csv')\n",
    "print(\"Generate sequential labelling {} => {}: {}\".format(chr21_index_csv, target_path, generate_sequence_labelling(chr21_index_csv, chr21_fasta, target_path)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_dir import (\n",
    "    chr1_index_csv, chr2_index_csv, chr3_index_csv, chr4_index_csv, chr5_index_csv, chr6_index_csv, chr7_index_csv, chr8_index_csv, chr9_index_csv, chr10_index_csv,\n",
    "    chr11_index_csv, chr12_index_csv, chr13_index_csv, chr14_index_csv, chr15_index_csv, chr16_index_csv, chr17_index_csv, chr18_index_csv, chr19_index_csv, chr20_index_csv,\n",
    "    chr21_index_csv, chr22_index_csv, chr23_index_csv, chr24_index_csv\n",
    ")\n",
    "from data_dir import (\n",
    "    chr1_fasta, chr2_fasta, chr3_fasta, chr4_fasta, chr5_fasta, chr6_fasta, chr7_fasta, chr8_fasta, chr9_fasta, chr10_fasta, \n",
    "\tchr11_fasta, chr12_fasta, chr13_fasta, chr14_fasta, chr15_fasta, chr16_fasta, chr17_fasta, chr18_fasta, chr19_fasta, chr20_fasta, \n",
    "\tchr21_fasta, chr22_fasta, chr23_fasta, chr24_fasta,\n",
    ")\n",
    "chr_fastas = [\n",
    "    chr1_fasta,\n",
    "\t#chr2_fasta, chr3_fasta, chr4_fasta, chr5_fasta, chr6_fasta, chr7_fasta, chr8_fasta, chr9_fasta, chr10_fasta,\n",
    "\t#chr11_fasta, chr12_fasta, chr13_fasta, chr14_fasta, chr15_fasta, chr16_fasta, chr17_fasta, chr18_fasta, chr19_fasta, chr20_fasta,\n",
    "\t#chr21_fasta, chr22_fasta, chr23_fasta, chr24_fasta\n",
    "]\n",
    "from data_dir import labseq_dir, labseq_names\n",
    "from data_preparation import generate_sequence_labelling\n",
    "chr_indices = [\n",
    "    chr1_index_csv, \n",
    "\t#chr2_index_csv, chr3_index_csv, chr4_index_csv, chr5_index_csv, chr6_index_csv, chr7_index_csv, chr8_index_csv, chr9_index_csv, chr10_index_csv,\n",
    "    #chr11_index_csv, chr12_index_csv, chr13_index_csv, chr14_index_csv, chr15_index_csv, chr16_index_csv, chr17_index_csv, chr18_index_csv, chr19_index_csv, chr20_index_csv,\n",
    "    #chr21_index_csv, chr22_index_csv, chr23_index_csv, chr24_index_csv\n",
    "]\n",
    "chr_labseq_path = [os.path.join(labseq_dir, fname) for fname in labseq_names[0:1]]\n",
    "for src, fasta, target in zip(chr_indices, chr_fastas, chr_labseq_path):\n",
    "    print(src, fasta, target)\n",
    "    #print(\"Generating sequential labelling for index {}, from fasta {}, to {}: {}\".format(src, fasta, target, generate_sequence_labelling(src, fasta, target, do_expand=True, expand_size=512)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sequential_labelling import Label_Dictionary\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "\n",
    "def prepare_sequence_from_csv(src_csv, label_dictionary=Label_Dictionary):\n",
    "    \"\"\"\n",
    "    Convert sequence into tokenized DNA sequence and label sequence.\n",
    "    CSV source has columns `sequence` and `label`. \n",
    "    `sequence` contains tokenized DNA sequence and `label` contains sequence of labels.\n",
    "    @param      src_csv (string): path to CSV source.\n",
    "    @param      label_dictionary (dict): dictionary to convert label into number.\n",
    "    @return     sequence, labels\n",
    "    \"\"\"\n",
    "    if not os.path.exists(src_csv):\n",
    "        raise FileNotFoundError(src_csv)\n",
    "    if label_dictionary == None:\n",
    "        raise Exception(\"Argument `label_dictionary` cannot be empty!\")\n",
    "    \n",
    "    df = pd.read_csv(src_csv)\n",
    "    return list(df['sequence']), list(df['label'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Create bundle from all expanded sequence from all genes from all chromosomes.\"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "src_dirs = [os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", f\"chr{i+1}.expanded\") for i in range(24)]\n",
    "bundle_path = os.path.join(\"data\", \"genome\", \"seqlab.positive.strand\", \"bundle.csv\")\n",
    "if os.path.exists(bundle_path):\n",
    "    os.remove(bundle_path)\n",
    "bundle = open(bundle_path, \"x\")\n",
    "bundle.write(\"sequence,label\\n\")\n",
    "for srcdir in tqdm(src_dirs, total=24):\n",
    "    files = [os.path.join(srcdir, f) for f in os.listdir(srcdir)]\n",
    "    for f in files:\n",
    "        df = pd.read_csv(f)\n",
    "        for i, r in df.iterrows():\n",
    "            bundle.write(f\"{r['sequence']},{r['label']}\\n\")\n",
    "bundle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 17 May 2022\n",
    "# From non overlapping kmer-ized genes, make bundle.\n",
    "import pandas as pd\n",
    "import os\n",
    "chrs = [f\"chr{i + 1}\" for i in range(24)]\n",
    "srcs = [os.path.join(\"workspace\", \"genlab\", \"seqlab.strand-positive.kmer.stride-510\", chr) for chr in chrs]\n",
    "dest = os.path.join(\"workspace\", \"seqlab\", \"seqlab.strand-positive.kmer.stride-510\", \"bundle.csv\")\n",
    "\n",
    "if os.path.exists(dest):\n",
    "    os.remove(dest)\n",
    "fdest = open(dest, \"x\")\n",
    "fdest.write(\"sequence,label\\n\")\n",
    "for src in srcs:\n",
    "    files = os.listdir(src)\n",
    "    files = [os.path.join(src, f) for f in files]\n",
    "    for f in files:\n",
    "        df = pd.read_csv(f)\n",
    "        for i,r in df.iterrows():\n",
    "            fdest.write(f\"{r['sequence']},{r['label']}\\n\")\n",
    "\n",
    "\n",
    "fdest.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bundling: 100%|██████████| 19235/19235 [11:28<00:00, 27.95it/s] \n"
     ]
    }
   ],
   "source": [
    "# Creating bundle from index.\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "gene_dir = os.path.join(\"data\", \"gene_dir_c510_k3\")\n",
    "index_dir = os.path.join(\"index\")\n",
    "#train_index = os.path.join(gene_dir, \"gene_train_index.csv\")\n",
    "#validation_index = os.path.join(gene_dir, \"gene_validation_index.csv\")\n",
    "#test_index = os.path.join(gene_dir, \"gene_test_index.csv\")\n",
    "whole_index = os.path.join(index_dir, \"gene_index.csv\")\n",
    "\n",
    "# Create training bundle from index.\n",
    "# Store file in by_sequence folder.\n",
    "bundle_path = os.path.join(\"workspace\", \"seqlab\", \"seqlab-3\")\n",
    "#train_bundle = os.path.join(bundle_path, \"gene_train_bundle.csv\")\n",
    "#validation_bundle = os.path.join(bundle_path, \"gene_validation_bundle.csv\")\n",
    "#test_bundle = os.path.join(bundle_path, \"gene_test_bundle.csv\")\n",
    "whole_bundle = os.path.join(bundle_path, \"gene_bundle.csv\")\n",
    "for s, d in zip([whole_index], [whole_bundle]):\n",
    "    df = pd.read_csv(s)\n",
    "    ddirname = os.path.dirname(d)\n",
    "    if not os.path.exists(ddirname):\n",
    "        os.makedirs(ddirname)\n",
    "    if os.path.exists(d):\n",
    "        os.remove(d)\n",
    "    dest_file = open(d, \"x\")\n",
    "    dest_file.write(\"sequence,label\\n\")\n",
    "    for i, r in tqdm(df.iterrows(), total=df.shape[0], desc=\"Bundling\"):\n",
    "        chr = r[\"chr\"]\n",
    "        gene = r[\"gene\"]\n",
    "        gene_path = os.path.join(gene_dir, chr, gene)\n",
    "        gene_df = pd.read_csv(gene_path)\n",
    "        for j, k in gene_df.iterrows():\n",
    "            dest_file.write(f\"{k['sequence']},{k['label']}\\n\")\n",
    "\n",
    "    dest_file.close()\n",
    "\n",
    "# 19235 genes seqlab\n",
    "# 19235 genes seqlab-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split gene_bundle into train, validation, and test set.\n",
    "import os \n",
    "import pandas as pd\n",
    "\n",
    "dirpath = os.path.join(\"workspace\", \"seqlab\", \"seqlab-3\")\n",
    "gene_bundle_path = os.path.join(dirpath, \"gene_bundle.csv\")\n",
    "gene_train_bundle_path = os.path.join(dirpath, \"gene_train_bundle.csv\")\n",
    "gene_validation_bundle_path = os.path.join(dirpath, \"gene_validation_bundle.csv\")\n",
    "gene_test_bundle_path = os.path.join(dirpath, \"gene_test_bundle.csv\")\n",
    "\n",
    "df = pd.read_csv(gene_bundle_path)\n",
    "train_df = df.sample(frac=0.8, random_state=1337)\n",
    "validation_df = df.drop(train_df.index)\n",
    "test_df = validation_df.sample(frac=0.5, random_state=1337)\n",
    "validation_df = validation_df.drop(test_df.index)\n",
    "train_df.to_csv(gene_train_bundle_path, index=False)\n",
    "validation_df.to_csv(gene_validation_bundle_path, index=False)\n",
    "test_df.to_csv(gene_test_bundle_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create small version of seqlab-3 sequence.\n",
    "import os \n",
    "import pandas as pd\n",
    "\n",
    "fractions = [0.1, 0.25]\n",
    "seqlab_3_dir = os.path.join(\"workspace\", \"seqlab\", \"seqlab-3\")\n",
    "train_df = pd.read_csv(os.path.join(seqlab_3_dir, \"gene_train_bundle.csv\"))\n",
    "validation_df = pd.read_csv(os.path.join(seqlab_3_dir, \"gene_validation_bundle.csv\"))\n",
    "test_df = pd.read_csv(os.path.join(seqlab_3_dir, \"gene_test_bundle.csv\"))\n",
    "\n",
    "for frac in fractions:\n",
    "    train_df.sample(frac=frac).to_csv(os.path.join(seqlab_3_dir, f\"gene_train_bundle.{frac * 100}.csv\"))\n",
    "    validation_df.sample(frac=frac).to_csv(os.path.join(seqlab_3_dir, f\"gene_validation_bundle.{frac * 100}.csv\"))\n",
    "    test_df.sample(frac=frac).to_csv(os.path.join(seqlab_3_dir, f\"gene_test_bundle.{frac * 100}.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing (2022-08-12 09:47:26.822747): 100%|██████████| 1753964/1753964 [13:33<00:00, 2156.62it/s]\n"
     ]
    }
   ],
   "source": [
    "# Generate SPLICE SITES, INTRON, and EXON only bundles.\n",
    "def at_least_one_exists(list, target_list):\n",
    "    # Check if at leats one element of list exists in target_list.\n",
    "    found = False\n",
    "    for elem in list:\n",
    "        if elem in target_list:\n",
    "            found = True\n",
    "    return found\n",
    "\n",
    "splice_sites = ['iiE', 'iEi', 'Eii', 'iEE', 'EEi', 'EiE']\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "gene_bundle_dirpath = os.path.join(\"workspace\", \"seqlab\", \"seqlab-3\")\n",
    "gene_bundle = os.path.join(gene_bundle_dirpath, \"gene_bundle.csv\")\n",
    "gene_ss_bundle = os.path.join(gene_bundle_dirpath, \"gene_ss_bundle.csv\")\n",
    "gene_exon_bundle = os.path.join(gene_bundle_dirpath, \"gene_exon_bundle.csv\")\n",
    "gene_intron_bundle = os.path.join(gene_bundle_dirpath, \"gene_intron_bundle.csv\")\n",
    "for p in [gene_bundle]:\n",
    "    df = pd.read_csv(p)\n",
    "    for a in [gene_ss_bundle, gene_exon_bundle, gene_intron_bundle]:\n",
    "        if os.path.exists(a):\n",
    "            os.remove(a)\n",
    "\n",
    "    gene_ss_bundle = open(gene_ss_bundle, \"x\")\n",
    "    gene_exon_bundle = open(gene_exon_bundle, \"x\")\n",
    "    gene_intron_bundle = open(gene_intron_bundle, \"x\")\n",
    "\n",
    "    for a in [gene_ss_bundle, gene_exon_bundle, gene_intron_bundle]:\n",
    "        a.write(\"sequence,label\\n\")\n",
    "\n",
    "    from datetime import datetime\n",
    "    cur_date = datetime.now()\n",
    "    for i, r in tqdm(df.iterrows(), total=df.shape[0], desc=f\"Processing ({cur_date})\"):\n",
    "        arr_labels = r[\"label\"].split(\" \")\n",
    "        if at_least_one_exists(splice_sites, arr_labels):\n",
    "            # raise NotImplementedError(\"TODO: write sequence into splice site bundle.\")\n",
    "            # Extract splice sites from gene_train_bundle.csv, gene_validation_bundle.csv, and gene_test_bundle.csv.\n",
    "            gene_ss_bundle.write(f\"{r['sequence']},{r['label']}\\n\")\n",
    "        elif all([a == \"iii\" for a in arr_labels]):\n",
    "            # raise NotImplementedError(\"TODO: write sequence into intron bundle.\")\n",
    "            # Extract introns.\n",
    "            gene_intron_bundle.write(f\"{r['sequence']},{r['label']}\\n\")\n",
    "        elif all([a == \"EEE\" for a in arr_labels]):\n",
    "            # raise NotImplementedError(\"TODO: write sequence into exon bundle.\")\n",
    "            # Extract exons.\n",
    "            gene_exon_bundle.write(f\"{r['sequence']},{r['label']}\\n\")\n",
    "\n",
    "    for a in [gene_ss_bundle, gene_intron_bundle, gene_exon_bundle]:\n",
    "        a.close()\n",
    "\n",
    "# Split SPLICE SITES bundle into train, validation, and test set.\n",
    "import os \n",
    "import pandas as pd\n",
    "\n",
    "dirpath = os.path.join(\"workspace\", \"seqlab\", \"seqlab-3\")\n",
    "splice_site_bundle_path = os.path.join(dirpath, \"gene_ss_bundle.csv\")\n",
    "splice_site_train_bundle_path = os.path.join(dirpath, \"gene_ss_train_bundle.csv\")\n",
    "splice_site_validation_bundle_path = os.path.join(dirpath, \"gene_ss_validation_bundle.csv\")\n",
    "splice_site_test_bundle_path = os.path.join(dirpath, \"gene_ss_test_bundle.csv\")\n",
    "\n",
    "df = pd.read_csv(splice_site_bundle_path)\n",
    "train_df = df.sample(frac=0.8, random_state=1337)\n",
    "validation_df = df.drop(train_df.index)\n",
    "test_df = validation_df.sample(frac=0.5, random_state=1337)\n",
    "validation_df = validation_df.drop(test_df.index)\n",
    "train_df.to_csv(splice_site_train_bundle_path, index=False)\n",
    "validation_df.to_csv(splice_site_validation_bundle_path, index=False)\n",
    "test_df.to_csv(splice_site_test_bundle_path, index=False)\n",
    "\n",
    "# 1753964 \n",
    "train_len = train_df.shape[0]\n",
    "validation_len = validation_df.shape[0]\n",
    "test_len = test_df.shape[0]\n",
    "total = train_len + validation_len + test_len\n",
    "print(f\"# Training instance {train_len} {train_len/total}\")\n",
    "print(f\"# Validation instance {validation_len} {validation_len/total}\")\n",
    "print(f\"# Test instance {test_len} {test_len/total}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_len = train_df.shape[0]\n",
    "validation_len = validation_df.shape[0]\n",
    "test_len = test_df.shape[0]\n",
    "total = train_len + validation_len + test_len\n",
    "print(f\"# Training instance {train_len} {train_len/total}\")\n",
    "print(f\"# Validation instance {validation_len} {validation_len/total}\")\n",
    "print(f\"# Test instance {test_len} {test_len/total}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene sequential labelling.\n",
    "# Create 10% and 25% sample of gene indices.\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "index_dir = os.path.join(\"index\")\n",
    "gene_train_index = os.path.join(index_dir, \"gene_train_index.csv\")\n",
    "gene_validation_index = os.path.join(index_dir, \"gene_validation_index.csv\")\n",
    "gene_test_index = os.path.join(index_dir, \"gene_test_index.csv\")\n",
    "\n",
    "gene_train_index_10 = os.path.join(index_dir, \"gene_train_index.10.csv\")\n",
    "gene_validation_index_10 = os.path.join(index_dir, \"gene_validation_index.10.csv\")\n",
    "gene_test_index_10 = os.path.join(index_dir, \"gene_test_index.10.csv\")\n",
    "\n",
    "gene_train_index_25 = os.path.join(index_dir, \"gene_train_index.25.csv\")\n",
    "gene_validation_index_25 = os.path.join(index_dir, \"gene_validation_index.25.csv\")\n",
    "gene_test_index_25 = os.path.join(index_dir, \"gene_test_index.25.csv\")\n",
    "\n",
    "srcs = [gene_train_index, gene_validation_index, gene_test_index]\n",
    "tens = [gene_train_index_10, gene_validation_index_10, gene_test_index_10]\n",
    "quarter = [gene_train_index_25, gene_validation_index_25, gene_test_index_25]\n",
    "\n",
    "for src, t, q in zip(srcs, tens, quarter):\n",
    "    src_df = pd.read_csv(src)\n",
    "    src_10_df = src_df.sample(frac=0.1, random_state=1337)\n",
    "    src_10_df.to_csv(t, index=False)\n",
    "    src_25_df = src_df.sample(frac=0.25, random_state=1337)\n",
    "    src_25_df.to_csv(q, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "def generate_bundle_with_marker(src_index, target_bundle, gene_dir):\n",
    "    gene_train_index = src_index\n",
    "    gene_train_bundle_csv = target_bundle\n",
    "    if os.path.exists(gene_train_bundle_csv):\n",
    "        os.remove(gene_train_bundle_csv)\n",
    "\n",
    "    target_bundle_dir = os.path.dirname(target_bundle)\n",
    "    os.makedirs(target_bundle_dir, exist_ok=True)\n",
    "\n",
    "    gene_train_bundle = open(gene_train_bundle_csv, \"x\")\n",
    "    gene_train_bundle.write(\"sequence,label,marker\\n\")\n",
    "\n",
    "    df = pd.read_csv(gene_train_index)\n",
    "    marker = True\n",
    "    for i, r in tqdm(df.iterrows(), total=df.shape[0], desc= f\"Processing {os.path.basename(src_index)}\"):\n",
    "        chr_dir = r[\"chr\"]\n",
    "        gene_file = r[\"gene\"]\n",
    "        gene_name = gene_file.split(\".\")[0]\n",
    "        gene_csv = os.path.join(gene_dir, chr_dir, gene_file)\n",
    "        gene_df = pd.read_csv(gene_csv)\n",
    "        for j, k in gene_df.iterrows():\n",
    "            sequence = k[\"sequence\"]\n",
    "            label = k[\"label\"]\n",
    "            gene_train_bundle.write(f\"{sequence},{label},{int(marker)}\\n\")\n",
    "        \n",
    "        marker = not marker\n",
    "\n",
    "    gene_train_bundle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene_train_index.10.csv: 100%|██████████| 1539/1539 [02:20<00:00, 10.98it/s]\n",
      "Processing gene_train_index.25.csv: 100%|██████████| 3847/3847 [03:28<00:00, 18.42it/s]\n",
      "Processing gene_train_index.csv: 100%|██████████| 15388/15388 [14:23<00:00, 17.83it/s]\n",
      "Processing gene_validation_index.10.csv: 100%|██████████| 192/192 [00:10<00:00, 18.67it/s]\n",
      "Processing gene_validation_index.25.csv: 100%|██████████| 481/481 [00:22<00:00, 21.73it/s]\n",
      "Processing gene_validation_index.csv: 100%|██████████| 1924/1924 [01:46<00:00, 18.13it/s]\n",
      "Processing gene_test_index.10.csv: 100%|██████████| 192/192 [00:12<00:00, 15.86it/s]\n",
      "Processing gene_test_index.25.csv: 100%|██████████| 481/481 [00:23<00:00, 20.22it/s]\n",
      "Processing gene_test_index.csv: 100%|██████████| 1923/1923 [01:38<00:00, 19.52it/s]\n"
     ]
    }
   ],
   "source": [
    "# Gene labelling: create bundle with marker that indicates a set of sequence belong to one long sequence.\n",
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "index_dir = os.path.join(\"index\")\n",
    "srcs = [\n",
    "    os.path.join(index_dir, \"gene_train_index.10.csv\"),\n",
    "    os.path.join(index_dir, \"gene_train_index.25.csv\"),\n",
    "    os.path.join(index_dir, \"gene_train_index.csv\"),\n",
    "    os.path.join(index_dir, \"gene_validation_index.10.csv\"),\n",
    "    os.path.join(index_dir, \"gene_validation_index.25.csv\"),\n",
    "    os.path.join(index_dir, \"gene_validation_index.csv\"),\n",
    "    os.path.join(index_dir, \"gene_test_index.10.csv\"),\n",
    "    os.path.join(index_dir, \"gene_test_index.25.csv\"),\n",
    "    os.path.join(index_dir, \"gene_test_index.csv\"),\n",
    "]\n",
    "workspace_dir = os.path.join(\"workspace\", \"genlab\", \"genlab-3\")\n",
    "dests = [\n",
    "    os.path.join(workspace_dir, \"gene_train_index_bundle.10.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_train_index_bundle.25.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_train_index_bundle.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_validation_index_bundle.10.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_validation_index_bundle.25.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_validation_index_bundle.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_test_index_bundle.10.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_test_index_bundle.25.csv\"),\n",
    "    os.path.join(workspace_dir, \"gene_test_index_bundle.csv\"),\n",
    "]\n",
    "# Generate gene bundles based on above.\n",
    "import os\n",
    "\n",
    "gene_dir = os.path.join(\"data\", \"gene_dir_c510_k3\")\n",
    "for a, b in zip(srcs, dests):\n",
    "    generate_bundle_with_marker(a, b, gene_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create gene sequence which contains splice sites at all position.\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "from utils.utils import is_exists_splice_site_in_sequence\n",
    "\n",
    "is_exists_splice_site_in_sequence([\"iii\", \"EEE\", \"EEE\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create gene sequence which contains splice sites at all position.\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "from tqdm import tqdm\n",
    "from utils.utils import kmer, str_kmer, is_exists_splice_site_in_sequence\n",
    "\n",
    "def generate_splice_site_all_pos_bundle(source_gene_dir, bundle_dest_dir, chunk_size, kmer_size):\n",
    "    \"\"\"\n",
    "    `source_gene_dir` - contains directories which each corresponds to chromosome.\n",
    "    Genes in chromosome folder is in raw format, not kmerized.\n",
    "    `bundle_dest_dir` - folder where the resulting bundle will be written.\n",
    "    \"\"\"\n",
    "    chr_names = os.listdir(source_gene_dir)\n",
    "    chr_dirs = [os.path.join(source_gene_dir, a) for a in chr_names]\n",
    "    chr_dirs = [a for a in chr_dirs if os.path.isdir(a)]\n",
    "\n",
    "    os.makedirs(bundle_dest_dir, exist_ok=True)\n",
    "    bundle_path = os.path.join(bundle_dest_dir, \"splice_site_all_pos.csv\")\n",
    "    if os.path.exists(bundle_path):\n",
    "        os.remove(bundle_path)\n",
    "    \n",
    "    bundle_file = open(bundle_path, \"x\")\n",
    "    bundle_file.write(\"sequence,label\\n\")\n",
    "\n",
    "    for d in tqdm(chr_dirs, total=len(chr_dirs), desc=\"Processing Chromosome\"):\n",
    "        filenames = os.listdir(d)\n",
    "        filepaths = [os.path.join(d, a) for a in filenames]\n",
    "        filepaths = [a for a in filepaths if os.path.isfile(a)]\n",
    "        \n",
    "        for f in filepaths:\n",
    "            df = pd.read_csv(f)\n",
    "            for i, r in df.iterrows():\n",
    "                sequence = r[\"sequence\"]\n",
    "                label = r[\"label\"]\n",
    "                len_sequence = len(sequence)\n",
    "                for i in range(0, len_sequence - chunk_size, 1):\n",
    "                    sublabel = label[i:i+chunk_size]\n",
    "                    arr_sublabel = kmer(sublabel, kmer_size)\n",
    "                    if is_exists_splice_site_in_sequence(arr_sublabel):\n",
    "                        subsequence = sequence[i:i+chunk_size]\n",
    "                        bundle_file.write(f\"{str_kmer(subsequence, kmer_size)},{' '.join(arr_sublabel)}\\n\")\n",
    "                    \n",
    "    bundle_file.close()                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Chromosome: 100%|██████████| 2/2 [08:54<00:00, 267.35s/it]\n"
     ]
    }
   ],
   "source": [
    "# Create sample index.\n",
    "import os \n",
    "gene_dir_path = os.path.join(\"data\", \"gene_dir_sample\")\n",
    "bundle_dest_dir = os.path.join(\"data\", \"gene_dir_sample\")\n",
    "chunk_size = 512\n",
    "kmer_size = 3\n",
    "generate_splice_site_all_pos_bundle(gene_dir_path, bundle_dest_dir, chunk_size, kmer_size)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "14ae8cb2141f3f34f4e0523006ff2d6cb0f7956c0f094e5497e312072e4d0d3c"
  },
  "kernelspec": {
   "display_name": "Python 3.7.9 64-bit ('sequence-processing': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "1bcc05615a70b396b2914747c544672e77157727153b1cb6572b3ac9e1c1c348"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
